// @abaplint/core/build/src/abap/types/basic/character_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CharacterType = void 0;
const _abstract_type_1 = require("./_abstract_type");
class CharacterType extends _abstract_type_1.AbstractType {
    constructor(length, extra) {
        super(extra);
        if (length <= 0) {
            throw new Error("Bad LENGTH, Character");
        }
        this.length = length;
    }
    cloneType(input) {
        const clone = Object.assign({}, this.getAbstractTypeData());
        if (input.qualifiedName) {
            clone.qualifiedName = input.qualifiedName;
        }
        if (input.ddicName) {
            clone.ddicName = input.ddicName;
        }
        if (input.derivedFromConstant) {
            clone.derivedFromConstant = input.derivedFromConstant;
        }
        return new CharacterType(this.length, clone);
    }
    getLength() {
        return this.length;
    }
    toText() {
        return "```c LENGTH " + this.getLength() + "```";
    }
    toABAP() {
        return "c LENGTH " + this.getLength();
    }
    isGeneric() {
        return false;
    }
    containsVoid() {
        return false;
    }
    getIdentifier() {
        return undefined;
    }
    toCDS() {
        return "abap.char( " + this.getLength() + " )";
    }
}
exports.CharacterType = CharacterType;
