// @abaplint/core/build/src/abap/types/basic/data_reference_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataReference = void 0;
const _abstract_type_1 = require("./_abstract_type");
const any_type_1 = require("./any_type");
class DataReference extends _abstract_type_1.AbstractType {
    constructor(type, qualifiedName) {
        super({ qualifiedName: qualifiedName });
        this.type = type;
    }
    toText(level) {
        return "Data REF TO " + this.type.toText(level + 1);
    }
    getType() {
        return this.type;
    }
    toABAP() {
        const type = this.type.toABAP();
        if (type.includes(" TABLE OF ")) {
            return ""; // hmm, should this return undefined?
        }
        return "REF TO " + this.type.toABAP();
    }
    isGeneric() {
        if (this.type instanceof any_type_1.AnyType) {
            return true;
        }
        return false;
    }
    containsVoid() {
        return this.type.containsVoid();
    }
    toCDS() {
        return "abap.TODO_REFERENCE";
    }
}
exports.DataReference = DataReference;
