// @abaplint/core/build/src/abap/types/basic/packed_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackedType = void 0;
const _abstract_type_1 = require("./_abstract_type");
class PackedType extends _abstract_type_1.AbstractType {
    constructor(length, decimals, extra) {
        super(extra);
        if (length <= 0) {
            throw new Error("Bad LENGTH, Packed");
        }
        else if (decimals < 0) {
            throw new Error("Bad DECIMALS, Packed");
        }
        this.length = length;
        this.decimals = decimals;
    }
    getLength() {
        return this.length;
    }
    getDecimals() {
        return this.decimals;
    }
    toText() {
        return "```p LENGTH " + this.getLength() + " DECIMALS " + this.getDecimals() + "```";
    }
    toABAP() {
        return "p LENGTH " + this.getLength() + " DECIMALS " + this.getDecimals();
    }
    isGeneric() {
        return false;
    }
    containsVoid() {
        return false;
    }
    toCDS() {
        return "abap.TODO_PACKED";
    }
}
exports.PackedType = PackedType;
