// @abaplint/core/build/src/abap/types/basic/string_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringType = void 0;
const _abstract_type_1 = require("./_abstract_type");
class StringType extends _abstract_type_1.AbstractType {
    static get(input) {
        const key = JSON.stringify(input);
        if (this.singletons.has(key)) {
            return this.singletons.get(key);
        }
        const ret = new StringType(input);
        this.singletons.set(key, ret);
        return ret;
    }
    constructor(input) {
        super(input);
    }
    /** fully qualified symbolic name of the type */
    getQualifiedName() {
        var _a;
        const qual = (_a = this.data) === null || _a === void 0 ? void 0 : _a.qualifiedName;
        if (qual === undefined) {
            return "STRING";
        }
        return qual;
    }
    toText() {
        return "```string```";
    }
    isGeneric() {
        return false;
    }
    toABAP() {
        return "string";
    }
    containsVoid() {
        return false;
    }
    toCDS() {
        return "abap.string";
    }
}
exports.StringType = StringType;
StringType.singletons = new Map();
