// @abaplint/core/build/src/abap/types/basic/structure_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructureType = void 0;
const _abstract_type_1 = require("./_abstract_type");
class StructureType extends _abstract_type_1.AbstractType {
    constructor(components, qualifiedName, ddicName, description) {
        super({
            qualifiedName: qualifiedName,
            ddicName: ddicName,
            description: description,
        });
        if (components.length === 0) {
            throw new Error("Structure does not contain any components");
        }
        this.indexed = {};
        for (const c of components) {
            const upper = c.name.toUpperCase();
            if (this.indexed[upper] !== undefined) {
                throw new Error("Structure, duplicate field name \"" + upper + "\", " + qualifiedName);
            }
            this.indexed[upper] = c.type;
        }
        this.components = components;
    }
    getComponents() {
        return this.components;
    }
    getComponentByName(name) {
        return this.indexed[name.toUpperCase()];
    }
    toText(level) {
        const compo = [];
        for (const c of this.components) {
            compo.push(c.name + " TYPE " + c.type.toText(level + 1));
        }
        const spaces = "  ".repeat(level);
        return "Structure\n" + spaces + "* " + compo.join("\n" + spaces + "* ");
    }
    isGeneric() {
        return false;
    }
    toABAP() {
        const ret = this.getQualifiedName();
        if (ret) {
            return ret;
        }
        return "StructureTypetoABAPtodo";
    }
    containsVoid() {
        return this.getComponents().some(c => { return c.type.containsVoid(); });
    }
    toCDS() {
        return "abap.TODO_STRUCTURE";
    }
}
exports.StructureType = StructureType;
