// @abaplint/core/build/src/abap/types/basic/table_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableType = exports.TableKeyType = exports.TableAccessType = void 0;
const _abstract_type_1 = require("./_abstract_type");
var TableAccessType;
(function (TableAccessType) {
    TableAccessType["standard"] = "STANDARD";
    TableAccessType["sorted"] = "SORTED";
    TableAccessType["hashed"] = "HASHED";
    TableAccessType["index"] = "INDEX";
    TableAccessType["any"] = "ANY";
})(TableAccessType || (exports.TableAccessType = TableAccessType = {}));
var TableKeyType;
(function (TableKeyType) {
    TableKeyType["default"] = "DEFAULT";
    TableKeyType["user"] = "USER";
    TableKeyType["empty"] = "EMPTY";
})(TableKeyType || (exports.TableKeyType = TableKeyType = {}));
class TableType extends _abstract_type_1.AbstractType {
    constructor(rowType, options, qualifiedName, description) {
        var _a;
        super({
            qualifiedName: qualifiedName,
            description: description,
        });
        this.rowType = rowType;
        this.options = options;
        if (((_a = options.primaryKey) === null || _a === void 0 ? void 0 : _a.type) === TableAccessType.standard && options.primaryKey.isUnique === true) {
            throw new Error("STANDARD tables cannot have UNIQUE key");
        }
    }
    getOptions() {
        return this.options;
    }
    isWithHeader() {
        return this.options.withHeader;
    }
    getAccessType() {
        var _a;
        return (_a = this.options.primaryKey) === null || _a === void 0 ? void 0 : _a.type;
    }
    getRowType() {
        return this.rowType;
    }
    toABAP() {
        // todo, this is used for downport, so use default key for now
        return "STANDARD TABLE OF " + this.rowType.toABAP() + " WITH DEFAULT KEY";
    }
    toText(level) {
        const type = this.rowType;
        if (this.options.withHeader === true) {
            return "Table with header of " + type.toText(level + 1);
        }
        else {
            return "Table of " + type.toText(level + 1);
        }
    }
    isGeneric() {
        var _a, _b;
        if (((_a = this.options.primaryKey) === null || _a === void 0 ? void 0 : _a.type) !== TableAccessType.standard
            && this.options.keyType === TableKeyType.user
            && ((_b = this.options.primaryKey) === null || _b === void 0 ? void 0 : _b.keyFields.length) === 0) {
            return true;
        }
        return this.rowType.isGeneric();
    }
    containsVoid() {
        return this.rowType.containsVoid();
    }
    toCDS() {
        return "abap.TODO_TABLE";
    }
}
exports.TableType = TableType;
