// @abaplint/core/build/src/abap/types/basic/void_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoidType = void 0;
const _abstract_type_1 = require("./_abstract_type");
class VoidType extends _abstract_type_1.AbstractType {
    static get(voided, qualifiedName) {
        const key = JSON.stringify({ voided, qualifiedName });
        if (this.singletons.has(key)) {
            return this.singletons.get(key);
        }
        const ret = new VoidType(voided, qualifiedName);
        this.singletons.set(key, ret);
        return ret;
    }
    constructor(voided, qualifiedName) {
        super({ qualifiedName: qualifiedName });
        this.voided = voided;
    }
    getVoided() {
        return this.voided;
    }
    toABAP() {
        return this.voided || "VOIDEDtoABAP";
    }
    toText() {
        return "Void(" + this.voided + ")";
    }
    isGeneric() {
        return false;
    }
    containsVoid() {
        return true;
    }
    toCDS() {
        return "abap.TODO_VOID";
    }
}
exports.VoidType = VoidType;
VoidType.singletons = new Map();
