// @abaplint/transpiler/build/src/handlers/handle_type_pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HandleTypePool = void 0;
const abaplint = require("@abaplint/core");
const chunk_1 = require("../chunk");
const transpile_types_1 = require("../transpile_types");
const traversal_1 = require("../traversal");
const structures_1 = require("../structures");
class HandleTypePool {
    runObject(obj, reg) {
        const spaghetti = new abaplint.SyntaxLogic(reg, obj).run().spaghetti;
        const spaghettiNode = spaghetti.getFirstChild()?.getFirstChild();
        if (spaghettiNode === undefined) {
            throw new Error("HandleTypePool: no spaghetti found");
        }
        const abapFile = obj.getABAPFiles()[0];
        if (abapFile === undefined) {
            throw new Error("HandleTypePool: no ABAP file found");
        }
        const chunk = new chunk_1.Chunk();
        chunk.appendString(`const pool = {};\n`);
        for (const v in spaghettiNode?.getData().vars) {
            const abs = spaghettiNode.getData().vars[v];
            const name = `pool['${v.toLowerCase()}']`;
            chunk.appendString(`${name} = ${transpile_types_1.TranspileTypes.toType(abs.getType())};\n`);
            chunk.appendString(traversal_1.Traversal.setValues(abs, name));
            // yea, this is a mess
            for (const cons of abapFile.getStructure()?.findAllStructures(abaplint.Structures.Constants) || []) {
                const cName = cons.findFirstExpression(abaplint.Expressions.DefinitionName)?.getFirstToken().getStr().toLowerCase();
                if (cName === v.toLocaleLowerCase()) {
                    chunk.appendString(structures_1.ConstantsTranspiler.handleValues(name, cons, new traversal_1.Traversal(spaghetti, abapFile, obj, reg)));
                }
            }
        }
        for (const t in spaghettiNode?.getData().types) {
            const abs = spaghettiNode.getData().types[t];
            chunk.appendString(`pool['${t.toLowerCase()}'] = ${transpile_types_1.TranspileTypes.toType(abs.getType())};\n`);
        }
        chunk.appendString(`abap.TypePools['${obj.getName()}'] = pool;`);
        const output = {
            object: {
                name: obj.getName(),
                type: obj.getType(),
            },
            filename: obj.getName().toLowerCase() + "." + obj.getType().toLowerCase() + ".mjs",
            chunk: chunk,
            requires: [],
            exports: [],
        };
        return [output];
    }
}
exports.HandleTypePool = HandleTypePool;
