// @abaplint/transpiler/build/src/handlers/handle_table_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HandleTableType = void 0;
const chunk_1 = require("../chunk");
const transpile_types_1 = require("../transpile_types");
class HandleTableType {
    runObject(obj, reg) {
        const filename = obj.getXMLFile()?.getFilename().replace(".xml", ".mjs").toLowerCase();
        if (filename === undefined) {
            return [];
        }
        const type = obj.parseType(reg);
        const chunk = new chunk_1.Chunk().appendString(`abap.DDIC["${obj.getName().toUpperCase()}"] = {
  "objectType": "TTYP",
  "type": ${transpile_types_1.TranspileTypes.toType(type)},
  "description": ${JSON.stringify(obj.getDescription())},
};`);
        const output = {
            object: {
                name: obj.getName(),
                type: obj.getType(),
            },
            filename: filename,
            chunk: chunk,
            requires: [],
            exports: [],
        };
        return [output];
    }
}
exports.HandleTableType = HandleTableType;
