// @abaplint/core/build/src/objects/_dynpros.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseDynpros = parseDynpros;
const xml_utils_1 = require("../xml_utils");
function parseDynpros(parsed) {
    var _a, _b, _c, _d;
    const dynpros = [];
    const xmlDynpros = (_c = (_b = (_a = parsed.abapGit) === null || _a === void 0 ? void 0 : _a["asx:abap"]) === null || _b === void 0 ? void 0 : _b["asx:values"]) === null || _c === void 0 ? void 0 : _c.DYNPROS;
    if (xmlDynpros !== undefined) {
        for (const d of (0, xml_utils_1.xmlToArray)(xmlDynpros.item)) {
            const fields = [];
            for (const f of (0, xml_utils_1.xmlToArray)((_d = d.FIELDS) === null || _d === void 0 ? void 0 : _d.RPY_DYFATC)) {
                fields.push({
                    name: f.NAME,
                    type: f.TYPE,
                    length: f.LENGTH,
                });
            }
            dynpros.push({
                number: d.HEADER.SCREEN,
                description: d.HEADER.DESCRIPT,
                nextScreen: d.HEADER.NEXTSCREEN,
                type: d.HEADER.TYPE,
                fields: fields,
            });
        }
    }
    return dynpros;
}
