// @abaplint/core/build/src/objects/_unknown_object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownObject = void 0;
const _abstract_object_1 = require("./_abstract_object");
const issue_1 = require("../issue");
const position_1 = require("../position");
const severity_1 = require("../severity");
class UnknownObject extends _abstract_object_1.AbstractObject {
    constructor(name, type) {
        super(name);
        this.type = type;
    }
    getType() {
        return this.type;
    }
    getAllowedNaming() {
        return {
            maxLength: 100,
            allowNamespace: true,
        };
    }
    getDescription() {
        return undefined;
    }
    getParsingIssues() {
        const pos = new position_1.Position(1, 1);
        const file = this.getFiles()[0];
        const message = "Unknown object type, currently not supported in abaplint, open issue on github";
        const issue = issue_1.Issue.atPosition(file, pos, message, "parser_error", severity_1.Severity.Error);
        return [issue];
    }
}
exports.UnknownObject = UnknownObject;
