// @abaplint/core/build/src/objects/authorization_check_field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthorizationCheckField = void 0;
const basic_1 = require("../abap/types/basic");
const ddic_1 = require("../ddic");
const _abstract_object_1 = require("./_abstract_object");
class AuthorizationCheckField extends _abstract_object_1.AbstractObject {
    getType() {
        return "AUTH";
    }
    getAllowedNaming() {
        return {
            maxLength: 30,
            allowNamespace: true,
        };
    }
    setDirty() {
        this.parsedXML = undefined;
        super.setDirty();
    }
    getDescription() {
        // todo
        return undefined;
    }
    getDataElementName() {
        var _a;
        this.parse();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.rollname;
    }
    parseType(reg) {
        var _a;
        this.parse();
        const references = [];
        const ddic = new ddic_1.DDIC(reg);
        if ((_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.rollname) {
            const found = ddic.lookupDataElement(this.parsedXML.rollname);
            if (found.object) {
                references.push({ object: found.object });
                reg.getDDICReferences().setUsing(this, references);
                return found.type;
            }
            else if (ddic.inErrorNamespace(this.parsedXML.rollname)) {
                return new basic_1.UnknownType(this.parsedXML.rollname + " not found");
            }
            else {
                return basic_1.VoidType.get(this.parsedXML.rollname);
            }
        }
        else {
            return new basic_1.UnknownType("Parsing error");
        }
    }
    parse() {
        var _a;
        if (this.parsedXML) {
            return { updated: false, runtime: 0 };
        }
        const start = Date.now();
        this.parsedXML = {};
        const parsed = super.parseRaw2();
        if (parsed === undefined
            || parsed.abapGit === undefined
            || parsed.abapGit["asx:abap"]["asx:values"] === undefined) {
            return { updated: false, runtime: 0 };
        }
        this.parsedXML.rollname = (_a = parsed.abapGit["asx:abap"]["asx:values"].AUTHX) === null || _a === void 0 ? void 0 : _a.ROLLNAME;
        const end = Date.now();
        return { updated: true, runtime: end - start };
    }
}
exports.AuthorizationCheckField = AuthorizationCheckField;
