// @abaplint/core/build/src/objects/enhancement_implementation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnhancementImplementation = void 0;
const _abstract_object_1 = require("./_abstract_object");
class EnhancementImplementation extends _abstract_object_1.AbstractObject {
    getType() {
        return "ENHO";
    }
    setDirty() {
        this.parsedXML = undefined;
        super.setDirty();
    }
    getAllowedNaming() {
        return {
            maxLength: 30,
            allowNamespace: true,
        };
    }
    getClassName() {
        var _a;
        this.parse();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.className;
    }
    getDescription() {
        var _a;
        this.parse();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.description;
    }
    parse() {
        if (this.parsedXML) {
            return { updated: false, runtime: 0 };
        }
        const start = Date.now();
        this.parsedXML = {};
        const parsed = super.parseRaw2();
        if (parsed === undefined
            || parsed.abapGit === undefined
            || parsed.abapGit["asx:abap"]["asx:values"] === undefined) {
            return { updated: false, runtime: 0 };
        }
        this.parsedXML.className = parsed.abapGit["asx:abap"]["asx:values"].CLASS;
        this.parsedXML.description = parsed.abapGit["asx:abap"]["asx:values"].SHORTTEXT;
        const end = Date.now();
        return { updated: true, runtime: end - start };
    }
}
exports.EnhancementImplementation = EnhancementImplementation;
