// @abaplint/core/build/src/objects/enhancement_spot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnhancementSpot = void 0;
const _abstract_object_1 = require("./_abstract_object");
const xml_utils_1 = require("../xml_utils");
class EnhancementSpot extends _abstract_object_1.AbstractObject {
    getType() {
        return "ENHS";
    }
    getDescription() {
        // todo
        return undefined;
    }
    getAllowedNaming() {
        return {
            maxLength: 30,
            allowNamespace: true,
        };
    }
    setDirty() {
        this.badis = undefined;
        super.setDirty();
    }
    listBadiDefinitions() {
        if (this.badis === undefined) {
            this.badis = this.parseXML();
        }
        return this.badis;
    }
    /////////////////
    parseXML() {
        var _a, _b;
        const parsed = super.parseRaw2();
        if (parsed === undefined) {
            return [];
        }
        const ret = [];
        for (const b of (0, xml_utils_1.xmlToArray)((_b = (_a = parsed.abapGit["asx:abap"]["asx:values"]) === null || _a === void 0 ? void 0 : _a.BADI_DATA) === null || _b === void 0 ? void 0 : _b.ENH_BADI_DATA)) {
            ret.push({
                name: b.BADI_NAME,
                interface: b.INTERFACE_NAME,
            });
        }
        return ret;
    }
}
exports.EnhancementSpot = EnhancementSpot;
