// @abaplint/core/build/src/objects/extension_index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionIndex = void 0;
const _abstract_object_1 = require("./_abstract_object");
class ExtensionIndex extends _abstract_object_1.AbstractObject {
    constructor() {
        super(...arguments);
        this.parsedXML = undefined;
    }
    getType() {
        return "XINX";
    }
    getAllowedNaming() {
        return {
            maxLength: 33,
            allowNamespace: true,
        };
    }
    setDirty() {
        this.parsedXML = undefined;
        super.setDirty();
    }
    getDescription() {
        var _a;
        this.parse();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.ddtext;
    }
    getTableName() {
        var _a;
        this.parse();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.sqltab;
    }
    parse() {
        var _a, _b, _c, _d, _e;
        if (this.parsedXML !== undefined) {
            return { updated: false, runtime: 0 };
        }
        const start = Date.now();
        this.parsedXML = {};
        const parsed = super.parseRaw2();
        if (parsed === undefined) {
            return { updated: false, runtime: 0 };
        }
        const xinx = (_c = (_b = (_a = parsed.abapGit) === null || _a === void 0 ? void 0 : _a["asx:abap"]) === null || _b === void 0 ? void 0 : _b["asx:values"]) === null || _c === void 0 ? void 0 : _c.XINX;
        this.parsedXML = {
            sqltab: (_d = xinx === null || xinx === void 0 ? void 0 : xinx.DD12V) === null || _d === void 0 ? void 0 : _d.SQLTAB,
            ddtext: (_e = xinx === null || xinx === void 0 ? void 0 : xinx.DD12V) === null || _e === void 0 ? void 0 : _e.DDTEXT,
        };
        const end = Date.now();
        return { updated: true, runtime: end - start };
    }
}
exports.ExtensionIndex = ExtensionIndex;
