// @abaplint/core/build/src/objects/lock_object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LockObject = void 0;
const basic_1 = require("../abap/types/basic");
const ddic_1 = require("../ddic");
const _abstract_object_1 = require("./_abstract_object");
class LockObject extends _abstract_object_1.AbstractObject {
    getType() {
        return "ENQU";
    }
    getAllowedNaming() {
        return {
            maxLength: 16,
            customRegex: /^(\/[A-Z_\d]{3,8}\/)?E[A-Z0-9_]+$/i,
            allowNamespace: true,
        };
    }
    setDirty() {
        this.parsedXML = undefined;
        super.setDirty();
    }
    getPrimaryTable() {
        var _a;
        this.parse();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.primaryTable;
    }
    parseType(reg) {
        var _a;
        this.parse();
        const references = [];
        const ddic = new ddic_1.DDIC(reg);
        if ((_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.primaryTable) {
            const found = ddic.lookupTableOrView2(this.parsedXML.primaryTable);
            if (found) {
                references.push({ object: found });
                reg.getDDICReferences().setUsing(this, references);
                return found.parseType(reg);
            }
            else if (ddic.inErrorNamespace(this.parsedXML.primaryTable)) {
                return new basic_1.UnknownType(this.parsedXML.primaryTable + " not found");
            }
            else {
                return basic_1.VoidType.get(this.parsedXML.primaryTable);
            }
        }
        else {
            return new basic_1.UnknownType("Parsing error");
        }
    }
    parse() {
        var _a, _b;
        if (this.parsedXML) {
            return { updated: false, runtime: 0 };
        }
        const start = Date.now();
        this.parsedXML = {};
        const parsed = super.parseRaw2();
        if (parsed === undefined
            || parsed.abapGit === undefined
            || parsed.abapGit["asx:abap"]["asx:values"] === undefined) {
            return { updated: false, runtime: 0 };
        }
        this.parsedXML.primaryTable = (_a = parsed.abapGit["asx:abap"]["asx:values"].DD25V) === null || _a === void 0 ? void 0 : _a.ROOTTAB;
        this.parsedXML.description = (_b = parsed.abapGit["asx:abap"]["asx:values"].DD25V) === null || _b === void 0 ? void 0 : _b.DDTEXT;
        const end = Date.now();
        return { updated: true, runtime: end - start };
    }
    getDescription() {
        var _a;
        this.parse();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.description;
    }
}
exports.LockObject = LockObject;
