// @abaplint/core/build/src/objects/table_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableType = void 0;
const _abstract_object_1 = require("./_abstract_object");
const Types = require("../abap/types/basic");
const ddic_1 = require("../ddic");
const basic_1 = require("../abap/types/basic");
const xml_utils_1 = require("../xml_utils");
class TableType extends _abstract_object_1.AbstractObject {
    constructor() {
        super(...arguments);
        this.parsedXML = undefined;
    }
    getType() {
        return "TTYP";
    }
    getAllowedNaming() {
        return {
            maxLength: 30,
            allowNamespace: true,
        };
    }
    getDescription() {
        var _a;
        this.parseXML();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.ddtext;
    }
    getRowType() {
        var _a;
        this.parseXML();
        return (_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.rowtype;
    }
    setDirty() {
        this.parsedXML = undefined;
        super.setDirty();
    }
    buildPrimaryKey() {
        var _a, _b, _c, _d, _e;
        const primaryKey = {
            isUnique: ((_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.keykind) === "U",
            type: basic_1.TableAccessType.standard,
            keyFields: [],
            name: "primary_key",
        };
        if (((_b = this.parsedXML) === null || _b === void 0 ? void 0 : _b.accessmode) === "S") {
            primaryKey.type = basic_1.TableAccessType.sorted;
        }
        else if (((_c = this.parsedXML) === null || _c === void 0 ? void 0 : _c.accessmode) === "H") {
            primaryKey.type = basic_1.TableAccessType.hashed;
        }
        for (const f of ((_d = this.parsedXML) === null || _d === void 0 ? void 0 : _d.dd42v) || []) {
            if (f.keyname === "") {
                primaryKey.keyFields.push(f.keyfield);
            }
        }
        if (((_e = this.parsedXML) === null || _e === void 0 ? void 0 : _e.keydef) === "T" && primaryKey.keyFields.length === 0) {
            primaryKey.keyFields.push("table_line");
        }
        return primaryKey;
    }
    buildTableOptions() {
        var _a, _b, _c, _d;
        let keyType = Types.TableKeyType.user;
        if (((_a = this.parsedXML) === null || _a === void 0 ? void 0 : _a.keydef) === "D") {
            keyType = Types.TableKeyType.default;
        }
        const tableOptions = {
            withHeader: false,
            keyType: keyType,
            primaryKey: this.buildPrimaryKey(),
            secondary: [],
        };
        for (const k of ((_b = this.parsedXML) === null || _b === void 0 ? void 0 : _b.dd43v) || []) {
            const fields = [];
            for (const f of ((_c = this.parsedXML) === null || _c === void 0 ? void 0 : _c.dd42v) || []) {
                if (f.keyname === k.keyname) {
                    fields.push(f.keyfield);
                }
            }
            let accessType = basic_1.TableAccessType.standard;
            switch (k.accessmode) {
                case "S":
                    accessType = basic_1.TableAccessType.sorted;
                    break;
                case "H":
                    accessType = basic_1.TableAccessType.hashed;
                    break;
                default:
                    break;
            }
            (_d = tableOptions.secondary) === null || _d === void 0 ? void 0 : _d.push({
                name: k.keyname,
                type: accessType,
                keyFields: fields,
                isUnique: k.unique,
            });
        }
        return tableOptions;
    }
    parseType(reg) {
        this.parseXML();
        const ddic = new ddic_1.DDIC(reg);
        const references = [];
        let type;
        const tableOptions = this.buildTableOptions();
        if (this.parsedXML === undefined) {
            type = new Types.UnknownType("Table Type, parser error", this.getName());
        }
        else if (this.parsedXML.rowkind === "S") {
            const lookup = ddic.lookupTableOrView(this.parsedXML.rowtype);
            type = new Types.TableType(lookup.type, tableOptions, this.getName(), this.getDescription());
            if (lookup.object) {
                references.push({ object: lookup.object });
            }
        }
        else if (this.parsedXML.rowkind === "E") {
            const lookup = ddic.lookupDataElement(this.parsedXML.rowtype);
            type = new Types.TableType(lookup.type, tableOptions, this.getName(), this.getDescription());
            if (lookup.object) {
                references.push({ object: lookup.object });
            }
        }
        else if (this.parsedXML.rowkind === "L") {
            const lookup = ddic.lookupTableType(this.parsedXML.rowtype);
            type = new Types.TableType(lookup.type, tableOptions, this.getName(), this.getDescription());
            if (lookup.object) {
                references.push({ object: lookup.object });
            }
        }
        else if (this.parsedXML.rowkind === "R" && this.parsedXML.rowtype === "OBJECT") {
            type = new Types.TableType(new basic_1.GenericObjectReferenceType(), tableOptions, this.getName(), this.getDescription());
        }
        else if (this.parsedXML.rowkind === "R" && this.parsedXML.rowtype === "DATA") {
            type = new Types.TableType(new basic_1.DataReference(new Types.DataType()), tableOptions, this.getName(), this.getDescription());
        }
        else if (this.parsedXML.rowkind === "R" && this.parsedXML.rowtype !== undefined) {
            const lookup = ddic.lookupObject(this.parsedXML.rowtype);
            type = new Types.TableType(lookup.type, tableOptions, this.getName(), this.getDescription());
            if (lookup.object) {
                references.push({ object: lookup.object });
            }
        }
        else if (this.parsedXML.rowkind === "") {
            if (this.parsedXML.datatype === undefined) {
                type = new Types.UnknownType("Table Type, empty DATATYPE" + this.getName(), this.getName());
            }
            else {
                const row = ddic.textToType({
                    text: this.parsedXML.datatype,
                    length: this.parsedXML.leng,
                    decimals: this.parsedXML.decimals,
                    infoText: this.getName(),
                });
                type = new Types.TableType(row, tableOptions, this.getName(), this.getDescription());
            }
        }
        else {
            type = new Types.UnknownType("Table Type, unknown kind \"" + this.parsedXML.rowkind + "\"" + this.getName(), this.getName());
        }
        reg.getDDICReferences().setUsing(this, references);
        return type;
    }
    ////////////////////
    parseXML() {
        var _a, _b;
        if (this.parsedXML !== undefined) {
            return;
        }
        this.parsedXML = {
            dd42v: [],
            dd43v: [],
        };
        const parsed = super.parseRaw2();
        if (parsed === undefined || parsed.abapGit === undefined) {
            return;
        }
        const values = parsed.abapGit["asx:abap"]["asx:values"];
        if (values === undefined) {
            return;
        }
        const dd40v = values.DD40V;
        this.parsedXML.rowtype = dd40v.ROWTYPE ? dd40v.ROWTYPE : "";
        this.parsedXML.rowkind = dd40v.ROWKIND ? dd40v.ROWKIND : "";
        this.parsedXML.datatype = dd40v.DATATYPE;
        this.parsedXML.leng = dd40v.LENG;
        this.parsedXML.decimals = dd40v.DECIMALS;
        this.parsedXML.accessmode = dd40v.ACCESSMODE;
        this.parsedXML.keykind = dd40v.KEYKIND;
        this.parsedXML.ddtext = dd40v.DDTEXT;
        this.parsedXML.keydef = dd40v.KEYDEF;
        for (const x of (0, xml_utils_1.xmlToArray)((_a = values.DD42V) === null || _a === void 0 ? void 0 : _a.DD42V)) {
            this.parsedXML.dd42v.push({
                keyname: x.SECKEYNAME || "",
                keyfield: x.KEYFIELD || "",
            });
        }
        for (const x of (0, xml_utils_1.xmlToArray)((_b = values.DD43V) === null || _b === void 0 ? void 0 : _b.DD43V)) {
            this.parsedXML.dd43v.push({
                keyname: x.SECKEYNAME || "",
                accessmode: x.ACCESSMODE || "",
                kind: x.KIND || "",
                unique: x.SECKEYUNIQUE === "X",
            });
        }
    }
}
exports.TableType = TableType;
