// @abaplint/core/build/src/abap/2_statements/statements/_statement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeSQL = exports.MacroContent = exports.MacroCall = exports.Empty = exports.Comment = exports.Unknown = void 0;
class Unknown {
    getMatcher() {
        throw new Error("Unknown Statement, get_matcher");
    }
}
exports.Unknown = Unknown;
class Comment {
    getMatcher() {
        throw new Error("Comment Statement, get_matcher");
    }
}
exports.Comment = Comment;
class Empty {
    getMatcher() {
        throw new Error("Empty Statement, get_matcher");
    }
}
exports.Empty = Empty;
class MacroCall {
    getMatcher() {
        throw new Error("MacroCall Statement, get_matcher");
    }
}
exports.MacroCall = MacroCall;
class MacroContent {
    getMatcher() {
        throw new Error("MacroContent Statement, get_matcher");
    }
}
exports.MacroContent = MacroContent;
class NativeSQL {
    getMatcher() {
        throw new Error("NativeSQL Statement, get_matcher");
    }
}
exports.NativeSQL = NativeSQL;
