// @abaplint/core/build/src/abap/4_file_information/_abap_file_information.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RiskLevel = exports.Duration = exports.MethodParameterDirection = exports.AttributeLevel = void 0;
// Only helper functions to get data from single file, no type information
var AttributeLevel;
(function (AttributeLevel) {
    AttributeLevel["Instance"] = "instance";
    AttributeLevel["Static"] = "static";
    AttributeLevel["Constant"] = "constant";
})(AttributeLevel || (exports.AttributeLevel = AttributeLevel = {}));
var MethodParameterDirection;
(function (MethodParameterDirection) {
    MethodParameterDirection["Importing"] = "importing";
    MethodParameterDirection["Exporting"] = "exporting";
    MethodParameterDirection["Changing"] = "changing";
    MethodParameterDirection["Returning"] = "returning";
})(MethodParameterDirection || (exports.MethodParameterDirection = MethodParameterDirection = {}));
var Duration;
(function (Duration) {
    Duration["short"] = "SHORT";
    Duration["medium"] = "MEDIUM";
    Duration["long"] = "LONG";
})(Duration || (exports.Duration = Duration = {}));
var RiskLevel;
(function (RiskLevel) {
    RiskLevel["harmless"] = "HARMLESS";
    RiskLevel["critical"] = "CRITICAL";
    RiskLevel["dangerous"] = "DANGEROUS";
})(RiskLevel || (exports.RiskLevel = RiskLevel = {}));
