// @abaplint/core/build/src/abap/4_file_information/abap_file_information.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ABAPFileInformation = void 0;
class ABAPFileInformation {
    constructor(parsed) {
        this.parsed = parsed;
    }
    listClassImplementations() {
        return this.parsed.implementations;
    }
    listInterfaceDefinitions() {
        return this.parsed.interfaces;
    }
    getInterfaceDefinitionByName(name) {
        const upper = name.toUpperCase();
        for (const i of this.listInterfaceDefinitions()) {
            if (i.identifier.getName().toUpperCase() === upper) {
                return i;
            }
        }
        return undefined;
    }
    listClassDefinitions() {
        return this.parsed.classes;
    }
    getClassDefinitionByName(name) {
        const upper = name.toUpperCase();
        for (const d of this.listClassDefinitions()) {
            if (d.identifier.getName().toUpperCase() === upper) {
                return d;
            }
        }
        return undefined;
    }
    getClassImplementationByName(name) {
        const upper = name.toUpperCase();
        for (const impl of this.listClassImplementations()) {
            if (impl.identifier.getName().toUpperCase() === upper) {
                return impl;
            }
        }
        return undefined;
    }
    listFormDefinitions() {
        return this.parsed.forms;
    }
}
exports.ABAPFileInformation = ABAPFileInformation;
