// @abaplint/runtime/build/src/statements/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Statements = void 0;
const append_1 = require("./append");
const assert_1 = require("./assert");
const assign_1 = require("./assign");
const commit_1 = require("./commit");
const concatenate_1 = require("./concatenate");
const condense_1 = require("./condense");
const convert_1 = require("./convert");
const create_data_1 = require("./create_data");
const delete_internal_1 = require("./delete_internal");
const describe_1 = require("./describe");
const find_1 = require("./find");
const collect_1 = require("./collect");
const overlay_1 = require("./overlay");
const cast_1 = require("./cast");
const get_bit_1 = require("./get_bit");
const read_report_1 = require("./read_report");
const get_reference_1 = require("./get_reference");
const raise_event_1 = require("./raise_event");
const receive_1 = require("./receive");
const get_locale_1 = require("./get_locale");
const unpack_1 = require("./unpack");
const get_parameter_1 = require("./get_parameter");
const set_locale_1 = require("./set_locale");
const get_run_time_1 = require("./get_run_time");
const get_time_1 = require("./get_time");
const insert_database_1 = require("./insert_database");
const insert_internal_1 = require("./insert_internal");
const delete_database_1 = require("./delete_database");
const loop_1 = require("./loop");
const message_1 = require("./message");
const modify_database_1 = require("./modify_database");
const modify_internal_1 = require("./modify_internal");
const move_corresponding_1 = require("./move_corresponding");
const read_table_1 = require("./read_table");
const replace_1 = require("./replace");
const rollback_1 = require("./rollback");
const select_1 = require("./select");
const set_bit_1 = require("./set_bit");
const shift_1 = require("./shift");
const sort_1 = require("./sort");
const wait_1 = require("./wait");
const fetch_next_cursor_1 = require("./fetch_next_cursor");
const open_cursor_1 = require("./open_cursor");
const close_cursor_1 = require("./close_cursor");
const set_handler_1 = require("./set_handler");
const split_1 = require("./split");
const translate_1 = require("./translate");
const call_transaction_1 = require("./call_transaction");
const update_database_1 = require("./update_database");
const write_1 = require("./write");
const call_function_1 = require("./call_function");
const trace_1 = require("../trace");
// this is a class, as statements like SELECT needs access to the database object instance
// and WRITE will access the Console
class Statements {
    append = append_1.append;
    assert = assert_1.assert;
    assign = assign_1.assign;
    cast = cast_1.cast;
    collect = collect_1.collect;
    commit = commit_1.commit;
    concatenate = concatenate_1.concatenate;
    condense = condense_1.condense;
    convert = convert_1.convert;
    createData = create_data_1.createData;
    deleteInternal = delete_internal_1.deleteInternal;
    describe = describe_1.describe;
    find = find_1.find;
    unpack = unpack_1.unpack;
    getBit = get_bit_1.getBit;
    readReport = read_report_1.readReport;
    getLocale = get_locale_1.getLocale;
    getParameter = get_parameter_1.getParameter;
    getRunTime = get_run_time_1.getRunTime;
    getTime = get_time_1.getTime;
    insertInternal = insert_internal_1.insertInternal;
    loop = loop_1.loop;
    modifyInternal = modify_internal_1.modifyInternal;
    moveCorresponding = move_corresponding_1.moveCorresponding;
    overlay = overlay_1.overlay;
    raiseEvent = raise_event_1.raiseEvent;
    readTable = read_table_1.readTable;
    replace = replace_1.replace;
    rollback = rollback_1.rollback;
    setBit = set_bit_1.setBit;
    setHandler = set_handler_1.setHandler;
    setLocale = set_locale_1.setLocale;
    getReference = get_reference_1.getReference;
    shift = shift_1.shift;
    sort = sort_1.sort;
    split = split_1.split;
    translate = translate_1.translate;
    wait = wait_1.wait;
    receive = receive_1.receive;
    callTransaction = call_transaction_1.callTransaction;
    context;
    constructor(context) {
        this.context = context;
    }
    _setTrace(min = 10, totals = false) {
        return new trace_1.Trace().setTrace(min, totals, this);
    }
    async openCursor(target, select, options) {
        const num = await (0, open_cursor_1.openCursor)(this.context, select, options);
        target.set(num);
    }
    async fetchNextCursor(cursor, target, packageSize) {
        await (0, fetch_next_cursor_1.fetchNextCursor)(this.context, cursor.get(), target, packageSize?.get() || 0);
    }
    async closeCursor(cursor) {
        await (0, close_cursor_1.closeCursor)(this.context, cursor.get());
    }
    async deleteDatabase(table, options) {
        await (0, delete_database_1.deleteDatabase)(table, options, this.context);
    }
    async insertDatabase(table, options) {
        return (0, insert_database_1.insertDatabase)(table, options, this.context);
    }
    async modifyDatabase(table, options) {
        return (0, modify_database_1.modifyDatabase)(table, options, this.context);
    }
    async select(target, select, runtimeOptions) {
        return (0, select_1.select)(target, select, runtimeOptions || {}, this.context);
    }
    async updateDatabase(table, options) {
        return (0, update_database_1.updateDatabase)(table, options, this.context);
    }
    async callFunction(options) {
        return new call_function_1.CallFunction(this.context).callFunction(options);
    }
    async message(options) {
        return new message_1.MessageStatement(this.context).message(options);
    }
    write(source, options) {
        return new write_1.WriteStatement(this.context).write(source, options);
    }
}
exports.Statements = Statements;
