// @abaplint/core/build/src/abap/5_syntax/_object_oriented.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectOriented = void 0;
const Statements = require("../2_statements/statements");
const Expressions = require("../2_statements/expressions");
const visibility_1 = require("../4_file_information/visibility");
const types_1 = require("../types");
// todo, think some of the public methods can be made private
// todo: changet this class to static? for performance
class ObjectOriented {
    constructor(scope) {
        this.scope = scope;
    }
    fromInterfaceByName(name, ignore) {
        const idef = this.scope.findInterfaceDefinition(name);
        if (idef === undefined || ignore.includes(name.toUpperCase())) {
            return [];
        }
        const ret = [name.toUpperCase()];
        for (const t of idef.getTypeDefinitions().getAll()) {
            const n = name + "~" + t.type.getName();
            this.scope.addTypeNamed(n, t.type);
        }
        this.scope.addListPrefix(idef.getAttributes().getConstants(), name + "~");
        this.scope.addListPrefix(idef.getAttributes().getStatic(), name + "~");
        this.scope.addListPrefix(idef.getAttributes().getInstance(), name + "~");
        for (const i of idef.getImplementing()) {
            if (ignore.includes(i.name.toUpperCase())) {
                continue;
            }
            ret.push(...this.fromInterfaceByName(i.name, ignore));
            ignore.push(i.name.toUpperCase());
            ret.push(i.name.toUpperCase());
        }
        return ret;
    }
    addAliasedAttributes(classDefinition) {
        for (const alias of classDefinition.getAliases() || []) {
            const comp = alias.getComponent();
            const idef = this.scope.findInterfaceDefinition(comp.split("~")[0]);
            if (idef) {
                const found = idef.getAttributes().findByName(comp.split("~")[1]);
                if (found) {
                    this.scope.addNamedIdentifier(alias.getName(), found);
                }
            }
        }
        const superName = classDefinition.getSuperClass();
        if (superName !== undefined) {
            const def = this.scope.findClassDefinition(superName);
            if (def) {
                this.addAliasedAttributes(def);
            }
        }
    }
    addAliasedTypes(classDefinition) {
        for (const alias of classDefinition.getAliases() || []) {
            const comp = alias.getComponent();
            const idef = this.scope.findInterfaceDefinition(comp.split("~")[0]);
            if (idef) {
                const found = idef.getTypeDefinitions().getByName(comp.split("~")[1]);
                if (found) {
                    this.scope.addType(found);
                }
            }
        }
        const superName = classDefinition.getSuperClass();
        if (superName !== undefined) {
            const def = this.scope.findClassDefinition(superName);
            if (def) {
                this.addAliasedTypes(def);
            }
        }
    }
    findMethodInInterface(interfaceName, methodName) {
        const idef = this.scope.findInterfaceDefinition(interfaceName);
        if (idef) {
            const methods = idef.getMethodDefinitions().getAll();
            for (const method of methods) {
                if (method.getName().toUpperCase() === methodName.toUpperCase()) {
                    return { method, def: idef };
                }
            }
            return this.findMethodViaAlias(methodName, idef);
        }
        return undefined;
    }
    findMethodViaAlias(methodName, def) {
        for (const a of def.getAliases() || []) {
            if (a.getName().toUpperCase() === methodName.toUpperCase()) {
                const comp = a.getComponent();
                const res = this.findMethodInInterface(comp.split("~")[0], comp.split("~")[1]);
                if (res) {
                    return res;
                }
            }
        }
        return undefined;
    }
    methodReferenceExtras(foundDef, ooName) {
        if (foundDef === undefined) {
            return {
                ooName: ooName,
                ooType: undefined,
            };
        }
        return {
            ooName: foundDef === null || foundDef === void 0 ? void 0 : foundDef.getName(),
            ooType: foundDef instanceof types_1.ClassDefinition ? "CLAS" : "INTF",
        };
    }
    findClassName(node) {
        if (!(node.get() instanceof Statements.ClassImplementation
            || node.get() instanceof Statements.ClassDefinition)) {
            throw new Error("findClassName, unexpected node type");
        }
        const className = node.findFirstExpression(Expressions.ClassName);
        if (className === undefined) {
            throw new Error("findClassName, unexpected node type");
        }
        return className.getFirstToken().getStr();
    }
    findInterfaces(cd) {
        var _a;
        const ret = [...cd.getImplementing()];
        for (const r of ret) {
            const nested = (_a = this.scope.findInterfaceDefinition(r.name)) === null || _a === void 0 ? void 0 : _a.getImplementing();
            if (nested) {
                ret.push(...nested);
            }
        }
        const sup = cd.getSuperClass();
        if (sup) {
            try {
                ret.push(...this.findInterfaces(this.findSuperDefinition(sup)));
            }
            catch (_b) {
                // ignore errors, they will show up as variable not found anyhow
            }
        }
        return ret;
    }
    searchEvent(def, name) {
        if (def === undefined || name === undefined) {
            return undefined;
        }
        const found = def.getEvents().find(e => e.getName().toUpperCase() === (name === null || name === void 0 ? void 0 : name.toUpperCase()));
        if (found) {
            return found;
        }
        for (const a of def.getAliases() || []) {
            if (a.getName().toUpperCase() === name.toUpperCase()) {
                const comp = a.getComponent();
                const res = this.searchEvent(this.scope.findObjectDefinition(comp.split("~")[0]), comp.split("~")[1]);
                if (res) {
                    return res;
                }
            }
        }
        const sup = def.getSuperClass();
        if (sup) {
            return this.searchEvent(this.findSuperDefinition(sup), name);
        }
        return undefined;
    }
    // search in via super class, interfaces and aliases
    searchAttributeName(def, name) {
        if (def === undefined || name === undefined) {
            return undefined;
        }
        const upper = name.toUpperCase();
        const attr = def.getAttributes();
        if (attr) {
            for (const a of attr.getAll()) {
                if (a.getName().toUpperCase() === upper) {
                    return a;
                }
            }
        }
        for (const a of def.getAliases() || []) {
            if (a.getName().toUpperCase() === upper) {
                const comp = a.getComponent();
                const res = this.searchAttributeName(this.scope.findObjectDefinition(comp.split("~")[0]), comp.split("~")[1]);
                if (res) {
                    return res;
                }
            }
        }
        if (name.includes("~")) {
            const interfaceName = upper.split("~")[0];
            if (this.listInterfacesRecursive(def).includes(interfaceName)) {
                return this.searchAttributeName(this.scope.findInterfaceDefinition(interfaceName), name.split("~")[1]);
            }
        }
        const sup = def.getSuperClass();
        if (sup) {
            return this.searchAttributeName(this.findSuperDefinition(sup), name);
        }
        return undefined;
    }
    // search in via super class, interfaces and aliases
    searchTypeName(def, name) {
        if (def === undefined || name === undefined) {
            return undefined;
        }
        const search = def.getTypeDefinitions().getByName(name);
        if (search) {
            return search;
        }
        for (const a of def.getAliases()) {
            if (a.getName().toUpperCase() === name.toUpperCase()) {
                const comp = a.getComponent();
                const res = this.searchTypeName(this.scope.findObjectDefinition(comp.split("~")[0]), comp.split("~")[1]);
                if (res) {
                    return res;
                }
            }
        }
        if (name.includes("~")) {
            const interfaceName = name.split("~")[0];
            if (def.getImplementing().some((a) => a.name.toUpperCase() === interfaceName.toUpperCase())) {
                return this.searchTypeName(this.scope.findInterfaceDefinition(interfaceName), name.split("~")[1]);
            }
        }
        const sup = def.getSuperClass();
        if (sup) {
            return this.searchTypeName(this.findSuperDefinition(sup), name);
        }
        return undefined;
    }
    // search in via super class, interfaces and aliases
    searchConstantName(def, name) {
        if (def === undefined || name === undefined) {
            return undefined;
        }
        const upper = name.toUpperCase();
        for (const a of def.getAttributes().getConstants()) {
            if (a.getName().toUpperCase() === upper) {
                return a;
            }
        }
        for (const a of def.getAliases()) {
            if (a.getName().toUpperCase() === upper) {
                const comp = a.getComponent();
                const res = this.searchConstantName(this.scope.findObjectDefinition(comp.split("~")[0]), comp.split("~")[1]);
                if (res) {
                    return res;
                }
            }
        }
        if (name.includes("~")) {
            const interfaceName = upper.split("~")[0];
            if (def.getImplementing().some((a) => a.name.toUpperCase() === interfaceName)) {
                return this.searchConstantName(this.scope.findInterfaceDefinition(interfaceName), name.split("~")[1]);
            }
        }
        const sup = def.getSuperClass();
        if (sup) {
            return this.searchConstantName(this.findSuperDefinition(sup), name);
        }
        return undefined;
    }
    // search in via super class, interfaces and aliases
    searchMethodName(def, name) {
        if (def === undefined || name === undefined) {
            return { method: undefined, def: undefined };
        }
        const methodDefinition = this.findMethod(def, name);
        if (methodDefinition) {
            return { method: methodDefinition, def };
        }
        let interfaceName = undefined;
        if (name.includes("~")) {
            interfaceName = name.split("~")[0];
            if (interfaceName && this.findInterfaces(def).some(i => i.name.toUpperCase() === (interfaceName === null || interfaceName === void 0 ? void 0 : interfaceName.toUpperCase())) === false) {
                return { method: undefined, def: undefined };
            }
        }
        // todo, this is not completely correct? hmm, why? visibility?
        if (methodDefinition === undefined && interfaceName) {
            name = name.split("~")[1];
            const found = this.findMethodInInterface(interfaceName, name);
            if (found) {
                return found;
            }
        }
        else if (methodDefinition === undefined) {
            const found = this.findMethodViaAlias(name, def);
            if (found) {
                return found;
            }
        }
        const sup = def.getSuperClass();
        if (methodDefinition === undefined && sup) {
            return this.searchMethodName(this.findSuperDefinition(sup), name);
        }
        return { method: undefined, def: undefined };
    }
    findMethod(def, methodName) {
        const defs = def.getMethodDefinitions();
        if (defs === undefined) {
            return undefined;
        }
        for (const method of defs.getAll()) {
            if (method.getName().toUpperCase() === methodName.toUpperCase()) {
                if (method.isRedefinition()) {
                    return this.findMethodInSuper(def, methodName);
                }
                else {
                    return method;
                }
            }
        }
        return undefined;
    }
    findMethodInSuper(child, methodName) {
        let sup = child.getSuperClass();
        while (sup !== undefined) {
            const cdef = this.findSuperDefinition(sup);
            const found = this.findMethod(cdef, methodName);
            if (found) {
                return found;
            }
            sup = cdef.getSuperClass();
        }
        return undefined;
    }
    findSuperDefinition(name) {
        const csup = this.scope.findClassDefinition(name);
        if (csup === undefined) {
            throw new Error("Super class \"" + name + "\" not found or contains errors");
        }
        return csup;
    }
    fromSuperClassesAndInterfaces(child) {
        const implemented = this.fromSuperClasses(child);
        this.fromInterfaces(child, implemented);
    }
    // returns list of interfaces implemented
    fromSuperClasses(child) {
        let sup = child.getSuperClass();
        const ignore = [];
        while (sup !== undefined) {
            const cdef = this.findSuperDefinition(sup);
            for (const a of cdef.getAttributes().getAll()) {
                if (a.getVisibility() !== visibility_1.Visibility.Private) {
                    this.scope.addIdentifier(a);
                    // todo, handle scope and instance vs static
                }
            }
            for (const c of cdef.getAttributes().getConstants()) {
                if (c.getVisibility() !== visibility_1.Visibility.Private) {
                    this.scope.addIdentifier(c);
                }
            }
            for (const t of cdef.getTypeDefinitions().getAll()) {
                if (t.visibility !== visibility_1.Visibility.Private) {
                    this.scope.addType(t.type);
                }
            }
            ignore.push(...this.fromInterfaces(cdef, ignore));
            sup = cdef.getSuperClass();
        }
        return ignore;
    }
    /** returns list of interfaces implemented, recursive */
    listInterfacesRecursive(definition) {
        const list = [];
        for (const i of definition.getImplementing()) {
            const upper = i.name.toUpperCase();
            list.push(upper);
            const def = this.scope.findInterfaceDefinition(upper);
            if (def) {
                list.push(...this.listInterfacesRecursive(def));
            }
        }
        return [...new Set(list)];
    }
    fromInterfaces(definition, skip) {
        const ignore = [];
        for (const i of definition.getImplementing()) {
            ignore.push(...this.fromInterfaceByName(i.name, ignore.concat(skip || [])));
        }
        return ignore;
    }
}
exports.ObjectOriented = ObjectOriented;
