// @abaplint/core/build/src/abap/5_syntax/_reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReferenceType = void 0;
var ReferenceType;
(function (ReferenceType) {
    /** for classes and interface references */
    ReferenceType["ObjectOrientedReference"] = "Object";
    ReferenceType["ObjectOrientedVoidReference"] = "Object (Void)";
    ReferenceType["ObjectOrientedUnknownReference"] = "Object (Unknown)";
    ReferenceType["TableReference"] = "Table";
    ReferenceType["TableVoidReference"] = "Table (Void)";
    ReferenceType["MethodReference"] = "Method";
    ReferenceType["BuiltinMethodReference"] = "Builtin Method";
    ReferenceType["ConstructorReference"] = "Constructor Reference";
    ReferenceType["MethodImplementationReference"] = "Method Implementation";
    ReferenceType["TypeReference"] = "Type";
    ReferenceType["BuiltinTypeReference"] = "Builtin Type";
    ReferenceType["VoidType"] = "Type (Void)";
    ReferenceType["InferredType"] = "Inferred Type";
    ReferenceType["FormReference"] = "Form";
    //  FormVoidReference = "Form (void)",
    ReferenceType["DataReadReference"] = "Read From";
    ReferenceType["DataWriteReference"] = "Write To";
})(ReferenceType || (exports.ReferenceType = ReferenceType = {}));
