// @abaplint/core/build/src/abap/5_syntax/_type_utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeUtils = void 0;
const types_1 = require("../types");
const basic_1 = require("../types/basic");
const cgeneric_type_1 = require("../types/basic/cgeneric_type");
const Expressions = require("../2_statements/expressions");
// todo: refactor to static? for performance
class TypeUtils {
    constructor(scope) {
        this.scope = scope;
    }
    isCharLikeStrict(type) {
        if (type === undefined) {
            return false;
        }
        else if (type instanceof basic_1.StructureType) {
            for (const c of type.getComponents()) {
                if (this.isCharLikeStrict(c.type) === false) {
                    return false;
                }
            }
            return true;
        }
        else if (type instanceof basic_1.TableType && type.isWithHeader()) {
            return this.isCharLikeStrict(type.getRowType());
        }
        else if (type instanceof basic_1.StringType
            || type instanceof basic_1.AnyType
            || type instanceof basic_1.DataType
            || type instanceof basic_1.CharacterType
            || type instanceof basic_1.SimpleType
            || type instanceof cgeneric_type_1.CGenericType
            || type instanceof basic_1.CLikeType
            || type instanceof basic_1.DateType
            || type instanceof basic_1.CSequenceType
            || type instanceof basic_1.NumericGenericType
            || type instanceof basic_1.NumericType
            || type instanceof basic_1.TimeType
            || type instanceof basic_1.UnknownType
            || type instanceof basic_1.VoidType) {
            return true;
        }
        return false;
    }
    isCharLike(type) {
        if (type === undefined) {
            return false;
        }
        else if (type instanceof basic_1.TableType && type.isWithHeader()) {
            return this.isCharLike(type.getRowType());
        }
        else if (type instanceof basic_1.StructureType) {
            for (const c of type.getComponents()) {
                if (this.isCharLike(c.type) === false) {
                    return false;
                }
            }
            return true;
        }
        else if (type instanceof basic_1.StringType
            || type instanceof basic_1.CharacterType
            || type instanceof basic_1.VoidType
            || type instanceof basic_1.AnyType
            || type instanceof basic_1.UnknownType
            || type instanceof basic_1.NumericType
            || type instanceof basic_1.IntegerType
            || type instanceof basic_1.Integer8Type
            || type instanceof basic_1.SimpleType
            || type instanceof basic_1.FloatType
            || type instanceof basic_1.FloatingPointType
            || type instanceof basic_1.DecFloatType
            || type instanceof basic_1.DecFloat16Type
            || type instanceof basic_1.DecFloat34Type
            || type instanceof basic_1.NumericGenericType
            || type instanceof basic_1.CSequenceType
            || type instanceof cgeneric_type_1.CGenericType
            || type instanceof basic_1.DateType
            || type instanceof basic_1.DataType
            || type instanceof basic_1.CLikeType
            || type instanceof basic_1.PackedType
            || type instanceof basic_1.TimeType) {
            return true;
        }
        return false;
    }
    isHexLike(type) {
        if (type === undefined) {
            return false;
        }
        else if (type instanceof basic_1.StructureType) {
            for (const c of type.getComponents()) {
                if (this.isHexLike(c.type) === false) {
                    return false;
                }
            }
            return true;
        }
        else if (type instanceof basic_1.TableType && type.isWithHeader()) {
            return this.isHexLike(type.getRowType());
        }
        else if (type instanceof basic_1.XStringType
            || type instanceof basic_1.HexType
            || type instanceof basic_1.VoidType
            || type instanceof basic_1.XGenericType
            || type instanceof basic_1.XSequenceType
            || type instanceof basic_1.AnyType
            || type instanceof basic_1.UnknownType) {
            return true;
        }
        return false;
    }
    isOOAssignable(source, target) {
        let sid = source.getIdentifier();
        let tid = target.getIdentifier();
        const tname = tid.getName().toUpperCase();
        const sname = sid.getName().toUpperCase();
        if (tname === sname) {
            return true;
        }
        /*
            console.dir(sid);
            console.dir(tid);
        */
        if (!(sid instanceof types_1.ClassDefinition || sid instanceof types_1.InterfaceDefinition)) {
            const found = this.scope.findObjectDefinition(sid.getName());
            if (found) {
                sid = found;
            }
            else {
                return false;
            }
        }
        if (!(tid instanceof types_1.ClassDefinition || tid instanceof types_1.InterfaceDefinition)) {
            const found = this.scope.findObjectDefinition(tid.getName());
            if (found) {
                tid = found;
            }
            else {
                return false;
            }
        }
        if (sid instanceof types_1.ClassDefinition && tid instanceof types_1.ClassDefinition) {
            if (sname === tname) {
                return true;
            }
            const slist = this.listAllSupers(sid);
            if (slist.indexOf(tname) >= 0) {
                return true;
            }
        }
        else if (sid instanceof types_1.ClassDefinition && tid instanceof types_1.InterfaceDefinition) {
            if (sid.getImplementing().some(i => i.name === tname)) {
                return true;
            }
            const slist = this.listAllInterfaces(sid);
            if (slist.indexOf(tname) >= 0) {
                return true;
            }
        }
        else if (sid instanceof types_1.InterfaceDefinition && tid instanceof types_1.InterfaceDefinition) {
            if (sname === tname) {
                return true;
            }
            if (sid.getImplementing().some(i => i.name === tname)) {
                return true;
            }
            const slist = this.listAllInterfaces(sid);
            if (slist.indexOf(tname) >= 0) {
                return true;
            }
        }
        return false;
    }
    listAllInterfaces(cdef) {
        var _a;
        const ret = new Set();
        const stack = [];
        // initialize
        cdef.getImplementing().forEach(i => stack.push(i.name));
        if (cdef instanceof types_1.ClassDefinition) {
            const supers = this.listAllSupers(cdef);
            for (const s of supers) {
                (_a = this.scope.findClassDefinition(s)) === null || _a === void 0 ? void 0 : _a.getImplementing().forEach(i => stack.push(i.name));
            }
        }
        // main loop
        while (stack.length > 0) {
            const intf = stack.pop().toUpperCase();
            ret.add(intf);
            const idef = this.scope.findInterfaceDefinition(intf);
            idef === null || idef === void 0 ? void 0 : idef.getImplementing().forEach(i => stack.push(i.name));
        }
        return Array.from(ret.values());
    }
    listAllSupers(cdef) {
        var _a, _b;
        const ret = [];
        let sup = cdef.getSuperClass();
        while (sup !== undefined) {
            ret.push(sup === null || sup === void 0 ? void 0 : sup.toUpperCase());
            sup = (_b = (_a = this.scope.findClassDefinition(sup)) === null || _a === void 0 ? void 0 : _a.getSuperClass()) === null || _b === void 0 ? void 0 : _b.toUpperCase();
        }
        return ret;
    }
    isConvable(source, target) {
        if (source instanceof basic_1.TableType && target instanceof basic_1.TableType) {
            // table key conversion, eg DEFAULT to EMPTY
            return this.isAssignable(source.getRowType(), target.getRowType());
        }
        return this.isAssignable(source, target);
    }
    isCastable(_source, _target) {
        // todo
        return true;
    }
    isCompareable(source1, source2, node1, node2) {
        /*
        console.dir(source1);
        console.dir(source2);
        */
        if (source1 === undefined || source2 === undefined) {
            return true;
        }
        if (source1 instanceof basic_1.HexType && this.isCalculated(node1) && source2 instanceof basic_1.IntegerType) {
            return false;
        }
        if (source2 instanceof basic_1.HexType && this.isCalculated(node2) && source1 instanceof basic_1.IntegerType) {
            return false;
        }
        return true;
    }
    structureContainsString(structure) {
        for (const c of structure.getComponents()) {
            if (c.type instanceof basic_1.StringType) {
                return true;
            }
        }
        return false;
    }
    structureContainsVoid(structure) {
        for (const c of structure.getComponents()) {
            if (c.type instanceof basic_1.VoidType) {
                return true;
            }
        }
        return false;
    }
    isCalculated(node) {
        /*
        if (node.getChildren().length === 1
            && node.get() instanceof Expressions.Source
            && node.getFirstChild()?.get() instanceof Expressions.MethodCallChain) {
          return false;
        }
        */
        const calculated = node.findFirstExpression(Expressions.MethodCallChain) !== undefined
            || node.findFirstExpression(Expressions.StringTemplate) !== undefined
            || node.findFirstExpression(Expressions.ArithOperator) !== undefined;
        return calculated;
    }
    isAssignableNew(source, target, node) {
        const calculated = (node === null || node === void 0 ? void 0 : node.findFirstExpression(Expressions.ArithOperator)) !== undefined;
        if (calculated && source instanceof basic_1.HexType && target instanceof basic_1.IntegerType) {
            return false;
        }
        return this.isAssignable(source, target);
    }
    isAssignableStrict(source, target, node) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const calculated = node ? this.isCalculated(node) : false;
        /*
            console.dir(source);
            console.dir(target);
            console.dir(calculated);
        */
        if (calculated && source instanceof basic_1.StringType && target instanceof basic_1.XStringType) {
            return false;
        }
        if (calculated) {
            return this.isAssignable(source, target);
        }
        if (source instanceof basic_1.CharacterType) {
            if (target instanceof basic_1.CharacterType) {
                if (((_a = source.getAbstractTypeData()) === null || _a === void 0 ? void 0 : _a.derivedFromConstant) === true) {
                    return source.getLength() <= target.getLength();
                }
                return source.getLength() === target.getLength();
            }
            else if (target instanceof basic_1.IntegerType) {
                if (((_b = source.getAbstractTypeData()) === null || _b === void 0 ? void 0 : _b.derivedFromConstant) === true) {
                    return true;
                }
                return false;
            }
            else if (target instanceof basic_1.XStringType) {
                if (((_c = source.getAbstractTypeData()) === null || _c === void 0 ? void 0 : _c.derivedFromConstant) === true) {
                    return true;
                }
                return false;
            }
            else if (target instanceof basic_1.StringType) {
                if (((_d = source.getAbstractTypeData()) === null || _d === void 0 ? void 0 : _d.derivedFromConstant) === true) {
                    return true;
                }
                return false;
            }
        }
        else if (source instanceof basic_1.HexType) {
            if (target instanceof basic_1.HexType) {
                if (((_e = source.getAbstractTypeData()) === null || _e === void 0 ? void 0 : _e.derivedFromConstant) === true) {
                    return source.getLength() <= target.getLength();
                }
                return source.getLength() === target.getLength();
            }
            else if (target instanceof basic_1.XStringType) {
                return false;
            }
            else if (target instanceof basic_1.IntegerType || target instanceof basic_1.Integer8Type) {
                if (((_f = source.getAbstractTypeData()) === null || _f === void 0 ? void 0 : _f.derivedFromConstant) === true) {
                    return true;
                }
                return false;
            }
        }
        else if (source instanceof basic_1.StringType) {
            if (target instanceof basic_1.StructureType && this.structureContainsString(target)) {
                return false;
                /*
              } else if (target instanceof CharacterType) {
                if (source.getAbstractTypeData()?.derivedFromConstant === true) {
                  return true;
                }
                return false;
                */
            }
            else if (target instanceof basic_1.IntegerType) {
                if (((_g = source.getAbstractTypeData()) === null || _g === void 0 ? void 0 : _g.derivedFromConstant) === true) {
                    return true;
                }
                return false;
            }
            else if (target instanceof cgeneric_type_1.CGenericType
                || target instanceof basic_1.GenericObjectReferenceType) {
                return false;
            }
            else if (target instanceof basic_1.XSequenceType || target instanceof basic_1.XStringType) {
                if (((_h = source.getAbstractTypeData()) === null || _h === void 0 ? void 0 : _h.derivedFromConstant) === true) {
                    return true;
                }
                return false;
            }
            return true;
        }
        else if (source instanceof basic_1.StructureType) {
            if (target instanceof basic_1.StructureType) {
                const sourceComponents = source.getComponents();
                const targetComponents = target.getComponents();
                if (sourceComponents.length !== targetComponents.length) {
                    return false;
                }
                for (let i = 0; i < sourceComponents.length; i++) {
                    if (this.isAssignableStrict(sourceComponents[i].type, targetComponents[i].type) === false) {
                        return false;
                    }
                }
                return true;
            }
            else if (target instanceof basic_1.SimpleType) {
                return this.isCharLikeStrict(source);
            }
            else if (target instanceof basic_1.CLikeType) {
                return this.isCharLikeStrict(source);
            }
            else if (target instanceof basic_1.VoidType
                || target instanceof basic_1.AnyType
                || target instanceof basic_1.DataType) {
                return true;
            }
            return false;
        }
        else if (source instanceof basic_1.Integer8Type) {
            if (target instanceof basic_1.IntegerType || target instanceof basic_1.StringType) {
                return false;
            }
        }
        else if (source instanceof basic_1.IntegerType) {
            if (target instanceof basic_1.StringType) {
                return false;
            }
            else if (target instanceof basic_1.Integer8Type || target instanceof basic_1.PackedType) {
                if (((_j = source.getAbstractTypeData()) === null || _j === void 0 ? void 0 : _j.derivedFromConstant) === true) {
                    return true;
                }
                return false;
            }
        }
        else if (source instanceof basic_1.FloatType) {
            if (target instanceof basic_1.IntegerType) {
                return false;
            }
        }
        else if (source instanceof basic_1.GenericObjectReferenceType) {
            if (target instanceof basic_1.ObjectReferenceType
                || target instanceof basic_1.StringType
                || target instanceof basic_1.CharacterType
                || target instanceof basic_1.TableType
                || target instanceof basic_1.CLikeType) {
                return false;
            }
        }
        else if (source instanceof basic_1.ObjectReferenceType) {
            if (target instanceof basic_1.XSequenceType
                || target instanceof basic_1.IntegerType
                || target instanceof basic_1.StructureType
                || target instanceof basic_1.TableType
                || target instanceof basic_1.XStringType) {
                return false;
            }
        }
        else if (source instanceof basic_1.XStringType) {
            if (target instanceof basic_1.CLikeType
                || target instanceof basic_1.IntegerType
                || target instanceof basic_1.StringType
                || target instanceof basic_1.ObjectReferenceType
                || target instanceof basic_1.HexType) {
                return false;
            }
        }
        else if (source instanceof basic_1.TableType) {
            if (target instanceof basic_1.TableType) {
                return this.isAssignableStrict(source.getRowType(), target.getRowType());
            }
            else if (target instanceof basic_1.UnknownType
                || target instanceof basic_1.AnyType
                || target instanceof basic_1.VoidType) {
                return true;
            }
        }
        return this.isAssignable(source, target);
    }
    isAssignable(source, target) {
        if (source === undefined || target === undefined) {
            return true;
        }
        /*
            console.dir(source);
            console.dir(target);
        */
        if (target instanceof basic_1.TableType) {
            if (target.isWithHeader()) {
                return this.isAssignable(source, target.getRowType());
            }
            if (source instanceof basic_1.VoidType
                || source instanceof basic_1.AnyType
                || source instanceof basic_1.DataType
                || source instanceof basic_1.UnknownType) {
                return true;
            }
            else if (source instanceof basic_1.TableType) {
                const targetRowType = target.getRowType();
                const sourceRowType = source.getRowType();
                if (targetRowType instanceof basic_1.VoidType || targetRowType instanceof basic_1.AnyType || targetRowType instanceof basic_1.UnknownType) {
                    return true;
                }
                else if (sourceRowType instanceof basic_1.VoidType || sourceRowType instanceof basic_1.AnyType || sourceRowType instanceof basic_1.UnknownType) {
                    return true;
                }
                if (targetRowType instanceof basic_1.StructureType
                    && this.structureContainsString(targetRowType)) {
                    if (!(sourceRowType instanceof basic_1.StructureType)) {
                        return false;
                    }
                    else if (!this.structureContainsString(sourceRowType)
                        && this.structureContainsVoid(sourceRowType) === false) {
                        return false;
                    }
                }
                else if (sourceRowType instanceof basic_1.StructureType
                    && this.structureContainsString(sourceRowType)) {
                    if (!(targetRowType instanceof basic_1.StructureType)) {
                        return false;
                    }
                    else if (!this.structureContainsString(targetRowType)
                        && this.structureContainsVoid(targetRowType) === false) {
                        return false;
                    }
                }
                return true;
            }
            return false;
        }
        else if (target instanceof basic_1.ObjectReferenceType && source instanceof basic_1.ObjectReferenceType) {
            return this.isOOAssignable(source, target);
        }
        else if (target instanceof basic_1.ObjectReferenceType
            || target instanceof basic_1.GenericObjectReferenceType) {
            if (source instanceof basic_1.ObjectReferenceType
                || source instanceof basic_1.GenericObjectReferenceType
                || source instanceof basic_1.VoidType
                || source instanceof basic_1.DataType
                || source instanceof basic_1.AnyType
                || source instanceof basic_1.UnknownType) {
                return true;
            }
            return false;
        }
        else if (target instanceof basic_1.DataReference) {
            if (source instanceof basic_1.DataReference
                || source instanceof basic_1.VoidType
                || source instanceof basic_1.DataType
                || source instanceof basic_1.AnyType
                || source instanceof basic_1.UnknownType) {
                return true;
            }
            return false;
        }
        else if (target instanceof basic_1.StructureType) {
            if (source instanceof basic_1.TableType && source.isWithHeader()) {
                return this.isAssignable(source.getRowType(), target);
            }
            else if (source instanceof basic_1.VoidType
                || source instanceof basic_1.AnyType
                || source instanceof basic_1.DataType
                || source instanceof basic_1.UnknownType) {
                return true;
            }
            else if (source instanceof basic_1.StructureType) {
                const targetDeep = this.structureContainsString(target);
                const sourceDeep = this.structureContainsString(source);
                if (targetDeep && !sourceDeep) {
                    return false;
                }
                const targetComponents = target.getComponents();
                const sourceComponents = source.getComponents();
                if (targetComponents.length !== sourceComponents.length) {
                    if (targetDeep === true || sourceDeep === true) {
                        return false;
                    }
                }
                for (let i = 0; i < targetComponents.length; i++) {
                    if (sourceComponents[i] === undefined) {
                        continue;
                    }
                    // hmm
                    if (sourceComponents[i].type instanceof basic_1.StringType && !(targetComponents[i].type instanceof basic_1.StringType)) {
                        return false;
                    }
                    else if (!(sourceComponents[i].type instanceof basic_1.StringType) && targetComponents[i].type instanceof basic_1.StringType) {
                        return false;
                    }
                }
                return true;
            }
            else if (target.containsVoid() === true) {
                return true;
            }
            else if (source instanceof basic_1.StringType && this.structureContainsString(target)) {
                return false;
            }
            else if (source instanceof basic_1.IntegerType) {
                return false;
            }
            else if (this.isCharLike(target) && this.isCharLike(source)) {
                return true;
            }
            return false;
        }
        else if (target instanceof basic_1.IntegerType
            || target instanceof basic_1.CharacterType
            || target instanceof basic_1.StringType) {
            if (source instanceof basic_1.TableType && source.isWithHeader() === false) {
                return false;
            }
            else if ((target instanceof basic_1.StringType || target instanceof basic_1.CharacterType)
                && source instanceof basic_1.StructureType
                && this.isCharLike(source)) {
                if (this.structureContainsString(source) === true) {
                    return false;
                }
                return true;
            }
            else if (source instanceof basic_1.StructureType) {
                return this.isCharLikeStrict(source);
            }
            else if (source instanceof basic_1.DataReference
                || source instanceof basic_1.ObjectReferenceType
                || source instanceof basic_1.GenericObjectReferenceType) {
                return false;
            }
            return true;
        }
        else if (target instanceof basic_1.SimpleType) {
            if (source instanceof basic_1.ObjectReferenceType
                || source instanceof basic_1.TableType) {
                return false;
            }
        }
        return true;
    }
}
exports.TypeUtils = TypeUtils;
