// @abaplint/core/build/src/abap/5_syntax/basic_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasicTypes = void 0;
const _typed_identifier_1 = require("../types/_typed_identifier");
const nodes_1 = require("../nodes");
const Expressions = require("../2_statements/expressions");
const Statements = require("../2_statements/statements");
const Types = require("../types/basic");
const _scope_type_1 = require("./_scope_type");
const _object_oriented_1 = require("./_object_oriented");
const class_constant_1 = require("../types/class_constant");
const identifier_1 = require("../1_lexer/tokens/identifier");
const _reference_1 = require("./_reference");
const basic_1 = require("../types/basic");
const field_chain_1 = require("./expressions/field_chain");
const types_1 = require("../types");
const expressions_1 = require("../2_statements/expressions");
const _builtin_1 = require("./_builtin");
const position_1 = require("../../position");
class BasicTypes {
    constructor(input) {
        this.input = input;
    }
    lookupQualifiedName(name) {
        var _a;
        // argh, todo, rewrite this entire method, more argh, again argh
        if (name === undefined) {
            return undefined;
        }
        const found = this.input.scope.findType(name);
        if (found) {
            return found;
        }
        if (name.includes("=>")) {
            const split = name.split("=>");
            const ooName = split[0];
            const typeName = split[1];
            const oo = this.input.scope.findObjectDefinition(ooName);
            if (oo) {
                if (typeName.includes("-")) {
                    const split = typeName.split("-");
                    const subTypeName = split[0];
                    const fieldName = split[1];
                    const stru = oo.getTypeDefinitions().getByName(subTypeName);
                    const struType = stru === null || stru === void 0 ? void 0 : stru.getType();
                    if (stru && struType instanceof basic_1.StructureType) {
                        let f = struType.getComponentByName(fieldName);
                        if (split[2] && f instanceof basic_1.StructureType) {
                            f = f.getComponentByName(split[2]);
                        }
                        if (f) {
                            return new _typed_identifier_1.TypedIdentifier(stru.getToken(), stru.getFilename(), f);
                        }
                    }
                }
                else {
                    const f = oo.getTypeDefinitions().getByName(typeName);
                    if (f) {
                        return f;
                    }
                }
            }
        }
        else if (name.includes("-")) {
            const split = name.split("-");
            const typeName = split[0];
            const fieldName = split[1];
            const type = this.input.scope.findType(typeName);
            if (type) {
                const stru = type.getType();
                if (stru instanceof basic_1.StructureType) {
                    let f = stru.getComponentByName(fieldName);
                    if (split[2] && f instanceof basic_1.StructureType) {
                        f = f.getComponentByName(split[2]);
                    }
                    if (f) {
                        return new _typed_identifier_1.TypedIdentifier(type.getToken(), type.getFilename(), f);
                    }
                }
            }
        }
        const lookup = this.input.scope.getDDIC().lookupNoVoid(name);
        const id = (_a = lookup === null || lookup === void 0 ? void 0 : lookup.object) === null || _a === void 0 ? void 0 : _a.getIdentifier();
        if (id && (lookup === null || lookup === void 0 ? void 0 : lookup.type)) {
            return new _typed_identifier_1.TypedIdentifier(id.getToken(), id.getFilename(), lookup.type);
        }
        const builtin = this.input.scope.getDDIC().lookupBuiltinType(name);
        if (builtin) {
            return new _typed_identifier_1.TypedIdentifier(new identifier_1.Identifier(new position_1.Position(1, 1), name), _builtin_1.BuiltIn.filename, builtin);
        }
        const type = this.input.scope.findTypePoolType(name);
        if (type) {
            return type;
        }
        return undefined;
    }
    resolveLikeName(node, headerLogic = true) {
        if (node === undefined) {
            return undefined;
        }
        let chain = node.findFirstExpression(Expressions.FieldChain);
        if (chain === undefined) {
            chain = node.findFirstExpression(Expressions.TypeName);
        }
        if (chain === undefined) {
            chain = node.findFirstExpression(Expressions.FieldSub);
        }
        if (chain === undefined) {
            chain = node.findFirstExpression(Expressions.SimpleFieldChain);
        }
        if (chain === undefined) {
            chain = node.findFirstExpression(Expressions.SimpleFieldChain2);
        }
        if (chain === undefined) {
            throw new Error("resolveLikeName, chain undefined");
        }
        const fullName = chain.concatTokens();
        let children = [...chain.getChildren()];
        if (children.length === 0) {
            return new Types.UnknownType("Type error, could not resolve \"" + fullName + "\", resolveLikeName1");
        }
        let type = undefined;
        if (children[1] && (children[1].getFirstToken().getStr() === "=>" || children[1].getFirstToken().getStr() === "->")) {
            type = field_chain_1.FieldChain.runSyntax(chain, this.input, _reference_1.ReferenceType.TypeReference);
        }
        else {
            const name = children.shift().getFirstToken().getStr();
            let found = this.input.scope.findVariable(name);
            const full = this.input.scope.findVariable(fullName); // workaround for eg "sy-repid"
            if (full) {
                children = [];
                found = full;
            }
            type = found === null || found === void 0 ? void 0 : found.getType();
            if (found === undefined) {
                found = this.input.scope.findExtraLikeType(name);
                type = found === null || found === void 0 ? void 0 : found.getType();
            }
            if (found) {
                this.input.scope.addReference(chain === null || chain === void 0 ? void 0 : chain.getFirstToken(), found, _reference_1.ReferenceType.TypeReference, this.input.filename);
            }
            if (type === undefined) {
                const found = this.input.scope.getDDIC().lookupNoVoid(name);
                if (found !== undefined) {
                    const using = { filename: this.input.filename, token: chain.getFirstToken(), object: found.object };
                    this.input.scope.getDDICReferences().addUsing(this.input.scope.getParentObj(), using);
                }
                type = found === null || found === void 0 ? void 0 : found.type;
            }
            if (type === undefined && this.input.scope.isAnyOO() === false && this.input.scope.getDDIC().inErrorNamespace(name) === false) {
                this.input.scope.addReference(chain.getChildren()[0].getFirstToken(), undefined, _reference_1.ReferenceType.VoidType, this.input.filename);
                return Types.VoidType.get(name);
            }
            while (children.length > 0) {
                const child = children.shift();
                if (child.getFirstToken().getStr() === "-") {
                    if (type instanceof Types.VoidType) {
                        return type;
                    }
                }
                else if (child.concatTokens() === "[]") {
                    if (type instanceof Types.TableType) {
                        type = new basic_1.TableType(type.getRowType(), { withHeader: false, keyType: Types.TableKeyType.default });
                    }
                }
                else if (child instanceof nodes_1.ExpressionNode && child.get() instanceof Expressions.FieldLength) {
                    const len = parseInt(child.concatTokens().replace("(", "").replace(")", ""), 10);
                    if (isNaN(len)) {
                        type = new Types.UnknownType("Type error, invalid field length");
                    }
                    else {
                        if (type instanceof Types.DateType) {
                            if (len > 8) {
                                type = new Types.UnknownType("Type error, offset too long");
                            }
                            else {
                                type = new Types.CharacterType(len);
                            }
                        }
                        else if (type instanceof Types.CharacterType) {
                            if (len <= type.getLength()) {
                                type = new Types.CharacterType(len);
                            }
                            else {
                                type = new Types.UnknownType("Type error, offset too long");
                            }
                        }
                    }
                }
                else { // field name
                    if (type instanceof Types.TableType) {
                        type = type.getRowType();
                    }
                    if (type instanceof Types.StructureType) {
                        type = type.getComponentByName(child.getFirstToken().getStr());
                        if (type === undefined) {
                            return new Types.UnknownType("Type error, field not part of structure " + fullName);
                        }
                    }
                    else if (!(type instanceof Types.VoidType)) {
                        return new Types.UnknownType("Type error, field not part of structure " + fullName);
                    }
                }
            }
            if (type instanceof Types.VoidType) {
                return type;
            }
            else if (type instanceof basic_1.TableType
                && type.isWithHeader()
                && headerLogic === true) {
                type = type.getRowType();
            }
            else if (type instanceof Types.TableType
                && type.isWithHeader() === true
                && type.getRowType() instanceof Types.VoidType) {
                return type.getRowType();
            }
        }
        if (!type) {
            return new Types.UnknownType("Type error, could not resolve \"" + fullName + "\", resolveLikeName2");
        }
        return type;
    }
    cloneType(type, qualifiedName) {
        // nested types(containing "-") will inherit the qualified names if possible
        // todo, this needs to be extended to all AbstractTypes instead of just CharacterType
        if (type instanceof basic_1.CharacterType
            && qualifiedName
            && qualifiedName.includes("-") === false) {
            type = type.cloneType({ qualifiedName });
        }
        return type;
    }
    resolveTypeName(typeName, length, decimals, qualifiedName) {
        var _a;
        if (typeName === undefined) {
            return undefined;
        }
        const chain = this.resolveTypeChain(typeName);
        if (chain) {
            return this.cloneType(chain, qualifiedName);
        }
        const chainText = typeName.concatTokens().toUpperCase();
        const f = this.input.scope.getDDIC().lookupBuiltinType(chainText, length, decimals, qualifiedName);
        if (f !== undefined) {
            return f;
        }
        const typ = this.input.scope.findType(chainText);
        if (typ) {
            const token = typeName.getFirstToken();
            if (chainText.includes("~")) {
                const name = chainText.split("~")[0];
                const idef = this.input.scope.findInterfaceDefinition(name);
                if (idef) {
                    this.input.scope.addReference(token, idef, _reference_1.ReferenceType.ObjectOrientedReference, this.input.filename, { ooType: "INTF", ooName: name });
                }
            }
            this.input.scope.addReference(token, typ, _reference_1.ReferenceType.TypeReference, this.input.filename);
            return typ.getType();
        }
        const type = (_a = this.input.scope.findTypePoolType(chainText)) === null || _a === void 0 ? void 0 : _a.getType();
        if (type) {
            //      this.scope.addReference(typeName.getFirstToken(), type, ReferenceType.TypeReference, this.filename);
            return type;
        }
        const ddic = this.input.scope.getDDIC().lookup(chainText);
        if (ddic) {
            this.input.scope.getDDICReferences().addUsing(this.input.scope.getParentObj(), { object: ddic.object, token: typeName.getFirstToken(), filename: this.input.filename });
            if (ddic.type instanceof _typed_identifier_1.TypedIdentifier) {
                this.input.scope.addReference(typeName.getFirstToken(), ddic.type, _reference_1.ReferenceType.TypeReference, this.input.filename);
            }
            else if (ddic.type instanceof basic_1.VoidType) {
                this.input.scope.addReference(typeName.getFirstToken(), undefined, _reference_1.ReferenceType.VoidType, this.input.filename);
            }
            return this.cloneType(ddic.type, qualifiedName);
        }
        return undefined;
    }
    simpleType(node, qualifiedNamePrefix) {
        let nameExpr = node.findFirstExpression(Expressions.NamespaceSimpleName);
        if (nameExpr === undefined) {
            nameExpr = node.findFirstExpression(Expressions.DefinitionName);
        }
        if (nameExpr === undefined) {
            return undefined;
        }
        let name = nameExpr.getFirstToken();
        if (nameExpr.countTokens() > 1) { // workaround for names with dashes
            name = new identifier_1.Identifier(name.getStart(), nameExpr.concatTokens());
        }
        let qualifiedName = undefined;
        if (node.get() instanceof Statements.Type) {
            if (this.input.scope.isTypePool() === true) {
                qualifiedName = name.getStr();
            }
            else {
                qualifiedName = (qualifiedNamePrefix || "") + name.getStr();
                if (this.input.scope.getType() === _scope_type_1.ScopeType.ClassDefinition
                    || this.input.scope.getType() === _scope_type_1.ScopeType.Interface) {
                    qualifiedName = this.input.scope.getName() + "=>" + qualifiedName;
                }
            }
        }
        else if (qualifiedNamePrefix) {
            qualifiedName = qualifiedNamePrefix + qualifiedName;
        }
        const found = this.parseType(node, qualifiedName);
        if (found) {
            return new _typed_identifier_1.TypedIdentifier(name, this.input.filename, found);
        }
        return undefined;
    }
    parseTable(node, name) {
        var _a, _b, _c, _d;
        const typename = node.findFirstExpression(Expressions.TypeName) || node.findFirstExpression(Expressions.EntityAssociation);
        const text = (_a = node.findFirstExpression(Expressions.TypeTable)) === null || _a === void 0 ? void 0 : _a.concatTokens().toUpperCase();
        if (text === undefined) {
            return undefined;
        }
        let type = undefined;
        if (text.startsWith("TYPE STANDARD TABLE ")
            || text.startsWith("TYPE TABLE ")
            || text.startsWith("LIKE TABLE ")
            || text.startsWith("LIKE STANDARD TABLE ")) {
            type = basic_1.TableAccessType.standard;
        }
        else if (text.startsWith("TYPE SORTED TABLE ")
            || text.startsWith("LIKE SORTED TABLE ")) {
            type = basic_1.TableAccessType.sorted;
        }
        else if (text.startsWith("TYPE HASHED TABLE ")
            || text.startsWith("LIKE HASHED TABLE ")) {
            type = basic_1.TableAccessType.hashed;
        }
        const typeTableKeys = node.findAllExpressions(expressions_1.TypeTableKey);
        const firstKey = typeTableKeys[0];
        const isNamed = (firstKey === null || firstKey === void 0 ? void 0 : firstKey.findDirectExpression(expressions_1.Field)) !== undefined
            && ((_b = firstKey === null || firstKey === void 0 ? void 0 : firstKey.findDirectExpression(expressions_1.Field)) === null || _b === void 0 ? void 0 : _b.concatTokens().toUpperCase()) !== "PRIMARY_KEY";
        const primaryKey = {
            name: "primary_key",
            type: type || basic_1.TableAccessType.standard,
            isUnique: isNamed ? false : (firstKey === null || firstKey === void 0 ? void 0 : firstKey.concatTokens().toUpperCase().includes("WITH UNIQUE ")) === true,
            keyFields: [],
        };
        let start = 1;
        if (isNamed === false) {
            for (const k of (firstKey === null || firstKey === void 0 ? void 0 : firstKey.findDirectExpressions(expressions_1.FieldSub)) || []) {
                primaryKey.keyFields.push(k.concatTokens().toUpperCase());
            }
        }
        else {
            start = 0;
        }
        const secondaryKeys = [];
        for (let i = start; i < typeTableKeys.length; i++) {
            const row = typeTableKeys[i];
            const name = (_c = row.findDirectExpression(expressions_1.Field)) === null || _c === void 0 ? void 0 : _c.concatTokens();
            if (name === undefined) {
                continue;
            }
            const secondary = {
                name: name,
                type: row.findDirectTokenByText("SORTED") ? basic_1.TableAccessType.sorted : basic_1.TableAccessType.hashed,
                isUnique: (row === null || row === void 0 ? void 0 : row.concatTokens().toUpperCase().includes("WITH UNIQUE ")) === true,
                keyFields: [],
            };
            for (const k of (row === null || row === void 0 ? void 0 : row.findDirectExpressions(expressions_1.FieldSub)) || []) {
                secondary.keyFields.push(k.concatTokens().toUpperCase());
            }
            secondaryKeys.push(secondary);
        }
        let keyType = Types.TableKeyType.user;
        if (text.includes(" EMPTY KEY")) {
            keyType = Types.TableKeyType.empty;
        }
        else if (text.includes(" DEFAULT KEY")) {
            keyType = Types.TableKeyType.default;
        }
        const options = {
            withHeader: text.includes(" WITH HEADER LINE"),
            keyType: keyType,
            primaryKey: primaryKey,
            secondary: secondaryKeys,
        };
        let found = undefined;
        if (text.startsWith("TYPE TABLE OF REF TO ")
            || text.startsWith("TYPE STANDARD TABLE OF REF TO ")
            || text.startsWith("TYPE SORTED TABLE OF REF TO ")
            || text.startsWith("TYPE HASHED TABLE OF REF TO ")) {
            found = this.resolveTypeRef(typename);
            if (found) {
                return new Types.TableType(found, options, name);
            }
        }
        else if (text.startsWith("TYPE TABLE OF ")
            || text.startsWith("TYPE STANDARD TABLE OF ")
            || text.startsWith("TYPE SORTED TABLE OF ")
            || text.startsWith("TYPE HASHED TABLE OF ")) {
            found = this.resolveTypeName(typename);
            if (found) {
                return new Types.TableType(found, options, name);
            }
        }
        else if (text.startsWith("LIKE TABLE OF ")
            || text.startsWith("LIKE STANDARD TABLE OF ")
            || text.startsWith("LIKE SORTED TABLE OF ")
            || text.startsWith("LIKE HASHED TABLE OF ")) {
            found = this.resolveLikeName(node);
            if (found) {
                return new Types.TableType(found, options, name);
            }
        }
        else if (text === "TYPE STANDARD TABLE"
            || text === "TYPE SORTED TABLE"
            || text === "TYPE HASHED TABLE"
            || text === "TYPE INDEX TABLE"
            || text === "TYPE ANY TABLE") {
            return new Types.TableType(Types.AnyType.get(), options);
        }
        else if (text.startsWith("TYPE RANGE OF ")) {
            const sub = node.findFirstExpression(Expressions.TypeName);
            found = this.resolveTypeName(sub);
            if (found === undefined) {
                return new Types.UnknownType("TYPE RANGE OF, could not resolve type");
            }
            const structure = new Types.StructureType([
                { name: "sign", type: new Types.CharacterType(1) },
                { name: "option", type: new Types.CharacterType(2) },
                { name: "low", type: found },
                { name: "high", type: found },
            ]);
            options.primaryKey.type = basic_1.TableAccessType.standard;
            return new Types.TableType(structure, options, name);
        }
        else if (text.startsWith("LIKE RANGE OF ")) {
            const sub = node.findFirstExpression(Expressions.FieldChain);
            found = this.resolveLikeName(sub);
            if (found === undefined) {
                return new Types.UnknownType("LIKE RANGE OF, could not resolve type");
            }
            const structure = new Types.StructureType([
                { name: "sign", type: new Types.CharacterType(1) },
                { name: "option", type: new Types.CharacterType(2) },
                { name: "low", type: found },
                { name: "high", type: found },
            ], name);
            options.primaryKey.type = basic_1.TableAccessType.standard;
            return new Types.TableType(structure, options);
        }
        else if (typename && (text.startsWith("TYPE TABLE FOR CREATE ")
            || text.startsWith("TYPE TABLE FOR READ ")
            || text.startsWith("TYPE TABLE FOR DELETE ")
            || text.startsWith("TYPE TABLE FOR UPDATE "))) {
            const name = typename.concatTokens();
            const type = (_d = this.input.scope.getDDIC().lookupDDLS(name)) === null || _d === void 0 ? void 0 : _d.type;
            if (type) {
                return new Types.TableType(basic_1.VoidType.get("RAP-TODO"), options);
            }
            else if (this.input.scope.getDDIC().inErrorNamespace(name)) {
                return new Types.UnknownType(`DDLS ${name} not found`);
            }
            else {
                return Types.VoidType.get(name);
            }
        }
        // fallback to old style syntax, OCCURS etc
        return this.parseType(node, name);
    }
    parseType(node, qualifiedName) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const typeName = node.findFirstExpression(Expressions.TypeName);
        let text = (_a = node.findFirstExpression(Expressions.Type)) === null || _a === void 0 ? void 0 : _a.concatTokens().toUpperCase();
        if (text === undefined) {
            text = (_b = node.findFirstExpression(Expressions.TypeParam)) === null || _b === void 0 ? void 0 : _b.concatTokens().toUpperCase();
        }
        if (text === undefined) {
            text = (_c = node.findFirstExpression(Expressions.TypeTable)) === null || _c === void 0 ? void 0 : _c.concatTokens().toUpperCase();
            if ((text === null || text === void 0 ? void 0 : text.startsWith("TYPE")) === false && (text === null || text === void 0 ? void 0 : text.startsWith("LIKE")) === false) {
                text = "TYPE";
            }
        }
        if (text === undefined) {
            text = (_d = node.findFirstExpression(Expressions.FormParamType)) === null || _d === void 0 ? void 0 : _d.concatTokens().toUpperCase();
        }
        if (text === undefined
            && node.get() instanceof Statements.Parameter
            && node.findDirectTokenByText("LIKE")) {
            text = "LIKE " + (typeName === null || typeName === void 0 ? void 0 : typeName.concatTokens());
        }
        if (text === undefined) {
            text = "TYPE";
        }
        let found = undefined;
        if (text.startsWith("LIKE LINE OF ")) {
            const name = (_e = node.findFirstExpression(Expressions.FieldChain)) === null || _e === void 0 ? void 0 : _e.concatTokens();
            let e = node.findFirstExpression(Expressions.Type);
            if (e === undefined) {
                e = node.findFirstExpression(Expressions.FormParamType);
            }
            if (e === undefined) {
                e = node.findFirstExpression(Expressions.FieldChain);
            }
            const type = this.resolveLikeName(e, false);
            if (type === undefined) {
                return new Types.UnknownType("Type error, could not resolve \"" + name + "\", parseType");
            }
            else if (type instanceof Types.TableType) {
                return type.getRowType();
            }
            else if (type instanceof Types.VoidType) {
                return type;
            }
            else {
                return new Types.UnknownType("Type error, not a table type " + name);
            }
        }
        else if (text.startsWith("LIKE REF TO ")) {
            const name = (_f = node.findFirstExpression(Expressions.FieldChain)) === null || _f === void 0 ? void 0 : _f.concatTokens();
            const type = this.resolveLikeName(node.findFirstExpression(Expressions.Type), false);
            if (type === undefined) {
                return new Types.UnknownType("Type error, could not resolve \"" + name + "\", parseType");
            }
            return new Types.DataReference(type, name);
        }
        else if (text === "TYPE STANDARD TABLE"
            || text === "TYPE SORTED TABLE"
            || text === "TYPE HASHED TABLE"
            || text === "TYPE INDEX TABLE"
            || text === "TYPE ANY TABLE") {
            return new Types.TableType(Types.AnyType.get(), { withHeader: node.concatTokens().toUpperCase().includes("WITH HEADER LINE"), keyType: Types.TableKeyType.default });
        }
        else if (text.startsWith("LIKE ")) {
            let sub = node.findFirstExpression(Expressions.Type);
            if (sub === undefined) {
                sub = node.findFirstExpression(Expressions.FormParamType);
            }
            if (sub === undefined) {
                sub = node.findFirstExpression(Expressions.TypeParam);
            }
            if (sub === undefined) {
                sub = node.findFirstExpression(Expressions.TypeName);
            }
            if (sub === undefined) {
                sub = node.findFirstExpression(Expressions.FieldChain);
            }
            found = this.resolveLikeName(sub);
            if (found && this.isOccurs(node)) {
                found = new Types.TableType(found, { withHeader: text.includes("WITH HEADER LINE"), keyType: Types.TableKeyType.default }, qualifiedName);
            }
            else if (text.endsWith(" WITH HEADER LINE") && found instanceof Types.TableType) {
                found = new Types.TableType(found.getRowType(), { withHeader: true, keyType: Types.TableKeyType.default }, qualifiedName);
            }
        }
        else if (text.startsWith("TYPE LINE OF ")) {
            const sub = node.findFirstExpression(Expressions.TypeName);
            found = this.resolveTypeName(sub);
            if (found instanceof _typed_identifier_1.TypedIdentifier) {
                found = found.getType();
            }
            if (found instanceof Types.TableType) {
                return found.getRowType();
            }
            else if (found instanceof Types.VoidType) {
                return found;
            }
            else if (found instanceof Types.UnknownType) {
                return new Types.UnknownType("TYPE LINE OF, unknown type, " + found.getError());
            }
            else {
                return new Types.UnknownType("TYPE LINE OF, unexpected type, " + (found === null || found === void 0 ? void 0 : found.constructor.name));
            }
        }
        else if (text.startsWith("TYPE REF TO ")) {
            found = this.resolveTypeRef(typeName);
        }
        else if (text.startsWith("TYPE")) {
            found = this.resolveTypeName(typeName, this.findLength(node), this.findDecimals(node), qualifiedName);
            const concat = node.concatTokens().toUpperCase();
            if (found && this.isOccurs(node)) {
                found = new Types.TableType(found, { withHeader: concat.includes(" WITH HEADER LINE"), keyType: Types.TableKeyType.default }, qualifiedName);
            }
            else if (found && concat.includes(" WITH HEADER LINE")) {
                if (found instanceof Types.VoidType) {
                    found = new Types.TableType(found, { withHeader: true, keyType: Types.TableKeyType.default });
                }
                else if (!(found instanceof Types.TableType)) {
                    throw new Error("WITH HEADER LINE can only be used with internal table");
                }
                else {
                    found = new Types.TableType(found.getRowType(), { withHeader: true, keyType: Types.TableKeyType.default });
                }
            }
            if (found === undefined && typeName === undefined) {
                let length = 1;
                const len = node.findDirectExpression(Expressions.ConstantFieldLength);
                if (len) {
                    const int = len.findDirectExpression(Expressions.Integer);
                    if (int) {
                        length = parseInt(int.concatTokens(), 10);
                    }
                }
                found = new Types.CharacterType(length, { qualifiedName: qualifiedName }); // fallback
                if (this.isOccurs(node)) {
                    found = new Types.TableType(found, { withHeader: concat.includes(" WITH HEADER LINE"), keyType: Types.TableKeyType.default }, qualifiedName);
                }
            }
        }
        if (text.includes(" WITH INDICATORS ")) {
            const componentName = (_h = (_g = node.findFirstExpression(Expressions.Type)) === null || _g === void 0 ? void 0 : _g.findDirectExpression(Expressions.ComponentName)) === null || _h === void 0 ? void 0 : _h.concatTokens().toUpperCase();
            if (componentName === undefined) {
                throw new Error("parseType, componentName expected");
            }
            if (found instanceof Types.StructureType) {
                const newComponents = [...found.getComponents()];
                newComponents.push({ name: componentName, type: Types.VoidType.get("INDICATORStodo") });
                found = new Types.StructureType(newComponents, qualifiedName);
            }
        }
        return found;
    }
    /////////////////////
    isOccurs(node) {
        var _a;
        if (node.findDirectTokenByText("OCCURS")) {
            return true;
        }
        else if ((_a = node.findFirstExpression(Expressions.TypeTable)) === null || _a === void 0 ? void 0 : _a.findDirectTokenByText("OCCURS")) {
            return true;
        }
        return false;
    }
    // todo, rewrite this method
    resolveTypeChain(expr) {
        var _a, _b, _c, _d;
        const chainText = expr.concatTokens().toUpperCase();
        if (chainText.includes("-")) {
            // workaround for stuff like "sy-repid"
            const built = this.input.scope.findType(chainText);
            if (built) {
                this.input.scope.addReference(expr.getFirstToken(), built, _reference_1.ReferenceType.TypeReference, this.input.filename);
                return built.getType();
            }
        }
        else if (chainText.includes("=>") === false && chainText.includes("-") === false) {
            return undefined;
        }
        let className;
        let rest = chainText;
        if (chainText.includes("=>")) {
            const split = chainText.split("=>");
            className = split[0];
            rest = split[1];
        }
        else if (chainText.includes("->")) {
            const split = chainText.split("->");
            className = split[0];
            rest = split[1];
        }
        const subs = rest.split("-");
        let foundType = undefined;
        if (className && chainText.includes("=>")) {
            const split = chainText.split("=>");
            const className = split[0];
            // the prefix might be itself
            if ((this.input.scope.getType() === _scope_type_1.ScopeType.Interface
                || this.input.scope.getType() === _scope_type_1.ScopeType.ClassDefinition)
                && this.input.scope.getName().toUpperCase() === className.toUpperCase()) {
                const foundId = this.input.scope.findType(subs[0]);
                foundType = foundId === null || foundId === void 0 ? void 0 : foundId.getType();
                if (foundType === undefined) {
                    return new Types.UnknownType("Could not resolve type " + chainText);
                }
                const token = (_a = expr.getChildren()[2]) === null || _a === void 0 ? void 0 : _a.getFirstToken();
                if (token) {
                    this.input.scope.addReference(token, foundId, _reference_1.ReferenceType.TypeReference, this.input.filename);
                }
            }
            else {
                // lookup in local and global scope
                const obj = this.input.scope.findObjectDefinition(className);
                if (obj === undefined && this.input.scope.getDDIC().inErrorNamespace(className) === false) {
                    this.input.scope.addReference(expr.getFirstToken(), undefined, _reference_1.ReferenceType.ObjectOrientedVoidReference, this.input.filename, { ooName: className.toUpperCase() });
                    return Types.VoidType.get(className);
                }
                else if (obj === undefined) {
                    return new Types.UnknownType("Could not resolve top " + className + ", resolveTypeChain");
                }
                const type = obj instanceof types_1.ClassDefinition ? "CLAS" : "INTF";
                this.input.scope.addReference(expr.getFirstToken(), obj, _reference_1.ReferenceType.ObjectOrientedReference, this.input.filename, { ooType: type, ooName: className });
                const byName = new _object_oriented_1.ObjectOriented(this.input.scope).searchTypeName(obj, subs[0]);
                foundType = byName === null || byName === void 0 ? void 0 : byName.getType();
                if (byName === undefined || foundType === undefined) {
                    return new Types.UnknownType(subs[0] + " not found in class or interface");
                }
                const token = (_b = expr.getChildren()[2]) === null || _b === void 0 ? void 0 : _b.getFirstToken();
                if (token) {
                    this.input.scope.addReference(token, byName, _reference_1.ReferenceType.TypeReference, this.input.filename);
                }
            }
        }
        else if (className && chainText.includes("->")) {
            const varVar = this.input.scope.findVariable(className);
            const foo = varVar === null || varVar === void 0 ? void 0 : varVar.getType();
            if (foo instanceof basic_1.ObjectReferenceType) {
                const typeName = subs[0];
                let id = foo.getIdentifier();
                if (!(id instanceof types_1.ClassDefinition || id instanceof types_1.InterfaceDefinition)) {
                    const found = this.input.scope.findObjectDefinition(foo.getIdentifierName());
                    if (found) {
                        id = found;
                    }
                    else {
                        return new Types.UnknownType(foo.getIdentifierName() + " not found in scope");
                    }
                }
                if (id instanceof types_1.ClassDefinition || id instanceof types_1.InterfaceDefinition) {
                    const type = id instanceof types_1.ClassDefinition ? "CLAS" : "INTF";
                    this.input.scope.addReference(expr.getFirstToken(), id, _reference_1.ReferenceType.ObjectOrientedReference, this.input.filename, { ooType: type, ooName: id.getName() });
                    const byName = new _object_oriented_1.ObjectOriented(this.input.scope).searchTypeName(id, typeName);
                    foundType = byName === null || byName === void 0 ? void 0 : byName.getType();
                    if (byName === undefined || foundType === undefined) {
                        return new Types.UnknownType(typeName + " not found in class or interface");
                    }
                    const token = (_c = expr.getChildren()[2]) === null || _c === void 0 ? void 0 : _c.getFirstToken();
                    if (token) {
                        this.input.scope.addReference(token, byName, _reference_1.ReferenceType.TypeReference, this.input.filename);
                    }
                }
                else {
                    return new Types.UnknownType("Not an object reference, " + className + ", " + id.constructor.name);
                }
            }
            else if (foo === undefined) {
                return new Types.UnknownType(className + " not found in scope");
            }
            else {
                return new Types.UnknownType("Not an object reference, " + className + ", " + foo.constructor.name);
            }
        }
        else {
            const found = this.input.scope.findType(subs[0]);
            foundType = found === null || found === void 0 ? void 0 : found.getType();
            if (foundType === undefined) {
                const typePoolType = (_d = this.input.scope.findTypePoolType(subs[0])) === null || _d === void 0 ? void 0 : _d.getType();
                if (typePoolType) {
                    //          this.scope.addReference(typeName.getFirstToken(), typePoolType, ReferenceType.TypeReference, this.filename);
                    foundType = typePoolType;
                }
                if (foundType === undefined) {
                    const f = this.input.scope.getDDIC().lookupTableOrView(subs[0]);
                    this.input.scope.getDDICReferences().addUsing(this.input.scope.getParentObj(), { object: f.object, filename: this.input.filename, token: expr.getFirstToken() });
                    if (f.type instanceof _typed_identifier_1.TypedIdentifier) {
                        foundType = f.type.getType();
                    }
                    else {
                        foundType = f.type;
                    }
                }
            }
            else {
                this.input.scope.addReference(expr.getFirstToken(), found, _reference_1.ReferenceType.TypeReference, this.input.filename);
            }
            if (foundType === undefined && this.input.scope.getDDIC().inErrorNamespace(subs[0]) === false) {
                this.input.scope.addReference(expr.getFirstToken(), undefined, _reference_1.ReferenceType.VoidType, this.input.filename);
                return Types.VoidType.get(subs[0]);
            }
            else if (foundType instanceof Types.VoidType) {
                this.input.scope.addReference(expr.getFirstToken(), undefined, _reference_1.ReferenceType.VoidType, this.input.filename);
                return foundType;
            }
            else if (foundType === undefined) {
                return new Types.UnknownType("Unknown type " + subs[0]);
            }
        }
        subs.shift();
        while (subs.length > 0) {
            if (foundType instanceof Types.UnknownType
                || foundType instanceof Types.VoidType) {
                return foundType;
            }
            else if (!(foundType instanceof Types.StructureType)) {
                return new Types.UnknownType("Not a structured type");
            }
            foundType = foundType.getComponentByName(subs[0]);
            if (foundType === undefined) {
                return new Types.UnknownType(`Field "${subs[0]}" not found in structure`);
            }
            subs.shift();
        }
        return foundType;
    }
    resolveConstantValue(expr) {
        var _a, _b;
        // todo: rewrite this method
        if (!(expr.get() instanceof Expressions.SimpleFieldChain)) {
            throw new Error("resolveConstantValue");
        }
        const firstNode = expr.getFirstChild();
        const firstToken = firstNode.getFirstToken();
        const firstName = firstToken.getStr();
        if (firstNode.get() instanceof Expressions.Field) {
            const found = this.input.scope.findVariable(firstName);
            const val = found === null || found === void 0 ? void 0 : found.getValue();
            if (typeof val === "string") {
                this.input.scope.addReference(firstToken, found, _reference_1.ReferenceType.DataReadReference, this.input.filename);
                return val;
            }
            else if ((found === null || found === void 0 ? void 0 : found.getType()) instanceof basic_1.StructureType) {
                this.input.scope.addReference(firstToken, found, _reference_1.ReferenceType.DataReadReference, this.input.filename);
            }
            return undefined;
        }
        else if (firstNode.get() instanceof Expressions.ClassName
            && firstName.toLowerCase() === this.input.scope.getName().toLowerCase()
            && (this.input.scope.getType() === _scope_type_1.ScopeType.Interface
                || this.input.scope.getType() === _scope_type_1.ScopeType.ClassDefinition)) {
            const children = expr.getChildren();
            const token = (_a = children[2]) === null || _a === void 0 ? void 0 : _a.getFirstToken();
            const found = this.input.scope.findVariable(token.getStr());
            const val = found === null || found === void 0 ? void 0 : found.getValue();
            if (typeof val === "string") {
                this.input.scope.addReference(firstToken, found, _reference_1.ReferenceType.DataReadReference, this.input.filename);
                return val;
            }
            return undefined;
        }
        else if (firstNode.get() instanceof Expressions.ClassName) {
            const obj = this.input.scope.findObjectDefinition(firstName);
            if (obj === undefined) {
                if (this.input.scope.existsObject(firstName) !== undefined) {
                    return undefined;
                }
                else if (this.input.scope.getDDIC().inErrorNamespace(firstName) === true) {
                    throw new Error("resolveConstantValue, not found: " + firstName);
                }
                else {
                    this.input.scope.addReference(firstNode.getFirstToken(), undefined, _reference_1.ReferenceType.ObjectOrientedVoidReference, this.input.filename, { ooName: firstName.toUpperCase() });
                    return undefined;
                }
            }
            const children = expr.getChildren();
            const token = (_b = children[2]) === null || _b === void 0 ? void 0 : _b.getFirstToken();
            const attr = token.getStr();
            const c = new _object_oriented_1.ObjectOriented(this.input.scope).searchConstantName(obj, attr);
            if (c instanceof class_constant_1.ClassConstant) {
                this.input.scope.addReference(firstToken, obj, _reference_1.ReferenceType.ObjectOrientedReference, this.input.filename, { ooName: obj.getName() });
                this.input.scope.addReference(token, c, _reference_1.ReferenceType.DataReadReference, this.input.filename);
                const val = c.getValue();
                if (typeof val === "string") {
                    return val;
                }
                else if (typeof val === "object" && children[4]) {
                    const name = children[4].getFirstToken().getStr();
                    if (val[name] !== undefined) {
                        return val[name];
                    }
                }
                return undefined;
            }
            throw new Error("resolveConstantValue, constant not found " + attr);
        }
        else {
            throw new Error("resolveConstantValue, unexpected structure");
        }
    }
    resolveTypeRef(chain) {
        var _a;
        if (chain === undefined) {
            return undefined;
        }
        const name = chain.getFirstToken().getStr();
        if (chain.getChildren().length === 1) {
            if (name.toUpperCase() === "OBJECT") {
                return new Types.GenericObjectReferenceType();
            }
            const search = this.input.scope.existsObject(name);
            if (search === null || search === void 0 ? void 0 : search.id) {
                this.input.scope.addReference(chain.getFirstToken(), search.id, _reference_1.ReferenceType.ObjectOrientedReference, this.input.filename, { ooType: search.ooType, ooName: name });
                return new Types.ObjectReferenceType(search.id, { qualifiedName: name, RTTIName: search.RTTIName });
            }
        }
        const found = this.resolveTypeName(chain);
        if (found && found instanceof Types.VoidType) {
            return found;
        }
        else if (found && !(found instanceof Types.UnknownType)) {
            return new Types.DataReference(found);
        }
        else if (chain.concatTokens().toUpperCase() === "DATA") {
            return new Types.DataReference(new Types.DataType());
        }
        if (this.input.scope.isBadiDef(name) === true) {
            return Types.VoidType.get(name);
        }
        if (((_a = this.input.scope.getDDIC()) === null || _a === void 0 ? void 0 : _a.inErrorNamespace(name)) === false) {
            //      this.scope.addReference(chain.getFirstToken(), undefined, ReferenceType.VoidType, this.filename);
            return Types.VoidType.get(name);
        }
        return new Types.UnknownType("REF, unable to resolve " + name);
    }
    findValue(node) {
        const val = node.findFirstExpression(Expressions.Value);
        if (val === undefined) {
            throw new Error("VALUE missing in expression");
        }
        if (val.concatTokens().toUpperCase() === "VALUE IS INITIAL") {
            return undefined;
        }
        const constant = val.findFirstExpression(Expressions.Constant);
        if (constant) {
            const conc = val.findFirstExpression(Expressions.ConcatenatedConstant);
            if (conc) {
                const first = conc.getFirstToken().getStr().substring(0, 1);
                let result = "";
                for (const token of conc.getAllTokens()) {
                    if (token.getStr() === "&") {
                        continue;
                    }
                    else {
                        result += token.getStr().substring(1, token.getStr().length - 1);
                    }
                }
                return first + result + first;
            }
            else {
                return constant.concatTokens();
            }
        }
        const chain = val.findFirstExpression(Expressions.SimpleFieldChain);
        if (chain) {
            return this.resolveConstantValue(chain);
        }
        throw new Error("findValue, unexpected");
    }
    findDecimals(node) {
        var _a, _b;
        const dec = (_b = (_a = node.findDirectExpression(Expressions.Decimals)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.Integer)) === null || _b === void 0 ? void 0 : _b.concatTokens();
        if (dec) {
            return parseInt(dec, 10);
        }
        return undefined;
    }
    findLength(node) {
        const val = node.findFirstExpression(Expressions.Length);
        const flen = node.findFirstExpression(Expressions.ConstantFieldLength);
        if (val && flen) {
            throw new Error("Only specify length once");
        }
        if (flen) {
            const cintExpr = flen.findFirstExpression(Expressions.Integer);
            if (cintExpr) {
                return this.parseInt(cintExpr.concatTokens());
            }
            const cchain = flen.findFirstExpression(Expressions.SimpleFieldChain);
            if (cchain) {
                const val = this.resolveConstantValue(cchain);
                return this.parseInt(val);
            }
        }
        if (val === undefined) {
            return 1;
        }
        const intExpr = val.findFirstExpression(Expressions.Integer);
        if (intExpr) {
            return this.parseInt(intExpr.concatTokens());
        }
        const strExpr = val.findFirstExpression(Expressions.ConstantString);
        if (strExpr) {
            return this.parseInt(strExpr.concatTokens());
        }
        const chain = val.findFirstExpression(Expressions.SimpleFieldChain);
        if (chain) {
            const val = this.resolveConstantValue(chain);
            return this.parseInt(val);
        }
        throw new Error("Unexpected, findLength");
    }
    parseInt(text) {
        if (text === undefined) {
            return undefined;
        }
        if (text.startsWith("'")) {
            text = text.split("'")[1];
        }
        else if (text.startsWith("`")) {
            text = text.split("`")[1];
        }
        return parseInt(text, 10);
    }
}
exports.BasicTypes = BasicTypes;
