// @abaplint/core/build/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rename = exports.PrettyPrinter = exports.Position = exports.CurrentScope = exports.RulesRunner = exports.SpaghettiScope = exports.SyntaxLogic = exports.ABAPObject = exports.Tokens = exports.ExpressionsCDS = exports.CDSParser = exports.LanguageServerTypes = exports.DDLParser = exports.AbstractFile = exports.ABAPFile = exports.MemoryFile = exports.Edits = exports.ReferenceType = exports.NativeSQL = exports.MacroContent = exports.MacroCall = exports.SpaghettiScopeNode = exports.Token = exports.ScopeType = exports.BasicTypes = exports.TypedIdentifier = exports.AbstractType = exports.VirtualPosition = exports.Comment = exports.Unknown = exports.Empty = exports.Identifier = exports.Nodes = exports.Types = exports.Expressions = exports.Statements = exports.Structures = exports.SkipLogic = exports.Objects = exports.ArtifactsRules = exports.ArtifactsObjects = exports.ArtifactsABAP = exports.BuiltIn = exports.MethodLengthStats = exports.LanguageServer = exports.Registry = exports.CyclomaticComplexityStats = exports.Version = exports.Config = exports.Issue = void 0;
exports.LSPEdit = exports.RuleTag = exports.Severity = exports.Visibility = exports.Info = exports.Diagnostics = void 0;
const issue_1 = require("./issue");
Object.defineProperty(exports, "Issue", { enumerable: true, get: function () { return issue_1.Issue; } });
const config_1 = require("./config");
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return config_1.Config; } });
const version_1 = require("./version");
Object.defineProperty(exports, "Version", { enumerable: true, get: function () { return version_1.Version; } });
const registry_1 = require("./registry");
Object.defineProperty(exports, "Registry", { enumerable: true, get: function () { return registry_1.Registry; } });
const method_length_stats_1 = require("./utils/method_length_stats");
Object.defineProperty(exports, "MethodLengthStats", { enumerable: true, get: function () { return method_length_stats_1.MethodLengthStats; } });
const language_server_1 = require("./lsp/language_server");
Object.defineProperty(exports, "LanguageServer", { enumerable: true, get: function () { return language_server_1.LanguageServer; } });
const LanguageServerTypes = require("./lsp/_interfaces");
exports.LanguageServerTypes = LanguageServerTypes;
const artifacts_1 = require("./abap/artifacts");
Object.defineProperty(exports, "ArtifactsABAP", { enumerable: true, get: function () { return artifacts_1.ArtifactsABAP; } });
const artifacts_objects_1 = require("./artifacts_objects");
Object.defineProperty(exports, "ArtifactsObjects", { enumerable: true, get: function () { return artifacts_objects_1.ArtifactsObjects; } });
const artifacts_rules_1 = require("./artifacts_rules");
Object.defineProperty(exports, "ArtifactsRules", { enumerable: true, get: function () { return artifacts_rules_1.ArtifactsRules; } });
const _abap_object_1 = require("./objects/_abap_object");
Object.defineProperty(exports, "ABAPObject", { enumerable: true, get: function () { return _abap_object_1.ABAPObject; } });
const syntax_1 = require("./abap/5_syntax/syntax");
Object.defineProperty(exports, "SyntaxLogic", { enumerable: true, get: function () { return syntax_1.SyntaxLogic; } });
const spaghetti_scope_1 = require("./abap/5_syntax/spaghetti_scope");
Object.defineProperty(exports, "SpaghettiScope", { enumerable: true, get: function () { return spaghetti_scope_1.SpaghettiScope; } });
Object.defineProperty(exports, "SpaghettiScopeNode", { enumerable: true, get: function () { return spaghetti_scope_1.SpaghettiScopeNode; } });
const _typed_identifier_1 = require("./abap/types/_typed_identifier");
Object.defineProperty(exports, "TypedIdentifier", { enumerable: true, get: function () { return _typed_identifier_1.TypedIdentifier; } });
const _abstract_type_1 = require("./abap/types/basic/_abstract_type");
Object.defineProperty(exports, "AbstractType", { enumerable: true, get: function () { return _abstract_type_1.AbstractType; } });
const _scope_type_1 = require("./abap/5_syntax/_scope_type");
Object.defineProperty(exports, "ScopeType", { enumerable: true, get: function () { return _scope_type_1.ScopeType; } });
const _current_scope_1 = require("./abap/5_syntax/_current_scope");
Object.defineProperty(exports, "CurrentScope", { enumerable: true, get: function () { return _current_scope_1.CurrentScope; } });
const Objects = require("./objects");
exports.Objects = Objects;
const abstract_token_1 = require("./abap/1_lexer/tokens/abstract_token");
Object.defineProperty(exports, "Token", { enumerable: true, get: function () { return abstract_token_1.AbstractToken; } });
const Statements = require("./abap/2_statements/statements");
exports.Statements = Statements;
const Expressions = require("./abap/2_statements/expressions");
exports.Expressions = Expressions;
const ExpressionsCDS = require("./cds/expressions");
exports.ExpressionsCDS = ExpressionsCDS;
const Structures = require("./abap/3_structures/structures");
exports.Structures = Structures;
const Nodes = require("./abap/nodes");
exports.Nodes = Nodes;
const BasicTypes = require("./abap/types/basic");
exports.BasicTypes = BasicTypes;
const Types = require("./abap/types");
exports.Types = Types;
const Tokens = require("./abap/1_lexer/tokens");
exports.Tokens = Tokens;
const position_1 = require("./position");
Object.defineProperty(exports, "Position", { enumerable: true, get: function () { return position_1.Position; } });
const virtual_position_1 = require("./virtual_position");
Object.defineProperty(exports, "VirtualPosition", { enumerable: true, get: function () { return virtual_position_1.VirtualPosition; } });
const _abstract_file_1 = require("./files/_abstract_file");
Object.defineProperty(exports, "AbstractFile", { enumerable: true, get: function () { return _abstract_file_1.AbstractFile; } });
const pretty_printer_1 = require("./pretty_printer/pretty_printer");
Object.defineProperty(exports, "PrettyPrinter", { enumerable: true, get: function () { return pretty_printer_1.PrettyPrinter; } });
const _statement_1 = require("./abap/2_statements/statements/_statement");
Object.defineProperty(exports, "Empty", { enumerable: true, get: function () { return _statement_1.Empty; } });
Object.defineProperty(exports, "Unknown", { enumerable: true, get: function () { return _statement_1.Unknown; } });
Object.defineProperty(exports, "Comment", { enumerable: true, get: function () { return _statement_1.Comment; } });
Object.defineProperty(exports, "MacroCall", { enumerable: true, get: function () { return _statement_1.MacroCall; } });
Object.defineProperty(exports, "MacroContent", { enumerable: true, get: function () { return _statement_1.MacroContent; } });
Object.defineProperty(exports, "NativeSQL", { enumerable: true, get: function () { return _statement_1.NativeSQL; } });
const edit_helper_1 = require("./edit_helper");
const _builtin_1 = require("./abap/5_syntax/_builtin");
Object.defineProperty(exports, "BuiltIn", { enumerable: true, get: function () { return _builtin_1.BuiltIn; } });
const abap_file_1 = require("./abap/abap_file");
Object.defineProperty(exports, "ABAPFile", { enumerable: true, get: function () { return abap_file_1.ABAPFile; } });
const memory_file_1 = require("./files/memory_file");
Object.defineProperty(exports, "MemoryFile", { enumerable: true, get: function () { return memory_file_1.MemoryFile; } });
const renamer_1 = require("./objects/rename/renamer");
Object.defineProperty(exports, "Rename", { enumerable: true, get: function () { return renamer_1.Renamer; } });
const Info = require("./abap/4_file_information/_abap_file_information");
exports.Info = Info;
const visibility_1 = require("./abap/4_file_information/visibility");
Object.defineProperty(exports, "Visibility", { enumerable: true, get: function () { return visibility_1.Visibility; } });
const _identifier_1 = require("./abap/4_file_information/_identifier");
Object.defineProperty(exports, "Identifier", { enumerable: true, get: function () { return _identifier_1.Identifier; } });
const severity_1 = require("./severity");
Object.defineProperty(exports, "Severity", { enumerable: true, get: function () { return severity_1.Severity; } });
const ddl_parser_1 = require("./ddl/ddl_parser");
Object.defineProperty(exports, "DDLParser", { enumerable: true, get: function () { return ddl_parser_1.DDLParser; } });
const cds_parser_1 = require("./cds/cds_parser");
Object.defineProperty(exports, "CDSParser", { enumerable: true, get: function () { return cds_parser_1.CDSParser; } });
const rules_runner_1 = require("./rules_runner");
Object.defineProperty(exports, "RulesRunner", { enumerable: true, get: function () { return rules_runner_1.RulesRunner; } });
const _irule_1 = require("./rules/_irule");
Object.defineProperty(exports, "RuleTag", { enumerable: true, get: function () { return _irule_1.RuleTag; } });
const cyclomatic_complexity_stats_1 = require("./utils/cyclomatic_complexity_stats");
Object.defineProperty(exports, "CyclomaticComplexityStats", { enumerable: true, get: function () { return cyclomatic_complexity_stats_1.CyclomaticComplexityStats; } });
const skip_logic_1 = require("./skip_logic");
Object.defineProperty(exports, "SkipLogic", { enumerable: true, get: function () { return skip_logic_1.SkipLogic; } });
const diagnostics_1 = require("./lsp/diagnostics");
Object.defineProperty(exports, "Diagnostics", { enumerable: true, get: function () { return diagnostics_1.Diagnostics; } });
const _edit_1 = require("./lsp/_edit");
Object.defineProperty(exports, "LSPEdit", { enumerable: true, get: function () { return _edit_1.LSPEdit; } });
const _reference_1 = require("./abap/5_syntax/_reference");
Object.defineProperty(exports, "ReferenceType", { enumerable: true, get: function () { return _reference_1.ReferenceType; } });
// do not include this file from anywhere within abaplint
// https://github.com/abaplint/abaplint/issues/873
const Edits = {
    applyEditSingle: edit_helper_1.applyEditSingle,
    applyEditList: edit_helper_1.applyEditList,
};
exports.Edits = Edits;
