// @abaplint/core/build/src/abap/types/_typed_identifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypedIdentifier = void 0;
const _identifier_1 = require("../4_file_information/_identifier");
class TypedIdentifier extends _identifier_1.Identifier {
    [Symbol.for("debug.description")]() {
        return `${this.constructor.name} ${this.getName()}:${this.getType().constructor.name}`;
    }
    static from(id, type, meta) {
        return new TypedIdentifier(id.getToken(), id.getFilename(), type, meta);
    }
    constructor(token, filename, type, meta, value) {
        super(token, filename);
        if (type instanceof TypedIdentifier) {
            this.type = type.getType();
        }
        else {
            this.type = type;
        }
        this.value = value;
        this.meta = [];
        if (meta) {
            this.meta = meta;
        }
    }
    toText() {
        return "Identifier: ```" + this.getName() + "```";
    }
    getType() {
        return this.type;
    }
    getMeta() {
        return this.meta;
    }
    getValue() {
        return this.value;
    }
}
exports.TypedIdentifier = TypedIdentifier;
