// @abaplint/core/build/src/abap/types/class_attributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Attributes = void 0;
const Structures = require("../3_structures/structures");
const Statements = require("../2_statements/statements");
const Expressions = require("../2_statements/expressions");
const class_attribute_1 = require("./class_attribute");
const class_constant_1 = require("./class_constant");
const nodes_1 = require("../nodes");
const visibility_1 = require("../4_file_information/visibility");
const class_data_1 = require("../5_syntax/statements/class_data");
const class_data_2 = require("../5_syntax/structures/class_data");
const data_1 = require("../5_syntax/statements/data");
const constant_1 = require("../5_syntax/statements/constant");
const data_2 = require("../5_syntax/structures/data");
const type_enum_1 = require("../5_syntax/structures/type_enum");
const constants_1 = require("../5_syntax/structures/constants");
const type_definitions_1 = require("./type_definitions");
const types_1 = require("../5_syntax/structures/types");
const type_1 = require("../5_syntax/statements/type");
const _reference_1 = require("../5_syntax/_reference");
const alias_1 = require("./alias");
class Attributes {
    constructor(node, input) {
        this.static = [];
        this.instance = [];
        this.constants = [];
        this.aliases = [];
        this.tlist = [];
        this.filename = input.filename;
        this.parse(node, input);
        this.types = new type_definitions_1.TypeDefinitions(this.tlist);
    }
    getTypes() {
        return this.types;
    }
    getStatic() {
        return this.static;
    }
    getAliases() {
        return this.aliases;
    }
    getAll() {
        let res = [];
        res = res.concat(this.static);
        res = res.concat(this.instance);
        return res;
    }
    getStaticsByVisibility(visibility) {
        const attributes = [];
        for (const attr of this.static) {
            if (attr.getVisibility() === visibility) {
                attributes.push(attr);
            }
        }
        return attributes;
    }
    getInstance() {
        return this.instance;
    }
    getInstancesByVisibility(visibility) {
        const attributes = [];
        for (const attr of this.instance) {
            if (attr.getVisibility() === visibility) {
                attributes.push(attr);
            }
        }
        return attributes;
    }
    getConstants() {
        return this.constants;
    }
    getConstantsByVisibility(visibility) {
        const attributes = [];
        for (const attr of this.constants) {
            if (attr.getVisibility() === visibility) {
                attributes.push(attr);
            }
        }
        return attributes;
    }
    // todo, optimize
    findByName(name) {
        const upper = name.toUpperCase();
        for (const a of this.getStatic()) {
            if (a.getName().toUpperCase() === upper) {
                return a;
            }
        }
        for (const a of this.getInstance()) {
            if (a.getName().toUpperCase() === upper) {
                return a;
            }
        }
        for (const a of this.getConstants()) {
            if (a.getName().toUpperCase() === upper) {
                return a;
            }
        }
        return undefined;
    }
    /////////////////////////////
    parse(node, input) {
        const cdef = node.findDirectStructure(Structures.ClassDefinition);
        if (cdef) {
            this.parseSection(cdef.findDirectStructure(Structures.PublicSection), visibility_1.Visibility.Public, input);
            this.parseSection(cdef.findDirectStructure(Structures.ProtectedSection), visibility_1.Visibility.Protected, input);
            this.parseSection(cdef.findDirectStructure(Structures.PrivateSection), visibility_1.Visibility.Private, input);
            return;
        }
        const idef = node.findDirectStructure(Structures.Interface);
        if (idef) {
            this.parseSection(idef.findDirectStructure(Structures.SectionContents), visibility_1.Visibility.Public, input);
            return;
        }
        throw new Error("MethodDefinition, expected ClassDefinition or InterfaceDefinition");
    }
    parseSection(node, visibility, input) {
        if (node === undefined) {
            return;
        }
        for (const c of node.getChildren()) {
            const ctyp = c.get();
            if (c instanceof nodes_1.StructureNode) {
                if (ctyp instanceof Structures.Data) {
                    const found = data_2.Data.runSyntax(c, input);
                    if (found !== undefined) {
                        const attr = new class_attribute_1.ClassAttribute(found, visibility, found.getMeta(), found.getValue());
                        this.instance.push(attr);
                        input.scope.addIdentifier(attr);
                    }
                }
                else if (ctyp instanceof Structures.ClassData) {
                    const found = new class_data_2.ClassData().runSyntax(c, input);
                    if (found !== undefined) {
                        const attr = new class_attribute_1.ClassAttribute(found, visibility, found.getMeta(), found.getValue());
                        this.static.push(attr);
                        input.scope.addIdentifier(attr);
                    }
                }
                else if (ctyp instanceof Structures.Constants) {
                    const { type: found, values } = new constants_1.Constants().runSyntax(c, input);
                    if (found !== undefined) {
                        const attr = new class_constant_1.ClassConstant(found, visibility, values);
                        this.constants.push(attr);
                        input.scope.addIdentifier(attr);
                    }
                }
                else if (ctyp instanceof Structures.TypeEnum) {
                    const { values, types } = new type_enum_1.TypeEnum().runSyntax(c, input);
                    for (const v of values) {
                        // for now add ENUM values as constants
                        const attr = new class_constant_1.ClassConstant(v, visibility, "novalueClassAttributeEnum");
                        this.constants.push(attr);
                        input.scope.addIdentifier(attr);
                    }
                    for (const t of types) {
                        this.tlist.push({ type: t, visibility });
                        //            scope.addIdentifier(attr);
                    }
                }
                else if (ctyp instanceof Structures.Types) {
                    const res = new types_1.Types().runSyntax(c, input);
                    if (res) {
                        input.scope.addType(res);
                        this.tlist.push({ type: res, visibility });
                    }
                }
                else {
                    // begin recursion
                    this.parseSection(c, visibility, input);
                }
            }
            else if (c instanceof nodes_1.StatementNode) {
                if (ctyp instanceof Statements.Data) {
                    this.instance.push(this.parseAttribute(c, visibility, input));
                }
                else if (ctyp instanceof Statements.ClassData) {
                    this.static.push(this.parseAttribute(c, visibility, input));
                }
                else if (ctyp instanceof Statements.Aliases) {
                    this.parseAlias(c, visibility, input);
                }
                else if (ctyp instanceof Statements.Constant) {
                    const found = new constant_1.Constant().runSyntax(c, input);
                    if (found) {
                        const attr = new class_constant_1.ClassConstant(found, visibility, found.getValue());
                        this.constants.push(attr);
                        input.scope.addIdentifier(attr);
                    }
                }
                else if (ctyp instanceof Statements.Type) {
                    const res = new type_1.Type().runSyntax(c, input);
                    if (res) {
                        input.scope.addType(res);
                        this.tlist.push({ type: res, visibility });
                    }
                }
            }
        }
    }
    parseAlias(node, visibility, input) {
        const aliasName = node.findFirstExpression(Expressions.SimpleName).getFirstToken();
        const compToken = node.findFirstExpression(Expressions.Field).getFirstToken();
        const compName = compToken.getStr();
        this.aliases.push(new alias_1.Alias(aliasName, visibility, compName, this.filename));
        if (compName.includes("~")) {
            const name = compName.split("~")[0];
            const idef = input.scope.findInterfaceDefinition(name);
            if (idef) {
                input.scope.addReference(compToken, idef, _reference_1.ReferenceType.ObjectOrientedReference, input.filename, { ooName: name.toUpperCase(), ooType: "INTF" });
                const foundType = idef.getTypeDefinitions().getByName(compName.split("~")[1]);
                if (foundType) {
                    input.scope.addTypeNamed(aliasName.getStr(), foundType);
                }
                const foundAttribute = idef.getAttributes().findByName(compName.split("~")[1]);
                if (foundAttribute) {
                    input.scope.addNamedIdentifier(aliasName.getStr(), foundAttribute);
                }
            }
        }
    }
    parseAttribute(node, visibility, input) {
        let found = undefined;
        const s = node.get();
        if (s instanceof Statements.Data) {
            found = new data_1.Data().runSyntax(node, input);
        }
        else if (s instanceof Statements.ClassData) {
            found = new class_data_1.ClassData().runSyntax(node, input);
        }
        else {
            throw new Error("ClassAttribute, unexpected node, 1, " + this.filename);
        }
        if (found === undefined) {
            throw new Error("ClassAttribute, unexpected node, " + this.filename);
        }
        input.scope.addIdentifier(found);
        return new class_attribute_1.ClassAttribute(found, visibility, found.getMeta(), found.getValue());
    }
}
exports.Attributes = Attributes;
