// @abaplint/core/build/src/abap/types/class_definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClassDefinition = void 0;
const method_definitions_1 = require("./method_definitions");
const expressions_1 = require("../2_statements/expressions");
const Statements = require("../2_statements/statements");
const Structures = require("../3_structures/structures");
const Expressions = require("../2_statements/expressions");
const class_attributes_1 = require("./class_attributes");
const _identifier_1 = require("../4_file_information/_identifier");
const _scope_type_1 = require("../5_syntax/_scope_type");
const event_definition_1 = require("./event_definition");
const visibility_1 = require("../4_file_information/visibility");
const _object_oriented_1 = require("../5_syntax/_object_oriented");
const _reference_1 = require("../5_syntax/_reference");
class ClassDefinition extends _identifier_1.Identifier {
    constructor(node, input) {
        var _a;
        if (!(node.get() instanceof Structures.ClassDefinition)) {
            throw new Error("ClassDefinition, unexpected node type");
        }
        const def = node.findFirstStatement(Statements.ClassDefinition);
        const name = def.findDirectExpression(Expressions.ClassName).getFirstToken();
        super(name, input.filename);
        input.scope.addClassDefinition(this);
        this.events = [];
        this.implementing = [];
        this.globalValue = def.findFirstExpression(Expressions.ClassGlobal) !== undefined;
        this.finalValue = def.findFirstExpression(Expressions.ClassFinal) !== undefined;
        input.scope.push(_scope_type_1.ScopeType.ClassDefinition, name.getStr(), name.getStart(), input.filename);
        this.superClass = this.findSuper(def, input);
        this.friends = this.findFriends(def, input);
        this.parse(input, node);
        const helper = new _object_oriented_1.ObjectOriented(input.scope);
        helper.fromSuperClassesAndInterfaces(this);
        this.attributes = new class_attributes_1.Attributes(node, input);
        this.types = this.attributes.getTypes();
        this.aliases = this.attributes.getAliases();
        const events = node.findAllStatements(Statements.Events);
        for (const e of events) {
            const eventName = (_a = e.findDirectExpression(Expressions.EventName)) === null || _a === void 0 ? void 0 : _a.concatTokens().toUpperCase();
            if (this.events.find(ev => ev.getName().toUpperCase() === eventName) !== undefined) {
                throw new Error("Event " + eventName + " already defined");
            }
            this.events.push(new event_definition_1.EventDefinition(e, visibility_1.Visibility.Public, input)); // todo, all these are not Public
        }
        this.methodDefs = new method_definitions_1.MethodDefinitions(node, input);
        input.scope.pop(node.getLastToken().getEnd());
        const concat = def.concatTokens().toUpperCase();
        this.testing = concat.includes(" FOR TESTING");
        this.sharedMemory = concat.includes(" SHARED MEMORY ENABLED");
        this.abstract = (def === null || def === void 0 ? void 0 : def.findDirectTokenByText("ABSTRACT")) !== undefined;
        // perform checks after everything has been initialized
        this.checkMethodsFromSuperClasses(input.scope);
        this.checkMethodNameLength();
    }
    getFriends() {
        return this.friends;
    }
    getEvents() {
        return this.events;
    }
    getMethodDefinitions() {
        return this.methodDefs;
    }
    getTypeDefinitions() {
        return this.types;
    }
    getSuperClass() {
        return this.superClass;
    }
    getAttributes() {
        return this.attributes;
    }
    isGlobal() {
        return this.globalValue;
    }
    isFinal() {
        return this.finalValue;
    }
    getImplementing() {
        return this.implementing;
    }
    getAliases() {
        return this.aliases;
    }
    isForTesting() {
        return this.testing;
    }
    isAbstract() {
        return this.abstract;
    }
    isSharedMemory() {
        return this.sharedMemory;
    }
    /*
      public getEvents() {
      }
    */
    ///////////////////
    findSuper(def, input) {
        var _a;
        const token = (_a = def === null || def === void 0 ? void 0 : def.findDirectExpression(expressions_1.SuperClassName)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
        this.addReference(token, input);
        const name = token === null || token === void 0 ? void 0 : token.getStr();
        return name;
    }
    checkMethodNameLength() {
        for (const m of this.methodDefs.getAll()) {
            if (m.getName().length > 30 && m.getName().includes("~") === false) {
                const message = `Method name "${m.getName()}" is too long, maximum length is 30 characters`;
                throw new Error(message);
            }
        }
    }
    checkMethodsFromSuperClasses(scope) {
        var _a;
        let sup = this.getSuperClass();
        const names = new Set();
        while (sup !== undefined) {
            const cdef = scope.findClassDefinition(sup);
            for (const m of ((_a = cdef === null || cdef === void 0 ? void 0 : cdef.getMethodDefinitions()) === null || _a === void 0 ? void 0 : _a.getAll()) || []) {
                const name = m.getName().toUpperCase();
                if (m.getVisibility() === visibility_1.Visibility.Private) {
                    continue;
                }
                else if (name === "CONSTRUCTOR"
                    || name === "DESTRUCTOR"
                    || name === "CLASS_CONSTRUCTOR") {
                    continue;
                }
                names.add(name);
            }
            for (const a of (cdef === null || cdef === void 0 ? void 0 : cdef.getAliases()) || []) {
                names.add(a.getName().toUpperCase());
            }
            sup = cdef === null || cdef === void 0 ? void 0 : cdef.getSuperClass();
        }
        for (const m of this.getMethodDefinitions().getAll()) {
            if (names.has(m.getName().toUpperCase())
                && m.isRedefinition() === false) {
                throw new Error(`${m.getName().toUpperCase()} already declared in superclass`);
            }
        }
    }
    findFriends(def, input) {
        var _a;
        const result = [];
        for (const n of ((_a = def === null || def === void 0 ? void 0 : def.findDirectExpression(Expressions.ClassFriends)) === null || _a === void 0 ? void 0 : _a.findDirectExpressions(Expressions.ClassName)) || []) {
            const token = n.getFirstToken();
            this.addReference(token, input);
            const name = token.getStr();
            result.push(name);
        }
        return result;
    }
    addReference(token, input) {
        const name = token === null || token === void 0 ? void 0 : token.getStr();
        if (name) {
            const s = input.scope.findClassDefinition(name);
            if (s) {
                input.scope.addReference(token, s, _reference_1.ReferenceType.ObjectOrientedReference, input.filename, { ooName: name.toUpperCase(), ooType: "CLAS" });
            }
            else if (input.scope.getDDIC().inErrorNamespace(name) === false) {
                input.scope.addReference(token, undefined, _reference_1.ReferenceType.ObjectOrientedVoidReference, input.filename);
            }
        }
    }
    parse(input, inputNode) {
        var _a;
        for (const node of inputNode.findAllStatements(Statements.InterfaceDef)) {
            const partial = node.findDirectTokenByText("PARTIALLY") !== undefined;
            const token = (_a = node.findFirstExpression(Expressions.InterfaceName)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
            if (token === undefined) {
                throw new Error("ClassDefinition, unable to find interface token");
            }
            const name = token.getStr().toUpperCase();
            this.implementing.push({ name, partial });
            const intf = input.scope.findInterfaceDefinition(name);
            if (intf) {
                input.scope.addReference(token, intf, _reference_1.ReferenceType.ObjectOrientedReference, input.filename, { ooName: name.toUpperCase(), ooType: "INTF" });
            }
            else if (input.scope.getDDIC().inErrorNamespace(name) === false) {
                input.scope.addReference(token, undefined, _reference_1.ReferenceType.ObjectOrientedVoidReference, input.filename, { ooName: name.toUpperCase(), ooType: "INTF" });
            }
            else {
                input.scope.addReference(token, undefined, _reference_1.ReferenceType.ObjectOrientedUnknownReference, input.filename, { ooName: name.toUpperCase(), ooType: "INTF" });
            }
        }
    }
}
exports.ClassDefinition = ClassDefinition;
