// @abaplint/core/build/src/abap/types/event_definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventDefinition = void 0;
const _identifier_1 = require("../4_file_information/_identifier");
const Expressions = require("../2_statements/expressions");
const events_1 = require("../2_statements/statements/events");
const expressions_1 = require("../2_statements/expressions");
const method_param_1 = require("../5_syntax/expressions/method_param");
class EventDefinition extends _identifier_1.Identifier {
    constructor(node, _visibility, input) {
        if (!(node.get() instanceof events_1.Events)) {
            throw new Error("MethodDefinition, expected MethodDef as part of input node");
        }
        const found = node.findFirstExpression(Expressions.EventName);
        if (found === undefined) {
            throw new Error("MethodDefinition, expected MethodDef as part of input node");
        }
        super(found.getFirstToken(), input.filename);
        this.parameters = [];
        this.parse(node, input);
        this.is_static = node.getFirstToken().getStr().toUpperCase() === "CLASS";
    }
    getParameters() {
        return this.parameters;
    }
    isStatic() {
        return this.is_static;
    }
    ///////////////
    parse(node, input) {
        for (const e of node.findAllExpressions(expressions_1.MethodParam)) {
            this.parameters.push(method_param_1.MethodParam.runSyntax(e, input, []));
        }
    }
}
exports.EventDefinition = EventDefinition;
