// @abaplint/core/build/src/abap/types/form_definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormDefinition = void 0;
const Statements = require("../2_statements/statements");
const Expressions = require("../2_statements/expressions");
const Tokens = require("../1_lexer/tokens");
const _identifier_1 = require("../4_file_information/_identifier");
const nodes_1 = require("../nodes");
const _typed_identifier_1 = require("./_typed_identifier");
const form_param_1 = require("../5_syntax/expressions/form_param");
const basic_1 = require("./basic");
class FormDefinition extends _identifier_1.Identifier {
    constructor(node, input) {
        const st = node instanceof nodes_1.StructureNode ? node.findFirstStatement(Statements.Form) : node;
        // FORMs can contain a dash in the name
        const formName = st.findFirstExpression(Expressions.FormName);
        const pos = formName.getFirstToken().getStart();
        const name = formName.concatTokens();
        const nameToken = new Tokens.Identifier(pos, name);
        super(nameToken, input.filename);
        this.node = st;
        this.tableParameters = this.findTables(input);
        this.usingParameters = this.findType(Expressions.FormUsing, input);
        this.changingParameters = this.findType(Expressions.FormChanging, input);
    }
    getTablesParameters() {
        return this.tableParameters;
    }
    getUsingParameters() {
        return this.usingParameters;
    }
    getChangingParameters() {
        return this.changingParameters;
    }
    ///////////////
    findTables(input) {
        const ret = [];
        const tables = this.node.findFirstExpression(Expressions.FormTables);
        if (tables === undefined) {
            return [];
        }
        for (const param of tables.findAllExpressions(Expressions.FormParam)) {
            if (param.getChildren().length === 1) {
                // untyped TABLES parameter
                ret.push(new _typed_identifier_1.TypedIdentifier(param.getFirstToken(), input.filename, basic_1.VoidType.get("FORM:UNTYPED"), ["form_parameter" /* IdentifierMeta.FormParameter */]));
            }
            else {
                const p = form_param_1.FormParam.runSyntax(param, input);
                let type = p.getType();
                const isStructure = param.findDirectTokenByText("STRUCTURE") !== undefined;
                if (isStructure) {
                    type = new basic_1.TableType(type, { withHeader: true, keyType: basic_1.TableKeyType.default });
                }
                if (type instanceof basic_1.TableType) {
                    type = new basic_1.TableType(type.getRowType(), { withHeader: true, keyType: basic_1.TableKeyType.default });
                }
                else if (!(type instanceof basic_1.UnknownType) && !(type instanceof basic_1.VoidType)) {
                    type = new basic_1.UnknownType("FORM TABLES type must be table type");
                }
                ret.push(new _typed_identifier_1.TypedIdentifier(p.getToken(), input.filename, type, ["form_parameter" /* IdentifierMeta.FormParameter */]));
            }
        }
        return ret;
    }
    findType(type, input) {
        const found = this.node.findFirstExpression(type);
        if (found === undefined) {
            return [];
        }
        return this.findParams(found, input);
    }
    findParams(node, input) {
        const res = [];
        for (const param of node.findAllExpressions(Expressions.FormParam)) {
            const p = form_param_1.FormParam.runSyntax(param, input);
            res.push(p);
        }
        return res;
    }
}
exports.FormDefinition = FormDefinition;
