// @abaplint/core/build/src/abap/types/function_module_definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionModuleDefinition = exports.FunctionModuleType = exports.FunctionModuleParameterDirection = void 0;
const xml_utils_1 = require("../../xml_utils");
var FunctionModuleParameterDirection;
(function (FunctionModuleParameterDirection) {
    FunctionModuleParameterDirection["importing"] = "importing";
    FunctionModuleParameterDirection["exporting"] = "exporting";
    FunctionModuleParameterDirection["changing"] = "changing";
    FunctionModuleParameterDirection["tables"] = "tables";
})(FunctionModuleParameterDirection || (exports.FunctionModuleParameterDirection = FunctionModuleParameterDirection = {}));
var FunctionModuleType;
(function (FunctionModuleType) {
    FunctionModuleType["regular"] = "regular";
    FunctionModuleType["remote"] = "remoteEnabled";
    FunctionModuleType["update"] = "update";
})(FunctionModuleType || (exports.FunctionModuleType = FunctionModuleType = {}));
class FunctionModuleDefinition {
    constructor(data) {
        this.parse(data);
    }
    getParameters() {
        return this.parameters;
    }
    getModuleType() {
        return this.moduleType;
    }
    getDescription() {
        return this.description;
    }
    getName() {
        return this.name;
    }
    isGlobalParameters() {
        return this.globalParameters;
    }
    ///////////////
    parse(data) {
        if (data.FUNCNAME === undefined) {
            throw new Error("Function module name undefined");
        }
        this.name = data.FUNCNAME;
        this.description = data.SHORT_TEXT;
        this.parameters = [];
        this.globalParameters = data.GLOBAL_FLAG === "X";
        this.moduleType = FunctionModuleType.regular;
        if (data.REMOTE_CALL === "R") {
            this.moduleType = FunctionModuleType.remote;
        }
        else if (data.UPDATE_TASK !== undefined) {
            this.moduleType = FunctionModuleType.update;
        }
        if (data.IMPORT) {
            for (const param of (0, xml_utils_1.xmlToArray)(data.IMPORT.RSIMP)) {
                if (param.PARAMETER === undefined) {
                    throw new Error("Function module name parameter undefined, importing");
                }
                this.parameters.push({
                    name: param.PARAMETER,
                    direction: FunctionModuleParameterDirection.importing,
                    type: param.TYP || param.DBFIELD,
                    optional: param.OPTIONAL === "X",
                    defaultValue: param.DEFAULT,
                });
            }
        }
        if (data.CHANGING) {
            for (const param of (0, xml_utils_1.xmlToArray)(data.CHANGING.RSCHA)) {
                if (param.PARAMETER === undefined) {
                    throw new Error("Function module name parameter undefined, changing");
                }
                this.parameters.push({
                    name: param.PARAMETER,
                    direction: FunctionModuleParameterDirection.changing,
                    type: param.TYP || param.DBFIELD,
                    optional: param.OPTIONAL === "X",
                    defaultValue: param.DEFAULT,
                });
            }
        }
        if (data.EXPORT) {
            for (const param of (0, xml_utils_1.xmlToArray)(data.EXPORT.RSEXP)) {
                if (param.PARAMETER === undefined) {
                    throw new Error("Function module name parameter undefined, exporting");
                }
                this.parameters.push({
                    name: param.PARAMETER,
                    direction: FunctionModuleParameterDirection.exporting,
                    type: param.TYP || param.DBFIELD,
                    optional: true,
                    defaultValue: undefined,
                });
            }
        }
        if (data.TABLES) {
            for (const param of (0, xml_utils_1.xmlToArray)(data.TABLES.RSTBL)) {
                if (param.PARAMETER === undefined) {
                    throw new Error("Function module name parameter undefined, tables");
                }
                this.parameters.push({
                    name: param.PARAMETER,
                    direction: FunctionModuleParameterDirection.tables,
                    // table types are stored in TYP
                    type: param.DBSTRUCT || param.TYP,
                    optional: param.OPTIONAL === "X",
                    defaultValue: undefined,
                });
            }
        }
    }
}
exports.FunctionModuleDefinition = FunctionModuleDefinition;
