// @abaplint/core/build/src/abap/types/interface_definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterfaceDefinition = void 0;
const _identifier_1 = require("../4_file_information/_identifier");
const Structures = require("../3_structures/structures");
const Statements = require("../2_statements/statements");
const Expressions = require("../2_statements/expressions");
const class_attributes_1 = require("./class_attributes");
const visibility_1 = require("../4_file_information/visibility");
const _scope_type_1 = require("../5_syntax/_scope_type");
const event_definition_1 = require("./event_definition");
const method_definitions_1 = require("./method_definitions");
const _reference_1 = require("../5_syntax/_reference");
const _object_oriented_1 = require("../5_syntax/_object_oriented");
class InterfaceDefinition extends _identifier_1.Identifier {
    constructor(node, input) {
        if (!(node.get() instanceof Structures.Interface)) {
            throw new Error("InterfaceDefinition, unexpected node type");
        }
        const name = node.findFirstStatement(Statements.Interface).findFirstExpression(Expressions.InterfaceName).getFirstToken();
        super(name, input.filename);
        input.scope.addInterfaceDefinition(this);
        this.events = [];
        this.implementing = [];
        this.globalValue = node.findFirstExpression(Expressions.ClassGlobal) !== undefined;
        input.scope.push(_scope_type_1.ScopeType.Interface, name.getStr(), node.getFirstToken().getStart(), input.filename);
        this.parse(input, node);
        input.scope.pop(node.getLastToken().getEnd());
        // perform checks after everything has been initialized
        this.checkMethodNameLength();
    }
    getSuperClass() {
        return undefined;
    }
    getImplementing() {
        return this.implementing;
    }
    getAliases() {
        return this.aliases;
    }
    getEvents() {
        return this.events;
    }
    getAttributes() {
        return this.attributes;
    }
    getTypeDefinitions() {
        return this.typeDefinitions;
    }
    isLocal() {
        return !this.globalValue;
    }
    isGlobal() {
        return this.globalValue;
    }
    getMethodDefinitions() {
        return this.methodDefinitions;
    }
    /////////////////
    checkMethodNameLength() {
        for (const m of this.methodDefinitions.getAll()) {
            if (m.getName().length > 30) {
                const message = `Method name "${m.getName()}" is too long, maximum length is 30 characters`;
                throw new Error(message);
            }
        }
    }
    checkInterfacesExists(input, node) {
        var _a;
        for (const i of node.findAllStatements(Statements.InterfaceDef)) {
            const token = (_a = i.findDirectExpression(Expressions.InterfaceName)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
            const name = token === null || token === void 0 ? void 0 : token.getStr();
            if (name) {
                this.implementing.push({ name, partial: false });
                const idef = input.scope.findInterfaceDefinition(name);
                if (idef) {
                    input.scope.addReference(token, idef, _reference_1.ReferenceType.ObjectOrientedReference, this.filename, { ooName: name.toUpperCase(), ooType: "INTF" });
                }
                else if (input.scope.getDDIC().inErrorNamespace(name) === false) {
                    input.scope.addReference(token, undefined, _reference_1.ReferenceType.ObjectOrientedVoidReference, this.filename);
                }
                else {
                    throw new Error("Interface " + name + " unknown");
                }
            }
        }
    }
    parse(input, node) {
        this.checkInterfacesExists(input, node);
        const helper = new _object_oriented_1.ObjectOriented(input.scope);
        helper.fromInterfaces(this);
        // todo, proper sequencing, the statements should be processed line by line
        this.attributes = new class_attributes_1.Attributes(node, input);
        this.typeDefinitions = this.attributes.getTypes();
        this.aliases = this.attributes.getAliases();
        const events = node.findAllStatements(Statements.Events);
        for (const e of events) {
            this.events.push(new event_definition_1.EventDefinition(e, visibility_1.Visibility.Public, input));
        }
        this.methodDefinitions = new method_definitions_1.MethodDefinitions(node, input);
        if (this.methodDefinitions.getByName("CONSTRUCTOR") !== undefined) {
            throw new Error("Interfaces cannot have constructor methods");
        }
    }
}
exports.InterfaceDefinition = InterfaceDefinition;
