// @abaplint/core/build/src/abap/types/method_parameters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodParameters = void 0;
const method_def_1 = require("../2_statements/statements/method_def");
const Expressions = require("../2_statements/expressions");
const nodes_1 = require("../nodes");
const _typed_identifier_1 = require("./_typed_identifier");
const basic_1 = require("./basic");
const method_def_returning_1 = require("../5_syntax/expressions/method_def_returning");
const method_param_1 = require("../5_syntax/expressions/method_param");
const _object_oriented_1 = require("../5_syntax/_object_oriented");
const _reference_1 = require("../5_syntax/_reference");
const identifier_1 = require("../1_lexer/tokens/identifier");
const _scope_type_1 = require("../5_syntax/_scope_type");
// todo:
// this.exceptions = [];
// also consider RAISING vs EXCEPTIONS
class MethodParameters {
    constructor(node, input, abstractMethod) {
        if (!(node.get() instanceof method_def_1.MethodDef)) {
            throw new Error("MethodDefinition, expected MethodDef as part of input node");
        }
        this.importing = [];
        this.exporting = [];
        this.changing = [];
        this.optional = [];
        this.defaults = {};
        this.returning = undefined;
        this.preferred = undefined;
        this.exceptions = [];
        this.filename = input.filename;
        // need the scope for LIKE typing inside method parameters
        const parentName = input.scope.getName();
        input.scope.push(_scope_type_1.ScopeType.MethodDefinition, "method definition", node.getStart(), input.filename);
        this.parse(node, input, parentName, abstractMethod);
        input.scope.pop(node.getEnd());
    }
    getFilename() {
        return this.filename;
    }
    getOptional() {
        return this.optional;
    }
    getAll() {
        const ret = [];
        const returning = this.getReturning();
        if (returning) {
            ret.push(returning);
        }
        ret.push(...this.getImporting());
        ret.push(...this.getExporting());
        ret.push(...this.getChanging());
        return ret;
    }
    getDefaultImporting() {
        if (this.importing.length === 0) {
            return undefined;
        }
        else if (this.importing.length === 1) {
            return this.importing[0].getName().toUpperCase();
        }
        else if (this.preferred) {
            return this.preferred;
        }
        let candidates = this.importing.map(i => i.getName().toUpperCase());
        candidates = candidates.filter(c => this.optional.indexOf(c) < 0);
        if (candidates.length === 1) {
            return candidates[0];
        }
        return undefined;
    }
    getImporting() {
        return this.importing;
    }
    getRequiredParameters() {
        var _a;
        const ret = [];
        for (const i of this.getImporting()) {
            if (this.getOptional().some(o => o.toUpperCase() === i.getName().toUpperCase()) === true) {
                continue;
            }
            else if (((_a = this.preferred) === null || _a === void 0 ? void 0 : _a.toUpperCase()) === i.getName().toUpperCase()) {
                continue;
            }
            ret.push(i);
        }
        for (const i of this.getChanging()) {
            if (this.getOptional().some(o => o.toUpperCase() === i.getName().toUpperCase()) === true) {
                continue;
            }
            ret.push(i);
        }
        return ret;
    }
    getExporting() {
        return this.exporting;
    }
    getChanging() {
        return this.changing;
    }
    getReturning() {
        return this.returning;
    }
    getExceptions() {
        return this.exceptions;
    }
    getParameterDefault(parameter) {
        return this.defaults[parameter.toUpperCase()];
    }
    ///////////////////
    parse(node, input, parentName, abstractMethod) {
        var _a, _b;
        const handler = node.findFirstExpression(Expressions.EventHandler);
        if (handler) {
            const nameToken = (_a = node.findFirstExpression(Expressions.ClassName)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
            const ooName = nameToken === null || nameToken === void 0 ? void 0 : nameToken.getStr();
            const def = input.scope.findObjectDefinition(ooName);
            const doVoid = def ? false : !input.scope.getDDIC().inErrorNamespace(ooName);
            if (def) {
                input.scope.addReference(nameToken, def, _reference_1.ReferenceType.ObjectOrientedReference, input.filename);
            }
            else if (doVoid && ooName) {
                input.scope.addReference(nameToken, undefined, _reference_1.ReferenceType.ObjectOrientedVoidReference, this.filename, { ooName: ooName.toUpperCase() });
            }
            const eventName = (_b = node.findFirstExpression(Expressions.EventName)) === null || _b === void 0 ? void 0 : _b.getFirstToken().getStr();
            const event = new _object_oriented_1.ObjectOriented(input.scope).searchEvent(def, eventName);
            for (const p of handler.findAllExpressions(Expressions.MethodParamName)) {
                const token = p.getFirstToken();
                const search = token.getStr().toUpperCase().replace("!", "");
                this.optional.push(search); // all parameters optional for event handlers
                if (search === "SENDER" && def) {
                    this.importing.push(new _typed_identifier_1.TypedIdentifier(token, this.filename, new basic_1.ObjectReferenceType(def), ["event_parameter" /* IdentifierMeta.EventParameter */]));
                    continue;
                }
                const found = event === null || event === void 0 ? void 0 : event.getParameters().find(p => p.getName().toUpperCase() === search);
                if (found) {
                    this.importing.push(new _typed_identifier_1.TypedIdentifier(token, this.filename, found.getType(), ["event_parameter" /* IdentifierMeta.EventParameter */]));
                }
                else if (doVoid) {
                    this.importing.push(new _typed_identifier_1.TypedIdentifier(token, this.filename, basic_1.VoidType.get(ooName), ["event_parameter" /* IdentifierMeta.EventParameter */]));
                }
                else {
                    const type = new basic_1.UnknownType(`handler parameter not found "${search}"`);
                    this.importing.push(new _typed_identifier_1.TypedIdentifier(token, this.filename, type, ["event_parameter" /* IdentifierMeta.EventParameter */]));
                }
            }
            return;
        }
        const importing = node.findFirstExpression(Expressions.MethodDefImporting);
        if (importing) {
            this.add(this.importing, importing, input, ["importing" /* IdentifierMeta.MethodImporting */], abstractMethod);
            if (importing.findDirectTokenByText("PREFERRED")) {
                this.preferred = importing.getLastToken().getStr().toUpperCase();
                if (this.preferred.startsWith("!")) {
                    this.preferred = this.preferred.substring(1);
                }
            }
        }
        const exporting = node.findFirstExpression(Expressions.MethodDefExporting);
        if (exporting) {
            this.add(this.exporting, exporting, input, ["exporting" /* IdentifierMeta.MethodExporting */], abstractMethod);
        }
        const changing = node.findFirstExpression(Expressions.MethodDefChanging);
        if (changing) {
            this.add(this.changing, changing, input, ["changing" /* IdentifierMeta.MethodChanging */], abstractMethod);
        }
        const returning = node.findFirstExpression(Expressions.MethodDefReturning);
        if (returning) {
            this.returning = method_def_returning_1.MethodDefReturning.runSyntax(returning, input, ["returning" /* IdentifierMeta.MethodReturning */]);
        }
        this.workaroundRAP(node, input, parentName);
    }
    workaroundRAP(node, input, parentName) {
        const resultName = node.findExpressionAfterToken("RESULT");
        const isRap = node.findExpressionAfterToken("IMPORTING");
        if (isRap) {
            for (const foo of node.findDirectExpressions(Expressions.MethodParamName)) {
                if (foo === resultName) {
                    continue;
                }
                this.importing.push(new _typed_identifier_1.TypedIdentifier(foo.getFirstToken(), input.filename, basic_1.VoidType.get("RapMethodParameter"), ["importing" /* IdentifierMeta.MethodImporting */]));
            }
            const concat = node.concatTokens().toUpperCase();
            if (concat.includes(" FOR VALIDATE ")
                || concat.includes(" FOR BEHAVIOR ")
                || concat.includes(" FOR DETERMINE ")
                || concat.includes(" FOR FEATURES ")
                || concat.includes(" FOR INSTANCE FEATURES ")
                || concat.includes(" FOR LOCK ")
                || concat.includes(" FOR NUMBERING ")
                || concat.includes(" FOR READ ")
                || concat.includes(" FOR VALIDATION ")
                || concat.includes(" FOR MODIFY ")) {
                const token = isRap.getFirstToken();
                this.exporting.push(new _typed_identifier_1.TypedIdentifier(new identifier_1.Identifier(token.getStart(), "failed"), input.filename, basic_1.VoidType.get("RapMethodParameter"), ["exporting" /* IdentifierMeta.MethodExporting */]));
                this.exporting.push(new _typed_identifier_1.TypedIdentifier(new identifier_1.Identifier(token.getStart(), "mapped"), input.filename, basic_1.VoidType.get("RapMethodParameter"), ["exporting" /* IdentifierMeta.MethodExporting */]));
                this.exporting.push(new _typed_identifier_1.TypedIdentifier(new identifier_1.Identifier(token.getStart(), "reported"), input.filename, basic_1.VoidType.get("RapMethodParameter"), ["exporting" /* IdentifierMeta.MethodExporting */]));
            }
        }
        if (resultName) {
            const token = resultName.getFirstToken();
            this.importing.push(new _typed_identifier_1.TypedIdentifier(token, input.filename, basic_1.VoidType.get("RapMethodParameter"), ["exporting" /* IdentifierMeta.MethodExporting */]));
        }
        // its some kind of magic
        if (parentName.toUpperCase() === "CL_ABAP_BEHAVIOR_SAVER") {
            const tempChanging = this.changing.map(c => new _typed_identifier_1.TypedIdentifier(c.getToken(), input.filename, basic_1.VoidType.get("RapMethodParameter"), c.getMeta()));
            while (this.changing.length > 0) {
                this.changing.shift();
            }
            this.changing.push(...tempChanging);
            const tempImporting = this.importing.map(c => new _typed_identifier_1.TypedIdentifier(c.getToken(), input.filename, basic_1.VoidType.get("RapMethodParameter"), c.getMeta()));
            while (this.importing.length > 0) {
                this.importing.shift();
            }
            this.importing.push(...tempImporting);
        }
    }
    add(target, source, input, meta, abstractMethod) {
        var _a, _b;
        for (const opt of source.findAllExpressions(Expressions.MethodParamOptional)) {
            const p = opt.findDirectExpression(Expressions.MethodParam);
            if (p === undefined) {
                continue;
            }
            const extraMeta = [];
            if (p.getFirstToken().getStr().toUpperCase() === "VALUE" && ((_a = p.getChildren()[1]) === null || _a === void 0 ? void 0 : _a.getFirstToken().getStr()) === "(") {
                extraMeta.push("pass_by_value" /* IdentifierMeta.PassByValue */);
            }
            else if (meta.includes("importing" /* IdentifierMeta.MethodImporting */)) {
                extraMeta.push("read_only" /* IdentifierMeta.ReadOnly */);
            }
            if (abstractMethod === true) {
                extraMeta.push("abstract" /* IdentifierMeta.Abstract */);
            }
            const id = method_param_1.MethodParam.runSyntax(p, input, [...meta, ...extraMeta]);
            input.scope.addIdentifier(id);
            target.push(id);
            if (opt.findDirectTokenByText("OPTIONAL")) {
                const name = target[target.length - 1].getName().toUpperCase();
                this.optional.push(name);
            }
            else if (opt.findFirstExpression(Expressions.Default)) {
                const name = target[target.length - 1].getName().toUpperCase();
                this.optional.push(name);
                const val = (_b = opt.findFirstExpression(Expressions.Default)) === null || _b === void 0 ? void 0 : _b.getLastChild();
                if (val && val instanceof nodes_1.ExpressionNode) {
                    this.defaults[name] = val;
                }
            }
        }
        if (target.length > 0) {
            return;
        }
        const params = source.findAllExpressions(Expressions.MethodParam);
        for (const param of params) {
            target.push(method_param_1.MethodParam.runSyntax(param, input, meta));
        }
    }
}
exports.MethodParameters = MethodParameters;
