// @abaplint/core/build/src/abap/types/method_definition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodDefinition = void 0;
const statements_1 = require("../2_statements/statements");
const Expressions = require("../2_statements/expressions");
const method_parameters_1 = require("./method_parameters");
const _identifier_1 = require("../4_file_information/_identifier");
const _reference_1 = require("../5_syntax/_reference");
class MethodDefinition extends _identifier_1.Identifier {
    // todo: final flag
    constructor(node, visibility, input) {
        var _a, _b, _c, _d;
        if (!(node.get() instanceof statements_1.MethodDef)) {
            throw new Error("MethodDefinition, expected MethodDef as part of input node");
        }
        const found = node.findDirectExpression(Expressions.MethodName);
        if (found === undefined) {
            throw new Error("MethodDefinition, expected MethodDef as part of input node");
        }
        super(found.getFirstToken(), input.filename);
        this.className = input.scope.getName();
        this.redefinition = false;
        if (node.findDirectExpression(Expressions.Redefinition)) {
            this.redefinition = true;
            const name = found.getFirstToken().getStr();
            if (name.includes("~")) {
                const idef = input.scope.findInterfaceDefinition(name.split("~")[0]);
                if (idef) {
                    input.scope.addReference(found.getFirstToken(), idef, _reference_1.ReferenceType.ObjectOrientedReference, input.filename);
                }
            }
        }
        this.eventHandler = false;
        const handler = node.findDirectExpression(Expressions.EventHandler);
        if (handler) {
            this.eventHandler = true;
            this.eventName = (_a = handler.findDirectExpression(Expressions.EventName)) === null || _a === void 0 ? void 0 : _a.concatTokens().toUpperCase();
            this.eventClass = (_b = handler.findDirectExpression(Expressions.ClassName)) === null || _b === void 0 ? void 0 : _b.concatTokens().toUpperCase();
        }
        this.abstract = false;
        if (node.findDirectExpression(Expressions.Abstract)) {
            this.abstract = true;
        }
        this.static = false;
        // checks for "CLASS-METHODS"
        if (node.getFirstToken().getStr().toUpperCase().startsWith("CLASS")) {
            this.static = true;
        }
        this.raising = [];
        for (const r of ((_c = node.findDirectExpression(Expressions.MethodDefRaising)) === null || _c === void 0 ? void 0 : _c.findAllExpressions(Expressions.ClassName)) || []) {
            const token = r.getFirstToken();
            const name = token.getStr();
            this.raising.push(name);
            const clas = input.scope.findClassDefinition(name);
            if (clas) {
                input.scope.addReference(token, clas, _reference_1.ReferenceType.ObjectOrientedReference, input.filename, { ooName: name.toUpperCase(), ooType: "CLAS" });
            }
            else if (input.scope.getDDIC().inErrorNamespace(name) === false) {
                input.scope.addReference(token, clas, _reference_1.ReferenceType.ObjectOrientedVoidReference, input.filename, { ooName: name.toUpperCase(), ooType: "CLAS" });
            }
            else {
                input.scope.addReference(token, clas, _reference_1.ReferenceType.ObjectOrientedUnknownReference, input.filename, { ooName: name.toUpperCase(), ooType: "CLAS" });
            }
        }
        this.exceptions = [];
        for (const r of ((_d = node.findDirectExpression(Expressions.MethodDefExceptions)) === null || _d === void 0 ? void 0 : _d.findAllExpressions(Expressions.NamespaceSimpleName)) || []) {
            const token = r.getFirstToken();
            const name = token.getStr();
            this.exceptions.push(name);
        }
        this.visibility = visibility;
        this.parameters = new method_parameters_1.MethodParameters(node, input, this.abstract);
    }
    getVisibility() {
        return this.visibility;
    }
    getClassName() {
        return this.className;
    }
    isRedefinition() {
        return this.redefinition;
    }
    isAbstract() {
        return this.abstract;
    }
    isStatic() {
        return this.static;
    }
    isEventHandler() {
        return this.eventHandler;
    }
    getEventName() {
        return this.eventName;
    }
    getEventClass() {
        return this.eventClass;
    }
    getParameters() {
        return this.parameters;
    }
    getRaising() {
        return this.raising;
    }
    getExceptions() {
        return this.exceptions;
    }
}
exports.MethodDefinition = MethodDefinition;
