// @abaplint/runtime/build/src/types/abap_object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ABAPObject = void 0;
const field_symbol_1 = require("./field_symbol");
class ABAPObject {
    value;
    qualifiedName;
    RTTIName;
    constructor(input) {
        this.qualifiedName = input?.qualifiedName;
        this.RTTIName = input?.RTTIName;
        this.clear();
    }
    clone() {
        const n = new ABAPObject({ qualifiedName: this.qualifiedName, RTTIName: this.RTTIName });
        n.value = this.value;
        return n;
    }
    get() {
        return this.value;
    }
    getInternalID() {
        return "OpenABAPInternalObjectId=" + this.value?.INTERNAL_ID;
    }
    clear() {
        this.value = undefined;
    }
    getQualifiedName() {
        return this.qualifiedName;
    }
    getRTTIName() {
        return this.RTTIName;
    }
    set(value) {
        if (value instanceof ABAPObject) {
            this.value = value.get();
        }
        else if (value instanceof field_symbol_1.FieldSymbol) {
            this.value = value.getPointer().get();
        }
        else {
            this.value = value;
        }
        return this;
    }
}
exports.ABAPObject = ABAPObject;
