// @abaplint/core/build/src/abap/5_syntax/global_definitions/find_global_definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FindGlobalDefinitions = void 0;
const interface_definition_1 = require("../../types/interface_definition");
const class_definition_1 = require("../../types/class_definition");
const _current_scope_1 = require("../_current_scope");
const Structures = require("../../3_structures/structures");
const interface_1 = require("../../../objects/interface");
const class_1 = require("../../../objects/class");
const BasicTypes = require("../../types/basic");
const objects_1 = require("../../../objects");
// todo: rewrite all of this to use a graph based deterministic approach instead
// this makes sure to cache global interface and class definitions in the corresponding object
class FindGlobalDefinitions {
    constructor(reg) {
        this.reg = reg;
    }
    run(progress) {
        const MAX_PASSES = 10;
        let lastPass = Number.MAX_SAFE_INTEGER;
        // the setDirty method in the objects clears the definitions
        let candidates = [];
        for (const o of this.reg.getObjects()) {
            if ((o instanceof interface_1.Interface || o instanceof class_1.Class) && o.getDefinition() === undefined) {
                candidates.push(o);
            }
            else if (o instanceof objects_1.DataElement
                || o instanceof objects_1.Table
                || o instanceof objects_1.TableType
                || o instanceof objects_1.View
                || o instanceof objects_1.LockObject
                || o instanceof objects_1.AuthorizationCheckField) {
                o.parseType(this.reg); // make sure the references are set after parsing finishes
            }
        }
        // make sure the sequence is always the same, disregarding the sequence they were added to the registry
        // this will hopefully make it easier to debug
        candidates.sort((a, b) => { return a.getName().localeCompare(b.getName()); });
        for (let i = 1; i <= MAX_PASSES; i++) {
            progress === null || progress === void 0 ? void 0 : progress.set(candidates.length, "Global OO types, pass " + i);
            let thisPass = 0;
            const next = [];
            for (const o of candidates) {
                progress === null || progress === void 0 ? void 0 : progress.tickSync("Global OO types(pass " + i + "), next pass: " + next.length);
                this.update(o);
                const untypedCount = this.countUntyped(o);
                if (untypedCount > 0) {
                    next.push(o);
                }
                thisPass = thisPass + untypedCount;
            }
            candidates = next;
            if (lastPass === thisPass || thisPass === 0) {
                break;
            }
            lastPass = thisPass;
        }
    }
    /////////////////////////////
    countUntyped(obj) {
        const def = obj.getDefinition();
        if (def === undefined) {
            return 1;
        }
        let count = 0;
        for (const t of def.getTypeDefinitions().getAll()) {
            count = count + this.count(t.type.getType());
        }
        for (const a of def.getAttributes().getAll()) {
            count = count + this.count(a.getType());
        }
        for (const a of def.getAttributes().getConstants()) {
            count = count + this.count(a.getType());
        }
        for (const m of def.getMethodDefinitions().getAll()) {
            for (const p of m.getParameters().getAll()) {
                count = count + this.count(p.getType());
            }
        }
        for (const e of def.getEvents() || []) {
            for (const p of e.getParameters()) {
                count = count + this.count(p.getType());
            }
        }
        return count;
    }
    count(type) {
        if (type instanceof BasicTypes.UnknownType || type instanceof BasicTypes.VoidType) {
            return 1;
        }
        else if (type instanceof BasicTypes.TableType) {
            return this.count(type.getRowType());
        }
        else if (type instanceof BasicTypes.DataReference) {
            return this.count(type.getType());
        }
        else if (type instanceof BasicTypes.StructureType) {
            let count = 0;
            for (const c of type.getComponents()) {
                count = count + this.count(c.type);
            }
            return count;
        }
        return 0;
    }
    update(obj) {
        const file = obj.getMainABAPFile();
        if (file === undefined) {
            obj.setDefinition(undefined);
            return;
        }
        const struc = file === null || file === void 0 ? void 0 : file.getStructure();
        if (struc === undefined) {
            obj.setDefinition(undefined);
            return;
        }
        const input = {
            filename: file.getFilename(),
            scope: _current_scope_1.CurrentScope.buildDefault(this.reg, obj),
            issues: [],
        };
        if (obj instanceof interface_1.Interface) {
            const found = struc.findFirstStructure(Structures.Interface);
            if (found) {
                try {
                    const def = new interface_definition_1.InterfaceDefinition(found, input);
                    obj.setDefinition(def);
                }
                catch (_a) {
                    obj.setDefinition(undefined);
                }
            }
            else {
                obj.setDefinition(undefined);
            }
        }
        else {
            const found = struc.findFirstStructure(Structures.ClassDefinition);
            if (found) {
                try {
                    const def = new class_definition_1.ClassDefinition(found, input);
                    obj.setDefinition(def);
                }
                catch (_b) {
                    obj.setDefinition(undefined);
                }
            }
            else {
                obj.setDefinition(undefined);
            }
        }
    }
}
exports.FindGlobalDefinitions = FindGlobalDefinitions;
