// @abaplint/core/build/src/rules/align_type_expressions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlignTypeExpressions = exports.AlignTypeExpressionsConf = void 0;
const _abap_rule_1 = require("./_abap_rule");
const _basic_rule_config_1 = require("./_basic_rule_config");
const objects_1 = require("../objects");
const ddic_1 = require("../ddic");
const _irule_1 = require("./_irule");
const issue_1 = require("../issue");
const position_1 = require("../position");
const Expressions = require("../abap/2_statements/expressions");
const Statements = require("../abap/2_statements/statements");
const Structures = require("../abap/3_structures/structures");
const edit_helper_1 = require("../edit_helper");
class AlignTypeExpressionsConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** Ignore global exception classes */
        this.ignoreExceptions = true;
    }
}
exports.AlignTypeExpressionsConf = AlignTypeExpressionsConf;
class AlignTypeExpressions extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new AlignTypeExpressionsConf();
    }
    getMetadata() {
        return {
            key: "align_type_expressions",
            title: "Align TYPE expressions",
            shortDescription: `Align TYPE expressions in statements`,
            extendedInformation: `
Currently works for METHODS + BEGIN OF

If BEGIN OF has an INCLUDE TYPE its ignored

Also note that clean ABAP does not recommend aligning TYPE clauses:
https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#dont-align-type-clauses`,
            tags: [_irule_1.RuleTag.SingleFile, _irule_1.RuleTag.Whitespace, _irule_1.RuleTag.Quickfix],
            badExample: `
TYPES: BEGIN OF foo,
         bar TYPE i,
         foobar TYPE i,
       END OF foo.

INTERFACE lif.
  METHODS bar
    IMPORTING
      foo TYPE i
      foobar TYPE i.
ENDINTERFACE.`,
            goodExample: `
TYPES: BEGIN OF foo,
         bar    TYPE i,
         foobar TYPE i,
       END OF foo.

INTERFACE lif.
  METHODS bar
    IMPORTING
      foo    TYPE i
      foobar TYPE i.
ENDINTERFACE.`,
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file, obj) {
        const issues = [];
        const stru = file.getStructure();
        if (stru === undefined) {
            return issues; // parser error
        }
        const ddic = new ddic_1.DDIC(this.reg);
        if (obj instanceof objects_1.Class) {
            const definition = obj.getClassDefinition();
            if (definition === undefined) {
                return [];
            }
            else if (this.conf.ignoreExceptions && ddic.isException(definition, obj)) {
                return [];
            }
        }
        issues.push(...this.checkTypes(stru, file));
        issues.push(...this.checkMethods(stru, file));
        issues.push(...this.checkEvents(stru, file));
        return issues;
    }
    check(fields, column, file) {
        const issues = [];
        const rows = new Set();
        for (const f of fields) {
            const row = f.after.getRow();
            if (rows.has(row)) {
                return [];
            }
            rows.add(row);
        }
        for (const f of fields) {
            if (f.after.getCol() === column) {
                continue;
            }
            let fix = undefined;
            if (f.after.getCol() < column) {
                fix = edit_helper_1.EditHelper.insertAt(file, f.after, " ".repeat(column - f.after.getCol()));
            }
            else {
                fix = edit_helper_1.EditHelper.deleteRange(file, new position_1.Position(f.after.getRow(), column), f.after);
            }
            const message = `Align TYPE expressions to column ${column}`;
            const issue = issue_1.Issue.atPosition(file, f.after, message, this.getMetadata().key, this.conf.severity, fix);
            issues.push(issue);
        }
        return issues;
    }
    checkMethods(stru, file) {
        const issues = [];
        const methods = stru.findAllStatements(Statements.MethodDef);
        for (const m of methods) {
            const fields = [];
            const params = m.findAllExpressions(Expressions.MethodParam);
            let column = 0;
            for (const p of params) {
                const children = p.getChildren();
                const name = children[children.length - 2];
                fields.push({
                    nameEnd: name.getLastToken().getEnd(),
                    after: p.findFirstExpression(Expressions.TypeParam).getFirstToken().getStart()
                });
                column = Math.max(column, name.getFirstToken().getEnd().getCol() + 1);
            }
            const ret = m.findFirstExpression(Expressions.MethodDefReturning);
            if (ret) {
                const children = ret.getChildren();
                const name = children[children.length - 2];
                fields.push({
                    nameEnd: name.getLastToken().getEnd(),
                    after: ret.findFirstExpression(Expressions.TypeParam).getFirstToken().getStart()
                });
                column = Math.max(column, name.getLastToken().getEnd().getCol() + 1);
            }
            issues.push(...this.check(fields, column, file));
        }
        return issues;
    }
    checkEvents(stru, file) {
        const issues = [];
        const events = stru.findAllStatements(Statements.Events);
        for (const e of events) {
            const fields = [];
            const params = e.findAllExpressions(Expressions.MethodParam);
            let column = 0;
            for (const p of params) {
                const children = p.getChildren();
                const name = children[children.length - 2];
                fields.push({
                    nameEnd: name.getLastToken().getEnd(),
                    after: p.findFirstExpression(Expressions.TypeParam).getFirstToken().getStart()
                });
                column = Math.max(column, name.getFirstToken().getEnd().getCol() + 1);
            }
            issues.push(...this.check(fields, column, file));
        }
        return issues;
    }
    checkTypes(stru, file) {
        const issues = [];
        const types = stru.findAllStructuresRecursive(Structures.Types);
        for (const t of types) {
            if (t.findDirectStatement(Statements.IncludeType)) {
                continue;
            }
            const fields = [];
            let column = 0;
            const st = t.findDirectStatements(Statements.Type);
            for (const s of st) {
                const name = s.getChildren()[1];
                fields.push({
                    nameEnd: name.getLastToken().getEnd(),
                    after: s.getChildren()[2].getFirstToken().getStart()
                });
                column = Math.max(column, name.getFirstToken().getEnd().getCol() + 1);
            }
            issues.push(...this.check(fields, column, file));
        }
        return issues;
    }
}
exports.AlignTypeExpressions = AlignTypeExpressions;
