// @abaplint/core/build/src/rules/allowed_object_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllowedObjectTypes = exports.AllowedObjectTypesConf = void 0;
const _irule_1 = require("./_irule");
const issue_1 = require("../issue");
const _basic_rule_config_1 = require("./_basic_rule_config");
const position_1 = require("../position");
class AllowedObjectTypesConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** List of allowed object types, example: ["CLAS", "INTF"]
         * @uniqueItems true
        */
        this.allowed = [];
    }
}
exports.AllowedObjectTypesConf = AllowedObjectTypesConf;
class AllowedObjectTypes {
    constructor() {
        this.conf = new AllowedObjectTypesConf();
    }
    getMetadata() {
        return {
            key: "allowed_object_types",
            title: "Check allowed object types",
            shortDescription: `Restricts the set of allowed object types.`,
            extendedInformation: `"allowed" is a list of 4 character R3TR object types, example: ["CLAS", "INTF"]`,
            tags: [_irule_1.RuleTag.SingleFile],
        };
    }
    initialize(_reg) {
        return this;
    }
    getDescription(objectType) {
        return "Object type " + objectType + " not allowed";
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    run(obj) {
        const allowed = this.getConfig().allowed;
        if (allowed === undefined || allowed.length === 0) {
            return [];
        }
        const objectType = obj.getType();
        if (allowed.indexOf(objectType) < 0) {
            const position = new position_1.Position(1, 1);
            const issue = issue_1.Issue.atPosition(obj.getFiles()[0], position, this.getDescription(objectType), this.getMetadata().key, this.conf.severity);
            return [issue];
        }
        return [];
    }
}
exports.AllowedObjectTypes = AllowedObjectTypes;
