// @abaplint/core/build/src/rules/cloud_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudTypes = exports.CloudTypesConf = void 0;
const _irule_1 = require("./_irule");
const issue_1 = require("../issue");
const version_1 = require("../version");
const Objects = require("../objects");
const _basic_rule_config_1 = require("./_basic_rule_config");
const position_1 = require("../position");
class CloudTypesConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.CloudTypesConf = CloudTypesConf;
class CloudTypes {
    constructor() {
        this.conf = new CloudTypesConf();
    }
    getMetadata() {
        return {
            key: "cloud_types",
            title: "Check cloud types",
            shortDescription: `Checks that the package does not contain any object types unsupported in cloud ABAP.`,
            tags: [_irule_1.RuleTag.SingleFile, _irule_1.RuleTag.Syntax],
        };
    }
    getDescription(objectType) {
        return "Object type " + objectType + " not supported in cloud";
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    initialize(reg) {
        this.reg = reg;
        return this;
    }
    static isCloud(obj) {
        return obj instanceof Objects.ApplicationJobCatalogEntry
            || obj instanceof Objects.ApplicationJobTemplate
            || obj instanceof Objects.APIReleaseState
            || obj instanceof Objects.AssignmentServiceToAuthorizationGroup
            || obj instanceof Objects.ATCCheckCategory
            || obj instanceof Objects.ApplicationDescriptorsFiori
            || obj instanceof Objects.ATCCheckObject
            || obj instanceof Objects.ATCCheckVariant
            || obj instanceof Objects.AuthorizationCheckField
            || obj instanceof Objects.AuthorizationObject
            || obj instanceof Objects.AuthorizationObjectExtension
            || obj instanceof Objects.BehaviorDefinition
            || obj instanceof Objects.BusinessCatalog
            || obj instanceof Objects.BusinessCatalogAppAssignment
            || obj instanceof Objects.CDSMetadataExtension
            || obj instanceof Objects.Class
            || obj instanceof Objects.CDSType
            || obj instanceof Objects.ChangeDocument
            || obj instanceof Objects.CDSEntityBuffer
            || obj instanceof Objects.ApplicationLogObject
            || obj instanceof Objects.CommunicationScenario
            || obj instanceof Objects.DataControl
            || obj instanceof Objects.Namespace
            || obj instanceof Objects.KnowledgeTransferDocument
            || obj instanceof Objects.DataDefinition
            || obj instanceof Objects.DataElement
            || obj instanceof Objects.Domain
            || obj instanceof Objects.EventBinding
            || obj instanceof Objects.EventConsumer
            || obj instanceof Objects.FunctionGroup
            || obj instanceof Objects.HttpService
            || obj instanceof Objects.IAMApp
            || obj instanceof Objects.InboundService
            || obj instanceof Objects.Interface
            || obj instanceof Objects.LockObject
            || obj instanceof Objects.MessageClass
            || obj instanceof Objects.NumberRange
            || obj instanceof Objects.OutboundService
            || obj instanceof Objects.Package
            || obj instanceof Objects.RestrictionField
            || obj instanceof Objects.RestrictionType
            || obj instanceof Objects.ServiceBinding
            || obj instanceof Objects.ServiceDefinition
            || obj instanceof Objects.Table
            || obj instanceof Objects.TableType
            || obj instanceof Objects.Transformation;
    }
    run(obj) {
        if (this.reg.getConfig().getVersion() !== version_1.Version.Cloud || CloudTypes.isCloud(obj)) {
            return [];
        }
        const position = new position_1.Position(1, 1);
        const issue = issue_1.Issue.atPosition(obj.getFiles()[0], position, this.getDescription(obj.getType()), this.getMetadata().key, this.conf.severity);
        return [issue];
    }
}
exports.CloudTypes = CloudTypes;
