// @abaplint/core/build/src/rules/downport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Downport = exports.DownportConf = void 0;
/* eslint-disable max-len */
const crypto = require("crypto");
const tokens_1 = require("../abap/1_lexer/tokens");
const Expressions = require("../abap/2_statements/expressions");
const Statements = require("../abap/2_statements/statements");
const statements_1 = require("../abap/2_statements/statements");
const _statement_1 = require("../abap/2_statements/statements/_statement");
const Structures = require("../abap/3_structures/structures");
const _builtin_1 = require("../abap/5_syntax/_builtin");
const _reference_1 = require("../abap/5_syntax/_reference");
const _scope_type_1 = require("../abap/5_syntax/_scope_type");
const syntax_1 = require("../abap/5_syntax/syntax");
const nodes_1 = require("../abap/nodes");
const _typed_identifier_1 = require("../abap/types/_typed_identifier");
const basic_1 = require("../abap/types/basic");
const config_1 = require("../config");
const edit_helper_1 = require("../edit_helper");
const issue_1 = require("../issue");
const objects_1 = require("../objects");
const _abap_object_1 = require("../objects/_abap_object");
const position_1 = require("../position");
const registry_1 = require("../registry");
const include_graph_1 = require("../utils/include_graph");
const version_1 = require("../version");
const virtual_position_1 = require("../virtual_position");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _irule_1 = require("./_irule");
// todo: refactor each sub-rule to new classes?
// todo: add configuration
class DownportConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.DownportConf = DownportConf;
class SkipToNextFile extends Error {
    constructor(issue) {
        super();
        this.issue = issue;
    }
}
class SpagHelper {
    constructor(spag) {
        this.spag = spag;
    }
    renameVariable(oldName, pos, lowFile, newName) {
        let fix = undefined;
        const references = this.findReferences(oldName, pos);
        references.sort((a, b) => {
            if (a.start.equals(b.start)) {
                return 0;
            }
            return a.start.isAfter(b.start) ? 1 : -1;
        });
        for (const r of references) {
            const replace = edit_helper_1.EditHelper.replaceRange(lowFile, r.start, r.end, newName);
            if (fix === undefined) {
                fix = replace;
            }
            else {
                fix = edit_helper_1.EditHelper.merge(replace, fix);
            }
        }
        return fix;
    }
    findReferences(name, pos) {
        var _a, _b;
        const positions = [];
        function has(element) {
            return positions.some(a => a.start.equals(element.start));
        }
        for (const r of this.spag.getData().references) {
            if (((_a = r.resolved) === null || _a === void 0 ? void 0 : _a.getName()) === name && ((_b = r.resolved) === null || _b === void 0 ? void 0 : _b.getStart().equals(pos))) {
                const sub = {
                    start: r.position.getStart(),
                    end: r.position.getEnd(),
                };
                if (has(sub) === false) {
                    positions.push(sub);
                }
            }
        }
        for (const child of this.spag.getChildren()) {
            const subPositions = new SpagHelper(child).findReferences(name, pos);
            for (const sub of subPositions) {
                if (has(sub) === false) {
                    positions.push(sub);
                }
            }
        }
        return positions;
    }
    findRecursiveDuplicate(name, skip) {
        var _a;
        const found = this.spag.findVariable(name);
        if ((found === null || found === void 0 ? void 0 : found.getStart().equals(skip)) === false) {
            return found;
        }
        for (const child of ((_a = this.spag) === null || _a === void 0 ? void 0 : _a.getChildren()) || []) {
            const sub = new SpagHelper(child).findRecursiveDuplicate(name, skip);
            if (sub) {
                return sub;
            }
        }
        return undefined;
    }
    isDuplicateName(name, pos) {
        let parent = this.spag.getParent();
        while ((parent === null || parent === void 0 ? void 0 : parent.getIdentifier().stype) === _scope_type_1.ScopeType.Let
            || (parent === null || parent === void 0 ? void 0 : parent.getIdentifier().stype) === _scope_type_1.ScopeType.For) {
            parent = parent.getParent();
        }
        if (parent === undefined) {
            return undefined;
        }
        return new SpagHelper(parent).findRecursiveDuplicate(name, pos) !== undefined;
    }
}
class Downport {
    constructor() {
        this.conf = new DownportConf();
    }
    getMetadata() {
        return {
            key: "downport",
            title: "Downport statement",
            shortDescription: `Downport functionality`,
            extendedInformation: `Much like the 'commented_code' rule this rule loops through unknown statements and tries parsing with
a higher level language version. If successful, various rules are applied to downport the statement.
Target downport version is always v702, thus rule is only enabled if target version is v702.

Current rules:
* NEW transformed to CREATE OBJECT, opposite of https://rules.abaplint.org/use_new/
* DATA() definitions are outlined, opposite of https://rules.abaplint.org/prefer_inline/
* FIELD-SYMBOL() definitions are outlined
* CONV is outlined
* COND is outlined
* REDUCE is outlined
* SWITCH is outlined
* FILTER is outlined
* APPEND expression is outlined
* INSERT expression is outlined
* EMPTY KEY is changed to DEFAULT KEY, opposite of DEFAULT KEY in https://rules.abaplint.org/avoid_use/
* CAST changed to ?=
* LOOP AT method_call( ) is outlined
* VALUE # with structure fields
* VALUE # with internal table lines
* Table Expressions are outlined
* SELECT INTO @DATA definitions are outlined
* Some occurrences of string template formatting option ALPHA changed to function module call
* SELECT/INSERT/MODIFY/DELETE/UPDATE "," in field list removed, "@" in source/targets removed
* PARTIALLY IMPLEMENTED removed, it can be quick fixed via rule implement_methods
* RAISE EXCEPTION ... MESSAGE
* Moving with +=, -=, /=, *=, &&= is expanded
* line_exists and line_index is downported to READ TABLE
* ENUMs, but does not nessesarily give the correct type and value
* MESSAGE with non simple source

Only one transformation is applied to a statement at a time, so multiple steps might be required to do the full downport.

Make sure to test the downported code, it might not always be completely correct.`,
            tags: [_irule_1.RuleTag.Downport, _irule_1.RuleTag.Quickfix],
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    initialize(reg) {
        this.lowReg = reg;
        const version = this.lowReg.getConfig().getVersion();
        if (version === version_1.Version.v702 || version === version_1.Version.OpenABAP) {
            this.initHighReg();
        }
        return this;
    }
    listMainForInclude(filename) {
        if (filename === undefined) {
            return [];
        }
        // only initialize this.graph if needed
        if (this.graph === undefined) {
            this.graph = new include_graph_1.IncludeGraph(this.lowReg);
        }
        return this.graph.listMainForInclude(filename);
    }
    containsError(highObj) {
        for (const file of highObj.getABAPFiles()) {
            for (const statement of file.getStatements()) {
                if (statement.get() instanceof _statement_1.Unknown) {
                    return true; // contains parser errors
                }
            }
            if (file.getStructure() === undefined) {
                return true;
            }
        }
        return false;
    }
    run(lowObj) {
        var _a;
        const ret = [];
        this.counter = 1;
        const version = this.lowReg.getConfig().getVersion();
        if (version !== version_1.Version.v702 && version !== version_1.Version.OpenABAP) {
            return ret;
        }
        else if (!(lowObj instanceof _abap_object_1.ABAPObject)) {
            return ret;
        }
        const highObj = this.highReg.getObject(lowObj.getType(), lowObj.getName());
        if (highObj === undefined || !(highObj instanceof _abap_object_1.ABAPObject)) {
            return ret;
        }
        let highSyntaxObj = highObj;
        if (this.containsError(highObj)) {
            return ret;
        }
        // for includes do the syntax check via a main program
        if (lowObj instanceof objects_1.Program && lowObj.isInclude()) {
            const mains = this.listMainForInclude((_a = lowObj.getMainABAPFile()) === null || _a === void 0 ? void 0 : _a.getFilename());
            if (mains.length <= 0) {
                return [];
            }
            const f = this.highReg.getFileByName(mains[0]);
            if (f === undefined) {
                return [];
            }
            highSyntaxObj = this.highReg.findObjectForFile(f);
        }
        for (const lowFile of lowObj.getABAPFiles()) {
            let highSyntax = undefined;
            const highFile = highObj.getABAPFileByName(lowFile.getFilename());
            if (highFile === undefined) {
                continue;
            }
            const lowStatements = lowFile.getStatements();
            const highStatements = highFile.getStatements();
            if (lowStatements.length !== highStatements.length) {
                // after applying a fix, there might be more statements in lowFile
                // should highReg be initialized again?
                /*
                const message = "Internal Error: Statement lengths does not match";
                ret.push(Issue.atStatement(lowFile, lowStatements[0], message, this.getMetadata().key));
                */
                // hmm, add some way to disable lazyUnknown() in statement_parser.ts
                // alternatively explicit enable it in vscode, its only relevant when a user is
                // actively editing the files
                continue;
            }
            if (highSyntax === undefined) {
                highSyntax = new syntax_1.SyntaxLogic(this.highReg, highSyntaxObj).run();
            }
            let containsUnknown = false;
            for (let i = 0; i < lowStatements.length; i++) {
                const low = lowStatements[i];
                const high = highStatements[i];
                if ((low.get() instanceof _statement_1.Unknown && !(high.get() instanceof _statement_1.Unknown))
                    || high.findFirstExpression(Expressions.InlineData)) {
                    containsUnknown = true;
                    try {
                        const issue = this.checkStatement(low, high, lowFile, highSyntax, highFile);
                        if (issue) {
                            ret.push(issue);
                        }
                    }
                    catch (e) {
                        if (e instanceof SkipToNextFile) {
                            ret.push(e.issue);
                            break;
                        }
                        else {
                            throw e;
                        }
                    }
                }
            }
            if (ret.length === 0 && containsUnknown) {
                // this is a hack in order not to change too many unit tests
                for (let i = 0; i < lowStatements.length; i++) {
                    const high = highStatements[i];
                    if (high.get() instanceof Statements.Data) {
                        const issue = this.anonymousTableType(high, lowFile, highSyntax);
                        if (issue) {
                            ret.push(issue);
                        }
                    }
                }
            }
            if (ret.length === 0 && lowFile.getRaw().includes(" xsdbool(")) {
                for (let i = 0; i < lowStatements.length; i++) {
                    const high = highStatements[i];
                    const issue = this.replaceXsdBool(high, lowFile, highSyntax);
                    if (issue) {
                        ret.push(issue);
                    }
                }
            }
        }
        return ret;
    }
    ////////////////////
    /** clones the orginal repository into highReg, and parses it with higher language version */
    initHighReg() {
        // use default configuration, ie. default target version
        const highConfig = config_1.Config.getDefault().get();
        const lowConfig = this.lowReg.getConfig().get();
        highConfig.syntax.errorNamespace = lowConfig.syntax.errorNamespace;
        highConfig.syntax.globalConstants = lowConfig.syntax.globalConstants;
        highConfig.syntax.globalMacros = lowConfig.syntax.globalMacros;
        this.highReg = new registry_1.Registry();
        for (const o of this.lowReg.getObjects()) {
            for (const f of o.getFiles()) {
                if (this.lowReg.isDependency(o) === true) {
                    this.highReg.addDependency(f);
                }
                else {
                    this.highReg.addFile(f);
                }
            }
        }
        this.highReg.parse();
    }
    /** applies one rule at a time, multiple iterations are required to transform complex statements */
    checkStatement(low, high, lowFile, highSyntax, highFile) {
        if (low.getFirstToken().getStart() instanceof virtual_position_1.VirtualPosition) {
            return undefined;
        }
        // downport XSDBOOL() early, as it is valid 702 syntax
        /*
        let found = this.replaceXsdBool(high, lowFile, highSyntax);
        if (found) {
          return found;
        }
        */
        let found = this.downportEnum(low, high, lowFile, highSyntax, highFile);
        if (found) {
            return found;
        }
        found = this.partiallyImplemented(high, lowFile);
        if (found) {
            return found;
        }
        found = this.raiseException(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.emptyKey(low, high, lowFile);
        if (found) {
            return found;
        }
        found = this.stringTemplateAlpha(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.moveWithOperator(low, high, lowFile);
        if (found) {
            return found;
        }
        found = this.moveWithSimpleValue(low, high, lowFile);
        if (found) {
            return found;
        }
        found = this.assignWithTable(low, high, lowFile);
        if (found) {
            return found;
        }
        found = this.assignComponent(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportRefSimple(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportCorrespondingSimple(high, lowFile);
        if (found) {
            return found;
        }
        found = this.downportRef(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportLoopGroup(high, lowFile, highSyntax, highFile);
        if (found) {
            return found;
        }
        found = this.callFunctionParameterSimple(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.moveWithTableTarget(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportSelectInline(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportSelectExistence(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportSQLMoveInto(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportSQLExtras(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineLoopInput(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineLoopTarget(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        let skipValue = false;
        let skipReduce = false;
        const valueBody = high.findFirstExpression(Expressions.ValueBody);
        const reduceBody = high.findFirstExpression(Expressions.ReduceBody);
        if (valueBody && reduceBody) {
            const valueToken = valueBody.getFirstToken();
            const reduceToken = reduceBody.getFirstToken();
            if (valueToken.getStart().isBefore(reduceToken.getStart())) {
                skipReduce = true;
            }
            else {
                skipValue = true;
            }
        }
        if (skipValue !== true) {
            found = this.outlineValue(low, high, lowFile, highSyntax);
            if (found) {
                return found;
            }
        }
        if (skipReduce !== true) {
            found = this.outlineReduce(low, high, lowFile, highSyntax);
            if (found) {
                return found;
            }
        }
        found = this.outlineCorresponding(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportSelectFields(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineSwitch(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineFilter(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineCast(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineConv(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineCond(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineCatchSimple(high, lowFile);
        if (found) {
            return found;
        }
        found = this.outlineGetReferenceSimple(high, lowFile);
        if (found) {
            return found;
        }
        found = this.outlineDataSimple(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineData(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.outlineFS(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.newToCreateObject(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.replaceLineFunctions(high, lowFile, highSyntax, highFile);
        if (found) {
            return found;
        }
        found = this.getReference(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.replaceContains(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.replaceMethodConditional(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.replaceTableExpression(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.replaceAppendExpression(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.replaceInsertExpression(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportMessageSource(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportMessage(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportReadTable(high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        return undefined;
    }
    //////////////////////////////////////////
    /** move INTO from after WHERE to after FROM */
    downportSQLMoveInto(low, high, lowFile, _highSyntax) {
        var _a;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        // note: SQLCond is also used in JOIN(FROM) conditions
        const where = (_a = high.findFirstExpression(Expressions.Select)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.SQLCond);
        if (where === undefined) {
            return undefined;
        }
        let into = high.findFirstExpression(Expressions.SQLIntoList);
        if (into === undefined) {
            into = high.findFirstExpression(Expressions.SQLIntoStructure);
        }
        if (into === undefined) {
            into = high.findFirstExpression(Expressions.SQLIntoTable);
        }
        if (into === undefined) {
            return undefined;
        }
        if (where.getLastToken().getEnd().isBefore(into.getFirstToken().getStart()) === false) {
            return undefined;
        }
        const from = high.findFirstExpression(Expressions.SQLFrom);
        if (from === undefined) {
            return undefined;
        }
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, from.getLastToken().getEnd(), ` ` + into.concatTokens());
        const fix2 = edit_helper_1.EditHelper.deleteRange(lowFile, into.getFirstToken().getStart(), into.getLastToken().getEnd());
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, low.getFirstToken(), "SQL, move INTO", this.getMetadata().key, this.conf.severity, fix);
    }
    /** removes @'s and commas */
    downportSQLExtras(low, high, lowFile, highSyntax) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        if (!(high.get() instanceof Statements.Select)
            && !(high.get() instanceof Statements.SelectLoop)
            && !(high.get() instanceof Statements.UpdateDatabase)
            && !(high.get() instanceof Statements.ModifyDatabase)
            && !(high.get() instanceof Statements.DeleteDatabase)
            && !(high.get() instanceof Statements.InsertDatabase)) {
            return undefined;
        }
        let fix = undefined;
        const addFix = (token) => {
            const add = edit_helper_1.EditHelper.deleteToken(lowFile, token);
            if (fix === undefined) {
                fix = add;
            }
            else {
                fix = edit_helper_1.EditHelper.merge(fix, add);
            }
        };
        const candidates = [high.findAllExpressionsRecursive(Expressions.SQLTarget),
            high.findAllExpressionsRecursive(Expressions.SQLSource),
            high.findAllExpressionsRecursive(Expressions.SQLSourceNoSpace),
            high.findAllExpressionsRecursive(Expressions.SQLSourceSimple)].flat();
        for (const c of candidates.reverse()) {
            if (c.getFirstToken() instanceof tokens_1.WAt
                || c.getFirstToken() instanceof tokens_1.At) {
                const tokens = c.getAllTokens();
                if (tokens[1] instanceof tokens_1.ParenLeftW && tokens[tokens.length - 1] instanceof tokens_1.WParenRightW) {
                    const source = c.findDirectExpression(Expressions.Source);
                    const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
                    const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getStart(), `DATA(${uniqueName}) = ${source === null || source === void 0 ? void 0 : source.concatTokens()}.\n`);
                    const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, c.getFirstToken().getStart(), c.getLastToken().getEnd(), "@" + uniqueName);
                    const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
                    return issue_1.Issue.atToken(lowFile, low.getFirstToken(), "SQL, outline complex @", this.getMetadata().key, this.conf.severity, fix);
                }
                else {
                    addFix(c.getFirstToken());
                }
            }
        }
        for (const fieldList of high.findAllExpressionsMulti([Expressions.SQLFieldList, Expressions.SQLFieldListLoop, Expressions.SQLOrderBy], true)) {
            for (const token of fieldList.getDirectTokens()) {
                if (token.getStr() === ",") {
                    addFix(token);
                }
            }
        }
        if (fix !== undefined) {
            return issue_1.Issue.atToken(lowFile, low.getFirstToken(), "SQL, remove @ and ,", this.getMetadata().key, this.conf.severity, fix);
        }
        for (const c of high.findAllExpressionsRecursive(Expressions.SQLIn)) {
            const children = c.getChildren();
            const first = children[1];
            if (!(first.get() instanceof tokens_1.WParenLeftW)) {
                continue;
            }
            const last = children[children.length - 1];
            if (last.get() instanceof tokens_1.WParenRightW || last.get() instanceof tokens_1.WParenRight) {
                const firstEnd = first.getFirstToken().getEnd();
                const endDelete = new position_1.Position(firstEnd.getRow(), firstEnd.getCol() + 1);
                const fix1 = edit_helper_1.EditHelper.deleteRange(lowFile, firstEnd, endDelete);
                const lastStart = last.getFirstToken().getStart();
                const startDelete = new position_1.Position(lastStart.getRow(), lastStart.getCol() - 1);
                const fix2 = edit_helper_1.EditHelper.deleteRange(lowFile, startDelete, lastStart);
                fix = edit_helper_1.EditHelper.merge(fix2, fix1);
                return issue_1.Issue.atToken(lowFile, low.getFirstToken(), "SQL, remove spaces", this.getMetadata().key, this.conf.severity, fix);
            }
        }
        return undefined;
    }
    downportSelectExistence(low, high, lowFile, highSyntax) {
        var _a, _b, _c, _d;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        else if (!(high.get() instanceof Statements.Select)) {
            return undefined;
        }
        const fieldList = high.findFirstExpression(Expressions.SQLFieldList);
        if ((fieldList === null || fieldList === void 0 ? void 0 : fieldList.concatTokens().toUpperCase()) !== "@ABAP_TRUE") {
            return undefined;
        }
        const fieldName = (_b = (_a = high.findFirstExpression(Expressions.SQLCond)) === null || _a === void 0 ? void 0 : _a.findFirstExpression(Expressions.SQLFieldName)) === null || _b === void 0 ? void 0 : _b.concatTokens();
        if (fieldName === undefined) {
            return undefined;
        }
        const into = high.findFirstExpression(Expressions.SQLIntoStructure);
        if (into === undefined) {
            return undefined;
        }
        const intoName = (_d = (_c = into.findFirstExpression(Expressions.SQLTarget)) === null || _c === void 0 ? void 0 : _c.findFirstExpression(Expressions.Target)) === null || _d === void 0 ? void 0 : _d.concatTokens();
        const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const fix1 = edit_helper_1.EditHelper.replaceRange(lowFile, fieldList.getFirstToken().getStart(), fieldList.getLastToken().getEnd(), fieldName);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, into === null || into === void 0 ? void 0 : into.getFirstToken().getStart(), into === null || into === void 0 ? void 0 : into.getLastToken().getEnd(), `INTO @DATA(${uniqueName})`);
        let fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        const fix3 = edit_helper_1.EditHelper.insertAt(lowFile, high.getLastToken().getEnd(), `\nCLEAR ${intoName}.\nIF sy-subrc = 0.\n  ${intoName} = abap_true.\nENDIF.`);
        fix = edit_helper_1.EditHelper.merge(fix, fix3);
        return issue_1.Issue.atToken(lowFile, low.getFirstToken(), "SQL, refactor existence check", this.getMetadata().key, this.conf.severity, fix);
    }
    downportSelectInline(low, high, lowFile, highSyntax) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        else if (!(high.get() instanceof Statements.Select) && !(high.get() instanceof Statements.SelectLoop)) {
            return undefined;
        }
        // as first step outline the @DATA, note that void types are okay, as long the field names are specified
        let found = this.downportSelectSingleInline(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        found = this.downportSelectTableInline(low, high, lowFile, highSyntax);
        if (found) {
            return found;
        }
        return undefined;
    }
    downportSelectFields(low, high, lowFile, _highSyntax) {
        var _a;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        else if (!(high.get() instanceof Statements.Select)) {
            return undefined;
        }
        const fields = high.findFirstExpression(Expressions.SQLFields);
        if (fields === undefined) {
            return undefined;
        }
        const code = (_a = fields.getLastChild()) === null || _a === void 0 ? void 0 : _a.concatTokens();
        if (code === undefined) {
            return undefined;
        }
        const fix1 = edit_helper_1.EditHelper.deleteRange(lowFile, fields.getFirstToken().getStart(), fields.getLastToken().getEnd());
        const fix2 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getEnd(), " " + code);
        const fix = edit_helper_1.EditHelper.merge(fix1, fix2);
        return issue_1.Issue.atToken(lowFile, fields.getFirstToken(), "Replace FIELDS", this.getMetadata().key, this.conf.severity, fix);
    }
    downportSelectSingleInline(low, high, lowFile, _highSyntax) {
        var _a, _b, _c, _d;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        const targets = ((_a = high.findFirstExpression(Expressions.SQLIntoStructure)) === null || _a === void 0 ? void 0 : _a.findDirectExpressions(Expressions.SQLTarget)) || [];
        if (targets.length !== 1) {
            return undefined;
        }
        const inlineData = targets[0].findFirstExpression(Expressions.InlineData);
        if (inlineData === undefined) {
            return undefined;
        }
        const sqlFrom = high.findAllExpressions(Expressions.SQLFromSource);
        if (sqlFrom.length !== 1) {
            return undefined;
        }
        const tableName = (_b = sqlFrom[0].findDirectExpression(Expressions.DatabaseTable)) === null || _b === void 0 ? void 0 : _b.concatTokens();
        if (tableName === undefined) {
            return undefined;
        }
        const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
        let fieldList = high.findFirstExpression(Expressions.SQLFieldList);
        if (fieldList === undefined) {
            fieldList = high.findFirstExpression(Expressions.SQLFieldListLoop);
        }
        if (fieldList === undefined) {
            return undefined;
        }
        let fieldDefinition = "";
        const fields = fieldList.findAllExpressions(Expressions.SQLFieldName);
        const name = ((_c = inlineData.findFirstExpression(Expressions.TargetField)) === null || _c === void 0 ? void 0 : _c.concatTokens()) || "error";
        if (fields.length === 1) {
            fieldDefinition = `DATA ${name} TYPE ${tableName}-${fields[0].concatTokens()}.`;
        }
        else if (fieldList.concatTokens() === "*") {
            fieldDefinition = `DATA ${name} TYPE ${tableName}.`;
        }
        else if (fieldList.concatTokens().toUpperCase() === "COUNT( * )") {
            fieldDefinition = `DATA ${name} TYPE i.`;
        }
        else if (fieldList.concatTokens().toUpperCase() === "@ABAP_TRUE"
            || fieldList.concatTokens().toUpperCase() === "@ABAP_FALSE") {
            fieldDefinition = `DATA ${name} TYPE abap_bool.`;
        }
        else if (fieldList.getChildren().length === 1 && fieldList.getChildren()[0].get() instanceof Expressions.SQLAggregation) {
            const c = fieldList.getChildren()[0];
            if (c instanceof nodes_1.ExpressionNode) {
                const concat = (_d = c.findFirstExpression(Expressions.SQLArithmetics)) === null || _d === void 0 ? void 0 : _d.concatTokens();
                fieldDefinition = `DATA ${name} TYPE ${tableName}-${concat}.`;
            }
        }
        else {
            for (const f of fields) {
                const fieldName = f.concatTokens();
                fieldDefinition += indentation + "        " + fieldName + " TYPE " + tableName + "-" + fieldName + ",\n";
            }
            fieldDefinition = `DATA: BEGIN OF ${name},
${fieldDefinition}${indentation}      END OF ${name}.`;
        }
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getStart(), `${fieldDefinition}
${indentation}`);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, inlineData.getFirstToken().getStart(), inlineData.getLastToken().getEnd(), name);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, inlineData.getFirstToken(), "Outline SELECT @DATA", this.getMetadata().key, this.conf.severity, fix);
    }
    downportSelectTableInline(low, high, lowFile, highSyntax) {
        var _a, _b, _c, _d, _e, _f, _g;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        const targets = ((_a = high.findFirstExpression(Expressions.SQLIntoTable)) === null || _a === void 0 ? void 0 : _a.findDirectExpressions(Expressions.SQLTarget)) || [];
        if (targets.length !== 1) {
            return undefined;
        }
        const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
        const inlineData = targets[0].findFirstExpression(Expressions.InlineData);
        if (inlineData === undefined) {
            return undefined;
        }
        const sqlFrom = high.findAllExpressions(Expressions.SQLFromSource);
        if (sqlFrom.length === 0) {
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Error outlining, sqlFrom not found", this.getMetadata().key, this.conf.severity);
        }
        let tableName = (_b = sqlFrom[0].findDirectExpression(Expressions.DatabaseTable)) === null || _b === void 0 ? void 0 : _b.concatTokens();
        if (tableName === undefined) {
            return undefined;
        }
        const tableMap = {};
        for (const from of sqlFrom) {
            const dbName = (_c = from.findDirectExpression(Expressions.DatabaseTable)) === null || _c === void 0 ? void 0 : _c.concatTokens().toUpperCase();
            if (dbName === undefined) {
                continue;
            }
            const asName = ((_d = from.findDirectExpression(Expressions.SQLAsName)) === null || _d === void 0 ? void 0 : _d.concatTokens().toUpperCase()) || dbName;
            tableMap[asName] = dbName;
        }
        const fieldList = high.findFirstExpression(Expressions.SQLFieldList);
        if (fieldList === undefined) {
            return undefined;
        }
        let fieldDefinitions = "";
        for (const f of fieldList.findAllExpressions(Expressions.SQLField)) {
            let fieldName = (_e = f.findFirstExpression(Expressions.SQLFieldName)) === null || _e === void 0 ? void 0 : _e.concatTokens();
            if (fieldName === undefined) {
                continue;
            }
            if (fieldName.includes("~")) {
                const split = fieldName.split("~");
                tableName = split[0];
                fieldName = split[1];
            }
            const translated = tableMap[tableName.toUpperCase()];
            const typeName = translated + "-" + fieldName;
            fieldName = ((_f = f.findFirstExpression(Expressions.SQLAsName)) === null || _f === void 0 ? void 0 : _f.concatTokens()) || fieldName;
            fieldDefinitions += indentation + "        " + fieldName + " TYPE " + typeName.toLowerCase() + ",\n";
        }
        const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const name = ((_g = inlineData.findFirstExpression(Expressions.TargetField)) === null || _g === void 0 ? void 0 : _g.concatTokens()) || "error";
        let fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getStart(), `TYPES: BEGIN OF ${uniqueName},
${fieldDefinitions}${indentation}      END OF ${uniqueName}.
${indentation}DATA ${name} TYPE STANDARD TABLE OF ${uniqueName} WITH DEFAULT KEY.
${indentation}`);
        if (fieldDefinitions === "") {
            fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getStart(), `DATA ${name} TYPE STANDARD TABLE OF ${tableName} WITH DEFAULT KEY.
${indentation}`);
        }
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, inlineData.getFirstToken().getStart(), inlineData.getLastToken().getEnd(), name);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, inlineData.getFirstToken(), "Outline SELECT @DATA", this.getMetadata().key, this.conf.severity, fix);
    }
    // the anonymous type might be used in inferred type statements, define it so it can be referred
    anonymousTableType(high, lowFile, highSyntax) {
        if (!(high.get() instanceof Statements.Data)) {
            return undefined;
        }
        const tt = high.findFirstExpression(Expressions.TypeTable);
        if (tt === undefined) {
            return undefined;
        }
        const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const code = `TYPES ${uniqueName} ${tt.concatTokens()}.\n`;
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getStart(), code);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, tt.getFirstToken().getStart(), tt.getLastToken().getEnd(), "TYPE " + uniqueName);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Add type for table definition", this.getMetadata().key, this.conf.severity, fix);
    }
    downportMessage(high, lowFile, highSyntax) {
        var _a, _b;
        if (!(high.get() instanceof Statements.Message)) {
            return undefined;
        }
        const foundWith = high.findExpressionAfterToken("WITH");
        if (foundWith === undefined) {
            return undefined;
        }
        const likeSource = high.findExpressionAfterToken("LIKE");
        for (const s of high.findAllExpressions(Expressions.Source)) {
            if (s === likeSource) {
                continue;
            }
            else if (s.getChildren().length === 1 && ((_a = s.getFirstChild()) === null || _a === void 0 ? void 0 : _a.get()) instanceof Expressions.Constant) {
                continue;
            }
            else if (s.getChildren().length === 1 && ((_b = s.getFirstChild()) === null || _b === void 0 ? void 0 : _b.get()) instanceof Expressions.FieldChain) {
                continue;
            }
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const firstToken = high.getFirstToken();
            const code = `DATA(${uniqueName}) = ${s.concatTokens()}.\n${indentation}`;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, firstToken.getStart(), code);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, s.getFirstToken().getStart(), s.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Refactor MESSAGE WITH source", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    replaceAppendExpression(high, lowFile, highSyntax) {
        if (!(high.get() instanceof Statements.Append)) {
            return undefined;
        }
        const children = high.getChildren();
        if (children[1].get() instanceof Expressions.Source) {
            const source = children[1];
            const target = high.findDirectExpression(Expressions.Target);
            if (target === undefined) {
                return undefined;
            }
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const firstToken = high.getFirstToken();
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, firstToken.getStart(), `DATA ${uniqueName} LIKE LINE OF ${target === null || target === void 0 ? void 0 : target.concatTokens()}.
${indentation}${uniqueName} = ${source.concatTokens()}.\n${indentation}`);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, source.getFirstToken().getStart(), source.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Outline APPEND source expression", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    downportReadTable(high, lowFile, highSyntax) {
        if (!(high.get() instanceof Statements.ReadTable)) {
            return undefined;
        }
        const source = high.findExpressionAfterToken("TABLE");
        if ((source === null || source === void 0 ? void 0 : source.get()) instanceof Expressions.Source) {
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const firstToken = high.getFirstToken();
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, firstToken.getStart(), `DATA(${uniqueName}) = ${source.concatTokens()}.\n` + indentation);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, source.getFirstToken().getStart(), source.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Outline table source", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    downportMessageSource(high, lowFile, highSyntax) {
        var _a;
        if (!(high.get() instanceof Statements.Message)) {
            return undefined;
        }
        const source = high.findExpressionAfterToken("MESSAGE");
        if ((source === null || source === void 0 ? void 0 : source.get()) instanceof Expressions.MessageSourceSource
            && ((_a = source.getFirstChild()) === null || _a === void 0 ? void 0 : _a.get()) instanceof Expressions.Source) {
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const firstToken = high.getFirstToken();
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, firstToken.getStart(), `DATA(${uniqueName}) = ${source.concatTokens()}.\n` + indentation);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, source.getFirstToken().getStart(), source.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Outline message source", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    replaceInsertExpression(high, lowFile, highSyntax) {
        if (!(high.get() instanceof Statements.InsertInternal)) {
            return undefined;
        }
        const children = high.getChildren();
        if (children[1].get() instanceof Expressions.Source) {
            const source = children[1];
            const target = high.findDirectExpression(Expressions.Target);
            if (target === undefined) {
                return undefined;
            }
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const firstToken = high.getFirstToken();
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, firstToken.getStart(), `DATA ${uniqueName} LIKE LINE OF ${target === null || target === void 0 ? void 0 : target.concatTokens()}.
${indentation}${uniqueName} = ${source.concatTokens()}.\n${indentation}`);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, source.getFirstToken().getStart(), source.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Outline INSERT source expression", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    replaceTableExpression(low, high, lowFile, highSyntax) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const fieldChain of high.findAllExpressionsRecursive(Expressions.FieldChain)) {
            const tableExpression = fieldChain.findDirectExpression(Expressions.TableExpression);
            if (tableExpression === undefined) {
                continue;
            }
            const concat = high.concatTokens().toUpperCase();
            if (concat.includes(" LINE_EXISTS( ") || concat.includes(" LINE_INDEX( ")) {
                // note: line_exists() must be replaced before handling table expressions
                continue;
            }
            let pre = "";
            let startToken = undefined;
            for (const child of fieldChain.getChildren()) {
                if (startToken === undefined) {
                    startToken = child.getFirstToken();
                }
                else if (child === tableExpression) {
                    break;
                }
                pre += child.concatTokens();
            }
            if (startToken === undefined) {
                continue;
            }
            const condition = this.tableCondition(tableExpression);
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const tabixBackup = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const firstToken = high.getFirstToken();
            // note that the tabix restore should be done before throwing the exception
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, firstToken.getStart(), `DATA ${uniqueName} LIKE LINE OF ${pre}.
${indentation}DATA ${tabixBackup} LIKE sy-tabix.
${indentation}${tabixBackup} = sy-tabix.
${indentation}READ TABLE ${pre} ${condition}INTO ${uniqueName}.
${indentation}sy-tabix = ${tabixBackup}.
${indentation}IF sy-subrc <> 0.
${indentation}  RAISE EXCEPTION TYPE cx_sy_itab_line_not_found.
${indentation}ENDIF.
${indentation}`);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, startToken.getStart(), tableExpression.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            if (high.get() instanceof Statements.ElseIf) {
                throw "downport, unable to downport table expression in ELSEIF";
            }
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Outline table expression", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    tableCondition(tableExpression) {
        let condition = "";
        let keyName = "";
        for (const c of tableExpression.getChildren() || []) {
            if (c.getFirstToken().getStr() === "[" || c.getFirstToken().getStr() === "]") {
                continue;
            }
            else if (c.get() instanceof Expressions.ComponentChainSimple && condition === "") {
                if (keyName === "") {
                    condition = "WITH KEY ";
                }
                else {
                    condition = "WITH TABLE KEY " + keyName + " COMPONENTS ";
                }
            }
            else if (c.get() instanceof Expressions.Source && condition === "") {
                condition = "INDEX ";
            }
            else if (c instanceof nodes_1.TokenNode && c.getFirstToken().getStr().toUpperCase() === "KEY") {
                continue;
            }
            else if (c.get() instanceof Expressions.SimpleName) {
                keyName = c.concatTokens();
                continue;
            }
            condition += c.concatTokens() + " ";
        }
        return condition;
    }
    outlineCatchSimple(node, lowFile) {
        // outlines "CATCH cx_bcs INTO DATA(lx_bcs_excep).", note that this does not need to look at types
        var _a, _b;
        if (this.lowReg.getConfig().getVersion() === version_1.Version.OpenABAP) {
            return undefined;
        }
        if (!(node.get() instanceof Statements.Catch)) {
            return undefined;
        }
        const target = node.findFirstExpression(Expressions.Target);
        if (!(((_a = target === null || target === void 0 ? void 0 : target.getFirstChild()) === null || _a === void 0 ? void 0 : _a.get()) instanceof Expressions.InlineData)) {
            return undefined;
        }
        const classNames = node.findDirectExpressions(Expressions.ClassName);
        if (classNames.length !== 1) {
            return undefined;
        }
        const className = classNames[0].concatTokens();
        const targetName = (_b = target.findFirstExpression(Expressions.TargetField)) === null || _b === void 0 ? void 0 : _b.concatTokens();
        const indentation = " ".repeat(node.getFirstToken().getStart().getCol() - 1);
        const code = `  DATA ${targetName} TYPE REF TO ${className}.
${indentation}CATCH ${className} INTO ${targetName}.`;
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, node.getStart(), node.getEnd(), code);
        return issue_1.Issue.atToken(lowFile, node.getFirstToken(), "Outline DATA", this.getMetadata().key, this.conf.severity, fix);
    }
    outlineGetReferenceSimple(node, lowFile) {
        var _a, _b, _c;
        if (!(node.get() instanceof Statements.GetReference)) {
            return undefined;
        }
        const target = node.findFirstExpression(Expressions.Target);
        if (!(((_a = target === null || target === void 0 ? void 0 : target.getFirstChild()) === null || _a === void 0 ? void 0 : _a.get()) instanceof Expressions.InlineData)) {
            return undefined;
        }
        const source = node.findFirstExpression(Expressions.Source);
        if (!(((_b = source === null || source === void 0 ? void 0 : source.getFirstChild()) === null || _b === void 0 ? void 0 : _b.get()) instanceof Expressions.FieldChain)) {
            return undefined;
        }
        const targetName = ((_c = target.findFirstExpression(Expressions.TargetField)) === null || _c === void 0 ? void 0 : _c.concatTokens()) || "errorError";
        const indentation = " ".repeat(node.getFirstToken().getStart().getCol() - 1);
        const firstToken = target.getFirstToken();
        const lastToken = target.getLastToken();
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, node.getStart(), `DATA ${targetName} LIKE REF TO ${source.concatTokens()}.\n${indentation}`);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, firstToken.getStart(), lastToken.getEnd(), targetName);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, node.getFirstToken(), "Outline DATA", this.getMetadata().key, this.conf.severity, fix);
    }
    outlineDataSimple(node, lowFile, highSyntax) {
        var _a, _b, _c;
        if (!(node.get() instanceof Statements.Move)) {
            return undefined;
        }
        if (this.lowReg.getConfig().getVersion() === version_1.Version.OpenABAP) {
            return undefined;
        }
        const target = node.findFirstExpression(Expressions.Target);
        if (!(((_a = target === null || target === void 0 ? void 0 : target.getFirstChild()) === null || _a === void 0 ? void 0 : _a.get()) instanceof Expressions.InlineData)) {
            return undefined;
        }
        let type = "";
        const source = node.findFirstExpression(Expressions.Source);
        if (source === undefined) {
            return undefined;
        }
        else if (source.getChildren().length !== 1) {
            return undefined;
        }
        else if (!(((_b = source.getFirstChild()) === null || _b === void 0 ? void 0 : _b.get()) instanceof Expressions.FieldChain)) {
            return undefined;
        }
        else if (source.findFirstExpression(Expressions.FieldOffset)) {
            return undefined;
        }
        else if (source.findFirstExpression(Expressions.FieldLength)) {
            return undefined;
        }
        else if (source.findFirstExpression(Expressions.TableExpression)) {
            const chain = source.findDirectExpression(Expressions.FieldChain);
            if (chain !== undefined
                && chain.getChildren().length === 2
                && chain.getChildren()[0].get() instanceof Expressions.SourceField
                && chain.getChildren()[1].get() instanceof Expressions.TableExpression) {
                type = "LINE OF " + chain.getChildren()[0].concatTokens();
            }
            else {
                return undefined;
            }
        }
        else {
            const spag = highSyntax.spaghetti.lookupPosition(source.getFirstToken().getStart(), lowFile.getFilename());
            if (spag) {
                const found = spag.findVariable(source.concatTokens());
                if (found && found.getType().isGeneric() === true) {
                    return undefined;
                }
            }
            type = source.concatTokens();
        }
        const targetName = (_c = target.findFirstExpression(Expressions.TargetField)) === null || _c === void 0 ? void 0 : _c.concatTokens();
        const indentation = " ".repeat(node.getFirstToken().getStart().getCol() - 1);
        const firstToken = node.getFirstToken();
        const lastToken = node.getLastToken();
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, firstToken.getStart(), `DATA ${targetName} LIKE ${type}.\n${indentation}`);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, firstToken.getStart(), lastToken.getEnd(), `${targetName} = ${source.concatTokens()}.`);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, node.getFirstToken(), "Outline DATA", this.getMetadata().key, this.conf.severity, fix);
    }
    partiallyImplemented(node, lowFile) {
        if (this.lowReg.getConfig().getVersion() === version_1.Version.OpenABAP) {
            return undefined;
        }
        if (node.get() instanceof Statements.InterfaceDef) {
            const partially = node.findDirectTokenByText("PARTIALLY");
            if (partially === undefined) {
                return undefined;
            }
            const implemented = node.findDirectTokenByText("IMPLEMENTED");
            if (implemented === undefined) {
                return undefined;
            }
            const fix = edit_helper_1.EditHelper.deleteRange(lowFile, partially.getStart(), implemented.getEnd());
            return issue_1.Issue.atToken(lowFile, partially, "Downport PARTIALLY IMPLEMENTED", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    raiseException(node, lowFile, highSyntax) {
        /*
        Note: IF_T100_DYN_MSG does not exist in 702, so this rule is mostly relevant for the transpiler
    
        DATA foo LIKE if_t100_message=>t100key.
        foo-msgid = 'ZHVAM'.
        foo-msgno = '001'.
        foo-attr1 = 'IF_T100_DYN_MSG~MSGV1'.
        foo-attr2 = 'IF_T100_DYN_MSG~MSGV2'.
        foo-attr3 = 'IF_T100_DYN_MSG~MSGV3'.
        foo-attr4 = 'IF_T100_DYN_MSG~MSGV4'.
        DATA bar TYPE REF TO zcl_hvam_exception.
        CREATE OBJECT bar EXPORTING textid = foo.
        bar->if_t100_dyn_msg~msgty = 'E'.
        bar->if_t100_dyn_msg~msgv1 = 'abc'.
        bar->if_t100_dyn_msg~msgv2 = 'abc'.
        bar->if_t100_dyn_msg~msgv3 = 'abc'.
        bar->if_t100_dyn_msg~msgv4 = 'abc'.
        RAISE EXCEPTION bar.
        */
        var _a, _b, _c, _d, _e, _f, _g;
        if (!(node.get() instanceof Statements.Raise)) {
            return undefined;
        }
        let id = undefined;
        let number = undefined;
        let startToken = node.findDirectTokenByText("ID");
        if (startToken) {
            const sources = node.findDirectExpressions(Expressions.Source);
            id = sources[0].concatTokens();
            const numberExpression = node.findExpressionAfterToken("NUMBER");
            if (numberExpression === undefined) {
                throw "downport raiseException, could not find number";
            }
            number = numberExpression.concatTokens();
            if (numberExpression.get() instanceof Expressions.MessageNumber) {
                number = "'" + number + "'";
            }
        }
        else {
            const s = node.findDirectExpression(Expressions.MessageSource);
            if (s === undefined) {
                return undefined;
            }
            if (s.findDirectExpression(Expressions.MessageClass)) {
                id = "'" + ((_b = (_a = s.findDirectExpression(Expressions.MessageClass)) === null || _a === void 0 ? void 0 : _a.concatTokens()) === null || _b === void 0 ? void 0 : _b.toUpperCase()) + "'";
            }
            else {
                id = (_c = s.findExpressionAfterToken("ID")) === null || _c === void 0 ? void 0 : _c.concatTokens();
            }
            if (s.findDirectExpression(Expressions.MessageTypeAndNumber)) {
                number = "'" + ((_d = s.findDirectExpression(Expressions.MessageTypeAndNumber)) === null || _d === void 0 ? void 0 : _d.concatTokens().substring(1)) + "'";
            }
            else {
                number = (_e = s.findExpressionAfterToken("NUMBER")) === null || _e === void 0 ? void 0 : _e.concatTokens();
            }
            startToken = node.getFirstToken();
        }
        const withs = ((_f = node.findDirectExpression(Expressions.RaiseWith)) === null || _f === void 0 ? void 0 : _f.findDirectExpressionsMulti([Expressions.SimpleSource1, Expressions.Source])) || [];
        const className = ((_g = node.findDirectExpression(Expressions.ClassName)) === null || _g === void 0 ? void 0 : _g.concatTokens()) || "ERROR";
        const uniqueName1 = this.uniqueName(node.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const uniqueName2 = this.uniqueName(node.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const indentation = " ".repeat(node.getFirstToken().getStart().getCol() - 1);
        let abap = `DATA ${uniqueName1} LIKE if_t100_message=>t100key.
${indentation}${uniqueName1}-msgid = ${id}.
${indentation}${uniqueName1}-msgno = ${number}.\n`;
        if (withs.length > 0) {
            abap += `${indentation}${uniqueName1}-attr1 = 'IF_T100_DYN_MSG~MSGV1'.
${indentation}${uniqueName1}-attr2 = 'IF_T100_DYN_MSG~MSGV2'.
${indentation}${uniqueName1}-attr3 = 'IF_T100_DYN_MSG~MSGV3'.
${indentation}${uniqueName1}-attr4 = 'IF_T100_DYN_MSG~MSGV4'.\n`;
        }
        abap += `${indentation}DATA ${uniqueName2} TYPE REF TO ${className}.
${indentation}CREATE OBJECT ${uniqueName2}.
${indentation}${uniqueName2}->if_t100_message~t100key = ${uniqueName1}.\n`;
        if (withs.length > 0) {
            abap += `${indentation}${uniqueName2}->if_t100_dyn_msg~msgty = 'E'.\n`;
        }
        let count = 1;
        for (const w of withs) {
            abap += `${indentation}${uniqueName2}->if_t100_dyn_msg~msgv${count} = ${w.concatTokens()}.\n`;
            count++;
        }
        abap += `${indentation}RAISE EXCEPTION ${uniqueName2}.`;
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, node.getStart(), node.getEnd(), abap);
        return issue_1.Issue.atToken(lowFile, startToken, "Downport RAISE EXCEPTION MESSAGE", this.getMetadata().key, this.conf.severity, fix);
    }
    emptyKey(low, node, lowFile) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (let i of node.findAllExpressions(Expressions.TypeTable)) {
            const key = i.findDirectExpression(Expressions.TypeTableKey);
            if (key === undefined) {
                continue;
            }
            i = key;
            const concat = i.concatTokens();
            if (concat.toUpperCase().includes("WITH EMPTY KEY") === false) {
                continue;
            }
            const token = i.findDirectTokenByText("EMPTY");
            if (token === undefined) {
                continue;
            }
            const fix = edit_helper_1.EditHelper.replaceToken(lowFile, token, "DEFAULT");
            return issue_1.Issue.atToken(lowFile, i.getFirstToken(), "Downport EMPTY KEY", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    callFunctionParameterSimple(high, lowFile, highSyntax) {
        if (!(high.get() instanceof Statements.CallFunction)) {
            return undefined;
        }
        let found = undefined;
        for (const p of high.findAllExpressions(Expressions.FunctionExportingParameter)) {
            found = p.findDirectExpression(Expressions.Source);
            if (found && (found.findDirectExpression(Expressions.FieldChain)
                || found.findDirectExpression(Expressions.Constant)
                || found.findDirectExpression(Expressions.TextElement))) {
                // its actually simple, ok
                found = undefined;
            }
            else if (found !== undefined) {
                break;
            }
        }
        if (found === undefined) {
            return undefined;
        }
        const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const code = `DATA(${uniqueName}) = ${found.concatTokens()}.\n`;
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), code);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, found.getFirstToken().getStart(), found.getLastToken().getEnd(), uniqueName);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, call function parameter", this.getMetadata().key, this.conf.severity, fix);
    }
    downportCorrespondingSimple(high, lowFile) {
        var _a;
        if (!(high.get() instanceof Statements.Move)
            || high.getChildren().length !== 4
            || high.getChildren()[2].getFirstToken().getStr().toUpperCase() !== "CORRESPONDING") {
            return undefined;
        }
        const target = high.findDirectExpression(Expressions.Target);
        if (target === undefined) {
            return undefined;
        }
        const sourceRef = (_a = high.findFirstExpression(Expressions.Source)) === null || _a === void 0 ? void 0 : _a.findFirstExpression(Expressions.CorrespondingBody);
        if ((sourceRef === null || sourceRef === void 0 ? void 0 : sourceRef.getChildren().length) === 1 && target.concatTokens().toUpperCase().startsWith("DATA(") === false) {
            const code = `MOVE-CORRESPONDING ${sourceRef.concatTokens()} TO ${target.concatTokens()}`;
            const start = high.getFirstToken().getStart();
            const end = high.getLastToken().getStart();
            const fix = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, code);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, simple CORRESPONDING move", this.getMetadata().key, this.conf.severity, fix);
        }
        else if ((sourceRef === null || sourceRef === void 0 ? void 0 : sourceRef.getChildren().length) === 2 && sourceRef.findDirectExpression(Expressions.CorrespondingBodyBase) !== undefined) {
            const base = sourceRef.findDirectExpression(Expressions.CorrespondingBodyBase);
            let code = `${target.concatTokens()} = ${base.getChildren()[2].concatTokens()}.\n`;
            code += `MOVE-CORRESPONDING ${sourceRef.getChildren()[1].concatTokens()} TO ${target.concatTokens()}`;
            const start = high.getFirstToken().getStart();
            const end = high.getLastToken().getStart();
            const fix = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, code);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, CORRESPONDING BASE move", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    downportRefSimple(high, lowFile, highSyntax) {
        var _a, _b;
        if (!(high.get() instanceof Statements.Move)
            || high.getChildren().length !== 4
            || high.getChildren()[2].getFirstToken().getStr().toUpperCase() !== "REF") {
            return undefined;
        }
        const target = high.findDirectExpression(Expressions.Target);
        if (target === undefined) {
            return undefined;
        }
        const sourceRef = (_a = high.findFirstExpression(Expressions.Source)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.Source);
        if (sourceRef === undefined
            || sourceRef.getChildren().length !== 1) {
            return undefined;
        }
        let code = "";
        if (sourceRef.findFirstExpression(Expressions.TableExpression)) {
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            code = `ASSIGN ${sourceRef.concatTokens()} TO FIELD-SYMBOL(<${uniqueName}>).
IF sy-subrc <> 0.
  RAISE EXCEPTION TYPE cx_sy_itab_line_not_found.
ENDIF.
GET REFERENCE OF <${uniqueName}> INTO ${target.concatTokens()}`;
        }
        else {
            const concat = target.concatTokens();
            code = `GET REFERENCE OF ${sourceRef.concatTokens()} INTO ${concat}`;
            // workaround for handling generic ANY type,
            if (concat.toUpperCase().startsWith("DATA(")) {
                const spag = highSyntax.spaghetti.lookupPosition(high.getFirstToken().getStart(), lowFile.getFilename());
                if (spag !== undefined) {
                    const found = spag.findVariable(sourceRef.concatTokens());
                    const tt = found === null || found === void 0 ? void 0 : found.getType();
                    if (tt instanceof basic_1.AnyType) {
                        const tname = (_b = target.findFirstExpression(Expressions.TargetField)) === null || _b === void 0 ? void 0 : _b.concatTokens();
                        code = `DATA ${tname} TYPE REF TO data.\nGET REFERENCE OF ${sourceRef.concatTokens()} INTO ${tname}`;
                    }
                }
            }
        }
        const start = high.getFirstToken().getStart();
        const end = high.getLastToken().getStart();
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, code);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, simple REF move", this.getMetadata().key, this.conf.severity, fix);
    }
    downportLoopGroup(high, lowFile, highSyntax, highFile) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
        if (!(high.get() instanceof Statements.Loop)) {
            return undefined;
        }
        const group = high.findDirectExpression(Expressions.LoopGroupBy);
        if (group === undefined) {
            return undefined;
        }
        const groupTargetName = ((_a = group.findFirstExpression(Expressions.TargetField)) === null || _a === void 0 ? void 0 : _a.concatTokens())
            || ((_b = group.findFirstExpression(Expressions.TargetFieldSymbol)) === null || _b === void 0 ? void 0 : _b.concatTokens().replace("<", "_").replace(">", "_"))
            || "nameNotFound";
        const loopSourceName = ((_c = high.findFirstExpression(Expressions.SimpleSource2)) === null || _c === void 0 ? void 0 : _c.concatTokens()) || "nameNotFound";
        const loopTargetName = ((_d = high.findFirstExpression(Expressions.TargetField)) === null || _d === void 0 ? void 0 : _d.concatTokens())
            || ((_e = high.findFirstExpression(Expressions.TargetFieldSymbol)) === null || _e === void 0 ? void 0 : _e.concatTokens())
            || "nameNotFound";
        const groupTarget = ((_f = group.findDirectExpression(Expressions.LoopGroupByTarget)) === null || _f === void 0 ? void 0 : _f.concatTokens()) || "";
        const isReference = (_g = high.findFirstExpression(Expressions.LoopTarget)) === null || _g === void 0 ? void 0 : _g.concatTokens().toUpperCase().startsWith("REFERENCE INTO ");
        let loopSourceRowType = "typeNotFound";
        const spag = highSyntax.spaghetti.lookupPosition(high.getFirstToken().getStart(), lowFile.getFilename());
        if (spag !== undefined) {
            const found = spag.findVariable(loopSourceName);
            const tt = found === null || found === void 0 ? void 0 : found.getType();
            if (tt instanceof basic_1.TableType) {
                loopSourceRowType = tt.getRowType().getQualifiedName() || "typeNotFound";
            }
        }
        let code = `TYPES: BEGIN OF ${groupTargetName}type,\n`;
        let condition = "";
        let groupCountName = undefined;
        let groupIndexName = undefined;
        for (const c of group.findAllExpressions(Expressions.LoopGroupByComponent)) {
            const name = c.findFirstExpression(Expressions.ComponentName);
            let type = ((_h = c.findFirstExpression(Expressions.Source)) === null || _h === void 0 ? void 0 : _h.concatTokens()) || "todo";
            if ((_j = c.concatTokens()) === null || _j === void 0 ? void 0 : _j.toUpperCase().endsWith(" = GROUP SIZE")) {
                type = "i";
                groupCountName = name === null || name === void 0 ? void 0 : name.concatTokens();
            }
            else if ((_k = c.concatTokens()) === null || _k === void 0 ? void 0 : _k.toUpperCase().endsWith(" = GROUP INDEX")) {
                type = "i";
                groupIndexName = name === null || name === void 0 ? void 0 : name.concatTokens();
            }
            else {
                if (condition !== "") {
                    condition += " ";
                }
                condition += c.concatTokens();
                type = type.replace(loopTargetName, loopSourceRowType);
                type = type.replace("->", "-");
            }
            code += `         ${name === null || name === void 0 ? void 0 : name.concatTokens()} TYPE ${type},\n`;
        }
        const s = group.findDirectExpression(Expressions.Source);
        let singleName = "";
        if (s) {
            let type = s.concatTokens();
            type = type.replace(loopTargetName, loopSourceRowType);
            type = type.replace("->", "-");
            singleName = s.concatTokens().split("-")[1];
            code += `         ${singleName} TYPE ${type},\n`;
            condition = singleName + " = " + s.concatTokens();
        }
        const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const uniqueFS = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const uniqueNameIndex = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        code += `         items LIKE ${loopSourceName},
       END OF ${groupTargetName}type.
DATA ${groupTargetName}tab TYPE STANDARD TABLE OF ${groupTargetName}type WITH DEFAULT KEY.
DATA ${uniqueName} LIKE LINE OF ${groupTargetName}tab.
LOOP AT ${loopSourceName} ${(_l = high.findFirstExpression(Expressions.LoopTarget)) === null || _l === void 0 ? void 0 : _l.concatTokens()}.\n`;
        if (groupIndexName !== undefined) {
            code += `DATA(${uniqueNameIndex}) = sy-tabix.\n`;
        }
        code += `READ TABLE ${groupTargetName}tab ASSIGNING FIELD-SYMBOL(<${uniqueFS}>) WITH KEY ${condition}.
IF sy-subrc = 0.\n`;
        if (groupCountName !== undefined) {
            code += `  <${uniqueFS}>-${groupCountName} = <${uniqueFS}>-${groupCountName} + 1.\n`;
        }
        code += `  INSERT ${loopTargetName}${isReference ? "->*" : ""} INTO TABLE <${uniqueFS}>-items.
ELSE.\n`;
        code += `  CLEAR ${uniqueName}.\n`;
        for (const c of group.findAllExpressions(Expressions.LoopGroupByComponent)) {
            const concat = c.concatTokens();
            //      console.dir(concat);
            if (concat.endsWith(" GROUP INDEX")) {
                code += `  ${uniqueName}-${groupIndexName} = ${uniqueNameIndex}.\n`;
            }
            else if (concat.endsWith(" GROUP SIZE")) {
                code += `  ${uniqueName}-${groupCountName} = 1.\n`;
            }
            else {
                code += `  ${uniqueName}-${concat}.\n`;
            }
        }
        if (singleName !== "") {
            code += `  ${uniqueName}-${singleName} = ${loopTargetName}-${singleName}.\n`;
        }
        code += `  INSERT ${loopTargetName}${isReference ? "->*" : ""} INTO TABLE ${uniqueName}-items.\n`;
        code += `  INSERT ${uniqueName} INTO TABLE ${groupTargetName}tab.\n`;
        code += `ENDIF.
ENDLOOP.
LOOP AT ${groupTargetName}tab ${groupTarget}.`;
        let fix = edit_helper_1.EditHelper.replaceRange(lowFile, high.getFirstToken().getStart(), high.getLastToken().getEnd(), code);
        for (const l of ((_m = highFile.getStructure()) === null || _m === void 0 ? void 0 : _m.findAllStructures(Structures.Loop)) || []) {
            // make sure to find the correct/current loop statement
            if (l.findDirectStatement(Statements.Loop) !== high) {
                continue;
            }
            for (const loop of l.findAllStatements(Statements.Loop)) {
                if ((_o = loop.concatTokens()) === null || _o === void 0 ? void 0 : _o.toUpperCase().startsWith("LOOP AT GROUP ")) {
                    const subLoopSource = loop.findFirstExpression(Expressions.SimpleSource2);
                    if (subLoopSource === undefined) {
                        continue;
                    }
                    const subLoopSourceName = (subLoopSource === null || subLoopSource === void 0 ? void 0 : subLoopSource.concatTokens()) || "nameNotFound";
                    const subCode = `LOOP AT ${subLoopSourceName}${isReference ? "->" : "-"}items`;
                    const subFix = edit_helper_1.EditHelper.replaceRange(lowFile, loop.getFirstToken().getStart(), subLoopSource.getLastToken().getEnd(), subCode);
                    fix = edit_helper_1.EditHelper.merge(subFix, fix);
                }
            }
        }
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, LOOP GROUP", this.getMetadata().key, this.conf.severity, fix);
    }
    downportRef(low, high, lowFile, highSyntax) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        let found = undefined;
        for (const s of high.findAllExpressionsRecursive(Expressions.Source)) {
            if (s.getFirstToken().getStr().toUpperCase() === "REF"
                && s.findDirectExpression(Expressions.TypeNameOrInfer)) {
                found = s;
            }
        }
        if (found === undefined) {
            return undefined;
        }
        const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const code = `DATA(${uniqueName}) = ${found.concatTokens()}.\n`;
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), code);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, found.getFirstToken().getStart(), found.getLastToken().getEnd(), uniqueName);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, REF", this.getMetadata().key, this.conf.severity, fix);
    }
    assignWithTable(low, high, lowFile) {
        var _a, _b;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        if (!(high.get() instanceof Statements.Assign)) {
            return undefined;
        }
        else if (high.getChildren().length !== 5) {
            return undefined;
        }
        const fieldChain = (_b = (_a = high.findDirectExpression(Expressions.AssignSource)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.Source)) === null || _b === void 0 ? void 0 : _b.findDirectExpression(Expressions.FieldChain);
        const tableExpression = fieldChain === null || fieldChain === void 0 ? void 0 : fieldChain.getLastChild();
        if (tableExpression === undefined
            || !(tableExpression.get() instanceof Expressions.TableExpression)
            || !(tableExpression instanceof nodes_1.ExpressionNode)) {
            return undefined;
        }
        let condition = "";
        if (tableExpression.getChildren().length === 3) {
            const index = tableExpression.findDirectExpression(Expressions.Source);
            if (index === undefined) {
                return undefined;
            }
            condition = `INDEX ${index.concatTokens()}`;
        }
        else {
            let concat = tableExpression.concatTokens();
            concat = concat.substring(2);
            concat = concat.substring(0, concat.length - 2);
            condition = `WITH KEY ${concat}`;
        }
        let pre = "";
        for (const c of fieldChain.getChildren()) {
            if (c === tableExpression) {
                break;
            }
            pre += c.concatTokens();
        }
        const fsTarget = high.findDirectExpression(Expressions.FSTarget);
        const code = `READ TABLE ${pre} ${condition} ASSIGNING ${fsTarget === null || fsTarget === void 0 ? void 0 : fsTarget.concatTokens()}.`;
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, high.getFirstToken().getStart(), high.getLastToken().getEnd(), code);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, ASSIGN table expr", this.getMetadata().key, this.conf.severity, fix);
    }
    assignComponent(low, high, lowFile, highSyntax) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        if (!(high.get() instanceof Statements.Assign)) {
            return undefined;
        }
        const assignSource = high.findDirectExpression(Expressions.AssignSource);
        if (assignSource === undefined || assignSource.getFirstToken().getStr().toUpperCase() !== "COMPONENT") {
            return undefined;
        }
        const componentSource = assignSource.findExpressionAfterToken("COMPONENT");
        if (componentSource === undefined || componentSource.get() instanceof Expressions.SimpleSource3) {
            return undefined;
        }
        const uniqueName = this.uniqueName(assignSource.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const code = `DATA(${uniqueName}) = ${componentSource.concatTokens()}.\n`;
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), code);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, componentSource.getFirstToken().getStart(), componentSource.getLastToken().getEnd(), uniqueName);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, ASSIGN COMPONENT source", this.getMetadata().key, this.conf.severity, fix);
    }
    moveWithSimpleValue(low, high, lowFile) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        if (!(high.get() instanceof Statements.Move)
            || high.getChildren().length !== 4) {
            return undefined;
        }
        const target = high.findDirectExpression(Expressions.Target);
        if (target === undefined) {
            return undefined;
        }
        const source = high.findDirectExpression(Expressions.Source);
        if (source === undefined) {
            return undefined;
        }
        const field = target.findDirectExpression(Expressions.TargetField);
        if (field === undefined) {
            return;
        }
        const valueBody = source.findDirectExpression(Expressions.ValueBody);
        if (valueBody === undefined) {
            return;
        }
        const type = source.findDirectExpression(Expressions.TypeNameOrInfer);
        if (type === undefined || (type === null || type === void 0 ? void 0 : type.concatTokens()) !== "#") {
            return;
        }
        const fieldAssignments = valueBody.findDirectExpressions(Expressions.FieldAssignment);
        if (fieldAssignments.length === 0) {
            return;
        }
        else if (fieldAssignments.length !== valueBody.getChildren().length) {
            return;
        }
        const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
        let code = `CLEAR ${target.concatTokens()}.\n`;
        for (const fieldAssignment of fieldAssignments) {
            code += indentation + target.concatTokens() + "-" + fieldAssignment.concatTokensWithLinebreaks() + `.\n`;
        }
        code = code.trimEnd();
        const start = high.getFirstToken().getStart();
        const end = high.getLastToken().getEnd();
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, code);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, simple move", this.getMetadata().key, this.conf.severity, fix);
    }
    // note, downporting ENUM does not give the correct types, but it will work in most cases?
    downportEnum(_low, high, lowFile, _highSyntax, highFile) {
        var _a, _b, _c, _d, _e;
        if (!(high.get() instanceof Statements.TypeEnumBegin)) {
            return undefined;
        }
        const enumStructure = (_a = highFile.getStructure()) === null || _a === void 0 ? void 0 : _a.findFirstStructure(Structures.TypeEnum);
        if (enumStructure === undefined) {
            return undefined;
        }
        if (enumStructure.getFirstStatement() !== high) {
            return undefined;
        }
        const enumName = (_b = high.findExpressionAfterToken("ENUM")) === null || _b === void 0 ? void 0 : _b.concatTokens();
        const structureName = (_c = high.findExpressionAfterToken("STRUCTURE")) === null || _c === void 0 ? void 0 : _c.concatTokens();
        // all ENUMS are char like?
        let code = `TYPES ${enumName} TYPE string.
CONSTANTS: BEGIN OF ${structureName},\n`;
        let count = 1;
        for (const e of enumStructure.findDirectStatements(Statements.TypeEnum).concat(enumStructure.findDirectStatements(Statements.Type))) {
            const name = (_d = e.findFirstExpression(Expressions.NamespaceSimpleName)) === null || _d === void 0 ? void 0 : _d.concatTokens();
            let value = (_e = e.findFirstExpression(Expressions.Value)) === null || _e === void 0 ? void 0 : _e.concatTokens();
            if (value === undefined) {
                value = "VALUE '" + count++ + "'";
            }
            code += `             ${name} TYPE ${enumName} ${value},\n`;
        }
        code += `           END OF ${structureName}.`;
        const start = enumStructure.getFirstToken().getStart();
        const end = enumStructure.getLastToken().getEnd();
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, code);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport ENUM", this.getMetadata().key, this.conf.severity, fix);
    }
    moveWithTableTarget(low, high, lowFile, highSyntax) {
        var _a;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        if (!(high.get() instanceof Statements.Move)) {
            return undefined;
        }
        const target = high.findDirectExpression(Expressions.Target);
        if (target === undefined) {
            return undefined;
        }
        const tableExpression = target.findDirectExpression(Expressions.TableExpression);
        if (tableExpression === undefined) {
            return undefined;
        }
        const index = tableExpression.findDirectExpression(Expressions.Source);
        if (index === undefined) {
            return undefined;
        }
        let uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        uniqueName = `<${uniqueName}>`;
        const tName = target.concatTokens().split("[")[0];
        const condition = this.tableCondition(tableExpression);
        const tabixBackup = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
        // restore tabix before exeption
        const code = `FIELD-SYMBOLS ${uniqueName} LIKE LINE OF ${tName}.
${indentation}DATA ${tabixBackup} LIKE sy-tabix.
${indentation}${tabixBackup} = sy-tabix.
${indentation}READ TABLE ${tName} ${condition}ASSIGNING ${uniqueName}.
${indentation}sy-tabix = ${tabixBackup}.
${indentation}IF sy-subrc <> 0.
${indentation}  RAISE EXCEPTION TYPE cx_sy_itab_line_not_found.
${indentation}ENDIF.
${indentation}${uniqueName}`;
        const start = target.getFirstToken().getStart();
        const end = (_a = tableExpression.findDirectTokenByText("]")) === null || _a === void 0 ? void 0 : _a.getEnd();
        if (end === undefined) {
            return undefined;
        }
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, code);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport, move with table target", this.getMetadata().key, this.conf.severity, fix);
    }
    moveWithOperator(low, high, lowFile) {
        var _a, _b, _c;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        else if (!(high.get() instanceof Statements.Move)) {
            return undefined;
        }
        const children = high.getChildren();
        const secondChild = children[1];
        if (secondChild === undefined) {
            return undefined;
        }
        const op = secondChild.getFirstToken();
        let operator = "";
        switch (op.getStr()) {
            case "+":
                operator = " + ";
                break;
            case "-":
                operator = " - ";
                break;
            case "/=":
                operator = " / ";
                break;
            case "*=":
                operator = " * ";
                break;
            case "&&=":
                operator = " && ";
                break;
            default:
                return undefined;
        }
        const target = (_a = high.findDirectExpression(Expressions.Target)) === null || _a === void 0 ? void 0 : _a.concatTokens();
        if (target === undefined) {
            return;
        }
        const sourceStart = (_c = (_b = high.findDirectExpression(Expressions.Source)) === null || _b === void 0 ? void 0 : _b.getFirstChild()) === null || _c === void 0 ? void 0 : _c.getFirstToken().getStart();
        if (sourceStart === undefined) {
            return;
        }
        const fix = edit_helper_1.EditHelper.replaceRange(lowFile, op.getStart(), sourceStart, "= " + target + operator);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Expand operator", this.getMetadata().key, this.conf.severity, fix);
    }
    stringTemplateAlpha(low, high, lowFile, highSyntax) {
        var _a, _b;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const child of high.findAllExpressionsRecursive(Expressions.StringTemplateSource)) {
            const formatting = (_a = child.findDirectExpression(Expressions.StringTemplateFormatting)) === null || _a === void 0 ? void 0 : _a.concatTokens().toUpperCase();
            if (formatting === undefined
                || (formatting === null || formatting === void 0 ? void 0 : formatting.startsWith("ALPHA = ")) === false) {
                continue;
            }
            let functionName = "";
            switch (formatting) {
                case "ALPHA = IN":
                    functionName = "CONVERSION_EXIT_ALPHA_INPUT";
                    break;
                case "ALPHA = OUT":
                    functionName = "CONVERSION_EXIT_ALPHA_OUTPUT";
                    break;
                default:
                    return undefined;
            }
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const source = (_b = child.findDirectExpression(Expressions.Source)) === null || _b === void 0 ? void 0 : _b.concatTokens();
            const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const code = `DATA ${uniqueName} TYPE string.
${indentation}CALL FUNCTION '${functionName}'
${indentation}  EXPORTING
${indentation}    input  = ${source}
${indentation}  IMPORTING
${indentation}    output = ${uniqueName}.\n`;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), code);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, child.getFirstToken().getStart(), child.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Downport ALPHA", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineLoopInput(low, high, lowFile, highSyntax) {
        var _a, _b;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        else if (!(high.get() instanceof Statements.Loop)) {
            return undefined;
        }
        else if ((_a = high.findDirectExpression(Expressions.LoopSource)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.SimpleSource2)) {
            return undefined;
        }
        // the first Source must be outlined
        const s = (_b = high.findDirectExpression(Expressions.LoopSource)) === null || _b === void 0 ? void 0 : _b.findDirectExpression(Expressions.Source);
        if (s === undefined) {
            return undefined;
        }
        const uniqueName = this.uniqueName(high.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
        const code = `DATA(${uniqueName}) = ${s.concatTokens()}.\n` +
            " ".repeat(high.getFirstToken().getStart().getCol() - 1);
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), code);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, s.getFirstToken().getStart(), s.getLastToken().getEnd(), uniqueName);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Outline LOOP input", this.getMetadata().key, this.conf.severity, fix);
    }
    outlineLoopTarget(node, lowFile, highSyntax) {
        var _a, _b, _c, _d, _e, _f, _g;
        // also allows outlining of voided types
        if (!(node.get() instanceof Statements.Loop)) {
            return undefined;
        }
        if (this.lowReg.getConfig().getVersion() === version_1.Version.OpenABAP) {
            return undefined;
        }
        const source = (_a = node.findDirectExpression(Expressions.LoopSource)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.SimpleSource2);
        if (source === undefined) {
            return undefined;
        }
        const sourceName = source === null || source === void 0 ? void 0 : source.concatTokens();
        if (sourceName === undefined) {
            return undefined;
        }
        let foundType = undefined;
        const spag = highSyntax.spaghetti.lookupPosition(source.getFirstToken().getStart(), lowFile.getFilename());
        if (spag) {
            foundType = spag.findVariable(source.concatTokens());
        }
        const concat = node.concatTokens().toUpperCase();
        if (concat.includes(" GROUP BY ") || concat.startsWith("LOOP AT GROUP ")) {
            return undefined;
        }
        const isReference = concat.includes(" REFERENCE INTO ");
        const indentation = " ".repeat(node.getFirstToken().getStart().getCol() - 1);
        const dataTarget = (_c = (_b = node.findDirectExpression(Expressions.LoopTarget)) === null || _b === void 0 ? void 0 : _b.findDirectExpression(Expressions.Target)) === null || _c === void 0 ? void 0 : _c.findDirectExpression(Expressions.InlineData);
        if (dataTarget) {
            const targetName = ((_d = dataTarget.findDirectExpression(Expressions.TargetField)) === null || _d === void 0 ? void 0 : _d.concatTokens()) || "DOWNPORT_ERROR";
            let code = `DATA ${targetName} LIKE LINE OF ${sourceName}.\n${indentation}`;
            if (isReference) {
                const likeName = this.uniqueName(node.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
                code = `DATA ${likeName} LIKE LINE OF ${sourceName}.\n${indentation}DATA ${targetName} LIKE REF TO ${likeName}.\n${indentation}`;
            }
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, node.getFirstToken().getStart(), code);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, dataTarget.getFirstToken().getStart(), dataTarget.getLastToken().getEnd(), targetName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, node.getFirstToken(), "Outline LOOP data target", this.getMetadata().key, this.conf.severity, fix);
        }
        const fsTarget = (_f = (_e = node.findDirectExpression(Expressions.LoopTarget)) === null || _e === void 0 ? void 0 : _e.findDirectExpression(Expressions.FSTarget)) === null || _f === void 0 ? void 0 : _f.findDirectExpression(Expressions.InlineFS);
        if (fsTarget) {
            const targetName = ((_g = fsTarget.findDirectExpression(Expressions.TargetFieldSymbol)) === null || _g === void 0 ? void 0 : _g.concatTokens()) || "DOWNPORT_ERROR";
            let type = `LIKE LINE OF ${sourceName}`;
            const f = foundType === null || foundType === void 0 ? void 0 : foundType.getType();
            if (f instanceof basic_1.TableType && f.getRowType() instanceof basic_1.AnyType) {
                type = "TYPE ANY";
            }
            const code = `FIELD-SYMBOLS ${targetName} ${type}.\n${indentation}`;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, node.getFirstToken().getStart(), code);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, fsTarget.getFirstToken().getStart(), fsTarget.getLastToken().getEnd(), targetName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, node.getFirstToken(), "Outline LOOP fs target", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineFor(forLoop, indentation, lowFile, highSyntax) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        let body = "";
        let end = "";
        const loopSource = (_a = forLoop.findFirstExpression(Expressions.Source)) === null || _a === void 0 ? void 0 : _a.concatTokens();
        let loopTargetFieldExpression = forLoop.findFirstExpression(Expressions.TargetField);
        let loopTargetFieldName = loopTargetFieldExpression === null || loopTargetFieldExpression === void 0 ? void 0 : loopTargetFieldExpression.concatTokens();
        const of = forLoop.findExpressionAfterToken("OF");
        if (of !== undefined) {
            loopTargetFieldExpression = of;
            loopTargetFieldName = of === null || of === void 0 ? void 0 : of.concatTokens();
        }
        if (((_c = (_b = forLoop.findDirectExpression(Expressions.InlineLoopDefinition)) === null || _b === void 0 ? void 0 : _b.getFirstChild()) === null || _c === void 0 ? void 0 : _c.get()) instanceof Expressions.TargetFieldSymbol) {
            loopTargetFieldExpression = undefined;
            loopTargetFieldName = undefined;
        }
        if (loopTargetFieldExpression) {
            const start = loopTargetFieldExpression.getFirstToken().getStart();
            const spag = highSyntax.spaghetti.lookupPosition(start, lowFile.getFilename());
            if (loopTargetFieldName && spag) {
                if (new SpagHelper(spag).isDuplicateName(loopTargetFieldName, start)) {
                    this.renameVariable(spag, loopTargetFieldName, start, lowFile, highSyntax);
                }
            }
        }
        let cond = ((_d = forLoop.findDirectExpression(Expressions.ComponentCond)) === null || _d === void 0 ? void 0 : _d.concatTokens()) || "";
        if (cond !== "") {
            cond = " WHERE " + cond;
        }
        const loop = forLoop.findDirectExpression(Expressions.InlineLoopDefinition);
        const indexInto = (_e = loop === null || loop === void 0 ? void 0 : loop.findExpressionAfterToken("INTO")) === null || _e === void 0 ? void 0 : _e.concatTokens();
        if (forLoop.findDirectTokenByText("UNTIL")
            || forLoop.findDirectTokenByText("WHILE")) {
            const fieldDef = forLoop.findDirectExpression(Expressions.InlineFieldDefinition);
            const field = (_f = fieldDef === null || fieldDef === void 0 ? void 0 : fieldDef.findFirstExpression(Expressions.Field)) === null || _f === void 0 ? void 0 : _f.concatTokens();
            const indexBackup = this.uniqueName(forLoop.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            body += indentation + "DATA " + field + " TYPE i.\n";
            const second = fieldDef === null || fieldDef === void 0 ? void 0 : fieldDef.getChildren()[2];
            if ((second === null || second === void 0 ? void 0 : second.get()) instanceof Expressions.Source) {
                body += indentation + field + " = " + second.concatTokens() + ".\n";
            }
            const not = forLoop.findDirectTokenByText("UNTIL") ? " NOT" : "";
            const cond = forLoop.findFirstExpression(Expressions.Cond);
            body += indentation + `DATA ${indexBackup} LIKE sy-index.\n`;
            body += indentation + `${indexBackup} = sy-index.\n`;
            body += indentation + `WHILE${not} ${cond === null || cond === void 0 ? void 0 : cond.concatTokens()}.\n`;
            body += indentation + `  sy-index = ${indexBackup}.\n`;
            const then = forLoop.findExpressionAfterToken("THEN");
            if (then) {
                end += `  ${field} = ${then.concatTokens()}.\n`;
            }
            else {
                end += `  ${field} = ${field} + 1.\n`;
            }
            end += indentation + "ENDWHILE";
        }
        else if (loopTargetFieldName !== undefined) {
            let from = (_g = forLoop.findExpressionAfterToken("FROM")) === null || _g === void 0 ? void 0 : _g.concatTokens();
            from = from ? " FROM " + from : "";
            let to = (_h = forLoop.findExpressionAfterToken("TO")) === null || _h === void 0 ? void 0 : _h.concatTokens();
            to = to ? " TO " + to : "";
            let gby = "";
            for (const lg of forLoop.findDirectExpressions(Expressions.LoopGroupByComponent)) {
                if (gby !== "") {
                    gby += " ";
                }
                gby += lg.concatTokens();
            }
            if (gby !== "") {
                gby = " GROUP BY ( " + gby + " )";
            }
            const fc = forLoop.findDirectExpression(Expressions.FieldChain);
            if (fc) {
                gby = " GROUP BY " + fc.concatTokens();
            }
            if (forLoop.findDirectTokenByText("ASCENDING")) {
                gby += " ASCENDING";
            }
            if (forLoop.findDirectTokenByText("DESCENDING")) {
                gby += " DESCENDING";
            }
            const groups = forLoop.findExpressionAfterToken("GROUPS");
            if (groups) {
                const concat = groups.concatTokens();
                if (concat.startsWith("<")) {
                    gby += " ASSIGNING FIELD-SYMBOL(" + concat + ")";
                }
                else {
                    gby += " INTO DATA(" + concat + ")";
                }
            }
            let inGroup = "";
            if (forLoop.concatTokens().toUpperCase().includes(" IN GROUP ")) {
                inGroup = "-items";
            }
            let into = "INTO DATA";
            if (loopTargetFieldName.startsWith("<")) {
                into = "ASSIGNING FIELD-SYMBOL";
            }
            // todo, also backup sy-index / sy-tabix here?
            body += indentation + `LOOP AT ${loopSource}${inGroup} ${into}(${loopTargetFieldName})${from}${to}${cond}${gby}.\n`;
            if (indexInto) {
                body += indentation + "  DATA(" + indexInto + ") = sy-tabix.\n";
            }
            end = "ENDLOOP";
        }
        else if (loopTargetFieldName === undefined) {
            // todo, also backup sy-index / sy-tabix here?
            const loopTargetFieldSymbol = (_j = forLoop.findFirstExpression(Expressions.TargetFieldSymbol)) === null || _j === void 0 ? void 0 : _j.concatTokens();
            body += indentation + `LOOP AT ${loopSource} ASSIGNING FIELD-SYMBOL(${loopTargetFieldSymbol})${cond}.\n`;
            if (indexInto) {
                body += indentation + "  DATA(" + indexInto + ") = sy-tabix.\n";
            }
            end = "ENDLOOP";
        }
        const l = forLoop.findDirectExpression(Expressions.Let);
        if (l) {
            body += this.outlineLet(l, indentation, highSyntax, lowFile);
        }
        return { body, end };
    }
    outlineFilter(low, high, lowFile, highSyntax) {
        var _a;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const i of high.findAllExpressionsRecursive(Expressions.Source)) {
            const firstToken = i.getFirstToken();
            if (firstToken.getStr().toUpperCase() !== "FILTER") {
                continue;
            }
            const filterBody = i.findDirectExpression(Expressions.FilterBody);
            if (filterBody === undefined) {
                continue;
            }
            const sourceName = (_a = filterBody.findDirectExpression(Expressions.Source)) === null || _a === void 0 ? void 0 : _a.concatTokens();
            if (sourceName === undefined) {
                continue;
            }
            let type = this.findType(i, lowFile, highSyntax);
            if (type === undefined) {
                type = "LIKE " + sourceName;
            }
            else {
                type = "TYPE " + type;
            }
            const uniqueName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
            const loopName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            let body = "";
            body += `DATA ${uniqueName} ${type}.\n`;
            body += `${indentation}LOOP AT ${sourceName} INTO DATA(${loopName}) ${filterBody.concatTokens().substring(sourceName.length + 1)}.\n`;
            body += `${indentation}  INSERT ${loopName} INTO TABLE ${uniqueName}.\n`;
            body += `${indentation}ENDLOOP.\n${indentation}`;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), body);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, firstToken.getStart(), i.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, firstToken, "Downport FILTER", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineSwitch(low, high, lowFile, highSyntax) {
        var _a, _b, _c, _d;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const i of high.findAllExpressionsRecursive(Expressions.Source)) {
            const firstToken = i.getFirstToken();
            if (firstToken.getStr().toUpperCase() !== "SWITCH") {
                continue;
            }
            let type = this.findType(i, lowFile, highSyntax);
            if (type === undefined) {
                if (high.get() instanceof Statements.Move
                    && high.findDirectExpression(Expressions.Source) === i
                    && ((_a = high.findDirectExpression(Expressions.Target)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.TargetField)) !== undefined) {
                    type = "LIKE " + ((_b = high.findDirectExpression(Expressions.Target)) === null || _b === void 0 ? void 0 : _b.concatTokens());
                }
                if (type === undefined) {
                    continue;
                }
            }
            else {
                type = "TYPE " + type;
            }
            const uniqueName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            let body = "";
            let name = "";
            const switchBody = i.findDirectExpression(Expressions.SwitchBody);
            if (switchBody === undefined) {
                continue;
            }
            for (const l of ((_c = switchBody === null || switchBody === void 0 ? void 0 : switchBody.findDirectExpression(Expressions.Let)) === null || _c === void 0 ? void 0 : _c.findDirectExpressions(Expressions.InlineFieldDefinition)) || []) {
                name = l.getFirstToken().getStr();
                body += indentation + `DATA(${name}) = ${(_d = switchBody.findFirstExpression(Expressions.Source)) === null || _d === void 0 ? void 0 : _d.concatTokens()}.\n`;
            }
            body += `DATA ${uniqueName} ${type}.\n`;
            let firstSource = false;
            let inWhen = false;
            for (const c of switchBody.getChildren()) {
                if (c.get() instanceof Expressions.Source && firstSource === false) {
                    body += indentation + `CASE ${c.concatTokens()}.`;
                    firstSource = true;
                }
                else if (c instanceof nodes_1.TokenNode && c.concatTokens().toUpperCase() === "THEN") {
                    inWhen = true;
                    body += ".\n";
                }
                else if (c instanceof nodes_1.TokenNode && c.concatTokens().toUpperCase() === "WHEN") {
                    inWhen = false;
                    body += `\n${indentation}  WHEN `;
                }
                else if (c instanceof nodes_1.TokenNode && c.concatTokens().toUpperCase() === "OR") {
                    body += ` OR `;
                }
                else if (c instanceof nodes_1.TokenNode && c.concatTokens().toUpperCase() === "ELSE") {
                    inWhen = true;
                    body += `\n${indentation}  WHEN OTHERS.\n`;
                }
                else if (inWhen === false) {
                    body += c.concatTokens();
                }
                else {
                    body += indentation + "    " + uniqueName + " = " + c.concatTokens() + ".";
                }
            }
            body += "\n" + indentation + "ENDCASE.\n" + indentation;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), body);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, firstToken.getStart(), i.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, firstToken, "Downport SWITCH", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineReduce(low, high, lowFile, highSyntax) {
        var _a, _b;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const i of high.findAllExpressionsRecursive(Expressions.Source)) {
            const firstToken = i.getFirstToken();
            if (firstToken.getStr().toUpperCase() !== "REDUCE") {
                continue;
            }
            const type = this.findType(i, lowFile, highSyntax);
            if (type === undefined) {
                continue;
            }
            const uniqueName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            let body = "";
            let name = "";
            const reduceBody = i.findDirectExpression(Expressions.ReduceBody);
            if (reduceBody === undefined) {
                continue;
            }
            const letNode = reduceBody.findDirectExpression(Expressions.Let);
            if (letNode) {
                body += this.outlineLet(letNode, indentation, highSyntax, lowFile);
            }
            let firstName = "";
            for (const init of reduceBody.findDirectExpressions(Expressions.InlineFieldDefinition)) {
                name = init.getFirstToken().getStr();
                if (firstName === "") {
                    firstName = name;
                }
                const spag = highSyntax.spaghetti.lookupPosition(init.getFirstToken().getStart(), lowFile.getFilename());
                if (spag && new SpagHelper(spag).isDuplicateName(name, init.getFirstToken().getStart())) {
                    this.renameVariable(spag, name, init.getFirstToken().getStart(), lowFile, highSyntax);
                }
                const s = (_a = init.findFirstExpression(Expressions.Source)) === null || _a === void 0 ? void 0 : _a.concatTokens();
                const t = (_b = init.findFirstExpression(Expressions.TypeName)) === null || _b === void 0 ? void 0 : _b.concatTokens();
                if (s) {
                    if (s.toUpperCase().startsWith("VALUE #")) {
                        body += indentation + `DATA(${name}) = ${s.replace("#", type)}.\n`;
                    }
                    else {
                        body += indentation + `DATA(${name}) = ${s}.\n`;
                    }
                }
                else {
                    body += indentation + `DATA ${name} TYPE ${t}.\n`;
                }
            }
            let end = "";
            for (const forLoop of (reduceBody === null || reduceBody === void 0 ? void 0 : reduceBody.findDirectExpressions(Expressions.For)) || []) {
                const outlineFor = this.outlineFor(forLoop, indentation, lowFile, highSyntax);
                body += outlineFor.body;
                end = outlineFor.end + `.\n` + end;
            }
            const next = reduceBody.findDirectExpression(Expressions.ReduceNext);
            if (next === undefined) {
                continue;
            }
            for (const n of next.getChildren()) {
                const concat = n.concatTokens();
                if (concat.toUpperCase() === "NEXT") {
                    continue;
                }
                else if (n.get() instanceof Expressions.SimpleTarget) {
                    body += indentation + "  " + concat + " ";
                }
                else if (n.get() instanceof Expressions.Source) {
                    body += " " + concat + ".\n";
                }
                else {
                    body += concat;
                }
            }
            body += indentation + end;
            body += indentation + `${uniqueName} = ${firstName}.\n`;
            const abap = `DATA ${uniqueName} TYPE ${type}.\n` +
                body +
                indentation;
            const reduceEnd = i.findDirectTokenByText(")");
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), abap);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, firstToken.getStart(), reduceEnd.getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, firstToken, "Downport REDUCE", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineCorresponding(low, high, lowFile, highSyntax) {
        var _a;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        const allSources = high.findAllExpressionsRecursive(Expressions.Source);
        for (const s of allSources) {
            const firstToken = s.getFirstToken();
            if (firstToken.getStr().toUpperCase() !== "CORRESPONDING") {
                continue;
            }
            const correspondingBody = s.findDirectExpression(Expressions.CorrespondingBody);
            if ((correspondingBody === null || correspondingBody === void 0 ? void 0 : correspondingBody.getFirstToken().getStr().toUpperCase()) === "BASE") {
                continue;
            }
            const uniqueName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            let type = this.findType(s, lowFile, highSyntax);
            if (type === undefined) {
                if (high.get() instanceof Statements.Move && high.findDirectExpression(Expressions.Source) === s) {
                    type = "LIKE " + ((_a = high.findDirectExpression(Expressions.Target)) === null || _a === void 0 ? void 0 : _a.concatTokens());
                }
                if (type === undefined) {
                    continue;
                }
            }
            else {
                type = "TYPE " + type;
            }
            const abap = `DATA ${uniqueName} ${type}.\n` +
                indentation + `CLEAR ${uniqueName}.\n` + // might be called inside a loop
                indentation + `MOVE-CORRESPONDING ${correspondingBody === null || correspondingBody === void 0 ? void 0 : correspondingBody.concatTokens()} TO ${uniqueName}.\n` +
                indentation;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), abap);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, firstToken.getStart(), s.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, firstToken, "Downport CORRESPONDING", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineValue(low, high, lowFile, highSyntax) {
        var _a, _b, _c, _d;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        const allSources = high.findAllExpressionsRecursive(Expressions.Source);
        for (const s of allSources) {
            const firstToken = s.getFirstToken();
            if (firstToken.getStr().toUpperCase() !== "VALUE") {
                continue;
            }
            const valueBody = s.findDirectExpression(Expressions.ValueBody);
            let type = this.findType(s, lowFile, highSyntax);
            if (type === undefined) {
                if (high.get() instanceof Statements.Move && high.findDirectExpression(Expressions.Source) === s) {
                    type = "LIKE " + ((_a = high.findDirectExpression(Expressions.Target)) === null || _a === void 0 ? void 0 : _a.concatTokens());
                }
                if (type === undefined) {
                    continue;
                }
            }
            else {
                type = "TYPE " + type;
            }
            const uniqueName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
            let indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            let body = "";
            const base = (_b = valueBody === null || valueBody === void 0 ? void 0 : valueBody.findDirectExpression(Expressions.ValueBase)) === null || _b === void 0 ? void 0 : _b.findExpressionAfterToken("BASE");
            if (base) {
                body += indentation + uniqueName + " = " + base.concatTokens() + ".\n";
            }
            let end = "";
            let structureName = uniqueName;
            let added = false;
            let data = "";
            let previous = undefined;
            if ((valueBody === null || valueBody === void 0 ? void 0 : valueBody.findDirectExpression(Expressions.ValueBodyLine)) !== undefined) {
                structureName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
                const extra = (valueBody === null || valueBody === void 0 ? void 0 : valueBody.findDirectExpression(Expressions.For)) ? "  " : "";
                data = indentation + extra + `DATA ${structureName} LIKE LINE OF ${uniqueName}.\n`;
            }
            for (const a of (valueBody === null || valueBody === void 0 ? void 0 : valueBody.getChildren()) || []) {
                if (a.get() instanceof Expressions.FieldAssignment) {
                    if (added === false) {
                        body += data;
                        added = true;
                    }
                    body += indentation + structureName + "-" + a.concatTokens() + ".\n";
                }
                else if (a instanceof nodes_1.ExpressionNode && a.get() instanceof Expressions.For) {
                    const outlineFor = this.outlineFor(a, indentation, lowFile, highSyntax);
                    body += outlineFor.body;
                    end = outlineFor.end + `.\n` + end;
                    indentation += "  ";
                }
                else if (a instanceof nodes_1.ExpressionNode && a.get() instanceof Expressions.Source) {
                    // special handling for superflous value expression
                    if ((valueBody === null || valueBody === void 0 ? void 0 : valueBody.getChildren().length) === 1) {
                        body += indentation + uniqueName + " = " + a.concatTokens() + `.\n`;
                    }
                }
                else if (a instanceof nodes_1.ExpressionNode && a.get() instanceof Expressions.Let) {
                    body += this.outlineLet(a, indentation, highSyntax, lowFile);
                }
                if (a instanceof nodes_1.ExpressionNode && a.get() instanceof Expressions.ValueBodyLine) {
                    let skip = false;
                    for (const b of (a === null || a === void 0 ? void 0 : a.getChildren()) || []) {
                        if (b.get() instanceof Expressions.FieldAssignment) {
                            if (added === false) {
                                body += data;
                                added = true;
                            }
                            body += indentation + structureName + "-" + b.concatTokensWithLinebreaks() + ".\n";
                        }
                        else if (b.get() instanceof Expressions.Source) {
                            // note: it wont work with APPEND for Hashed/Sorted Tables, so use INSERT,
                            body += indentation + "INSERT " + b.concatTokens() + ` INTO TABLE ${uniqueName}.\n`;
                            skip = true;
                        }
                        else if (b.get() instanceof Expressions.ValueBodyLines) {
                            body += indentation + "INSERT " + b.concatTokens() + ` INTO TABLE ${uniqueName}.\n`;
                            skip = true;
                        }
                        else if (b.concatTokens() === ")") {
                            if (added === false && (previous === null || previous === void 0 ? void 0 : previous.concatTokens()) === "(") {
                                body += data;
                                added = true;
                            }
                            if (skip === false) {
                                body += indentation + `INSERT ${structureName} INTO TABLE ${uniqueName}.\n`;
                            }
                        }
                        previous = b;
                    }
                }
            }
            if (body === "" && ((_c = valueBody === null || valueBody === void 0 ? void 0 : valueBody.getLastChild()) === null || _c === void 0 ? void 0 : _c.getFirstToken().getStr().toUpperCase()) === "OPTIONAL") {
                const fieldChain = valueBody.findFirstExpression(Expressions.FieldChain);
                const rowName = this.uniqueName(firstToken.getStart(), lowFile.getFilename(), highSyntax);
                let tableExpression = undefined;
                let tabName = "";
                let after = "";
                for (const c of (fieldChain === null || fieldChain === void 0 ? void 0 : fieldChain.getChildren()) || []) {
                    if (c.get() instanceof Expressions.TableExpression && c instanceof nodes_1.ExpressionNode) {
                        tableExpression = c;
                    }
                    else if (tableExpression === undefined) {
                        tabName += c.concatTokens();
                    }
                    else {
                        after += c.concatTokens();
                    }
                }
                let condition = "";
                if ((tableExpression === null || tableExpression === void 0 ? void 0 : tableExpression.getChildren().length) === 3) {
                    condition = "INDEX " + ((_d = tableExpression === null || tableExpression === void 0 ? void 0 : tableExpression.findDirectExpression(Expressions.Source)) === null || _d === void 0 ? void 0 : _d.concatTokens());
                }
                else {
                    condition = "WITH KEY " + (tableExpression === null || tableExpression === void 0 ? void 0 : tableExpression.concatTokens().replace("[ ", "").replace(" ]", ""));
                }
                body +=
                    indentation + `READ TABLE ${tabName} INTO DATA(${rowName}) ${condition}.\n` +
                        indentation + `IF sy-subrc = 0.\n` +
                        indentation + `  ${uniqueName} = ${rowName}${after}.\n` +
                        indentation + `ENDIF.\n`;
                if (type.includes("LIKE DATA(")) {
                    type = `LIKE LINE OF ${tabName}`;
                }
            }
            if (end !== "") {
                indentation = indentation.substring(2);
                body += indentation + end;
            }
            const abap = `DATA ${uniqueName} ${type}.\n` +
                indentation + `CLEAR ${uniqueName}.\n` + // might be called inside a loop
                body +
                indentation;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), abap);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, firstToken.getStart(), s.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, firstToken, "Downport VALUE", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineLet(node, indentation, highSyntax, lowFile) {
        var _a;
        let ret = "";
        for (const f of node.findDirectExpressions(Expressions.InlineFieldDefinition)) {
            const c = f.getFirstChild();
            if (c === undefined) {
                continue;
            }
            const name = c.concatTokens().toLowerCase();
            const spag = highSyntax.spaghetti.lookupPosition(c.getFirstToken().getStart(), lowFile.getFilename());
            if (spag === undefined) {
                continue;
            }
            if (new SpagHelper(spag).isDuplicateName(name, c.getFirstToken().getStart())) {
                this.renameVariable(spag, name, c.getFirstToken().getStart(), lowFile, highSyntax);
            }
            const found = spag.findVariable(name);
            if (found === undefined) {
                const source = f.findFirstExpression(Expressions.Source);
                if (source) {
                    ret += indentation + "DATA(" + name + `) = ${source.concatTokens()}.\n`;
                }
                continue;
            }
            const type = found.getType().getQualifiedName() ? (_a = found.getType().getQualifiedName()) === null || _a === void 0 ? void 0 : _a.toLowerCase() : found.getType().toABAP();
            ret += indentation + "DATA " + name + ` TYPE ${type}.\n`;
            const source = f.findFirstExpression(Expressions.Source);
            if (source) {
                ret += indentation + name + ` = ${source.concatTokens()}.\n`;
            }
        }
        return ret;
    }
    renameVariable(spag, name, pos, lowFile, highSyntax) {
        const newName = this.uniqueName(pos, lowFile.getFilename(), highSyntax);
        const fix = new SpagHelper(spag).renameVariable(name, pos, lowFile, newName);
        const issue = issue_1.Issue.atPosition(lowFile, pos, "Rename before outline", this.getMetadata().key, this.conf.severity, fix);
        throw new SkipToNextFile(issue);
    }
    findType(i, lowFile, highSyntax, ref = false) {
        var _a, _b;
        const expr = i.findDirectExpression(Expressions.TypeNameOrInfer);
        if (expr === undefined) {
            return undefined;
        }
        const firstToken = expr.getFirstToken();
        const concat = expr.concatTokens().toLowerCase();
        if (concat !== "#") {
            return ref ? "REF TO " + concat : concat;
        }
        const spag = highSyntax.spaghetti.lookupPosition(firstToken.getStart(), lowFile.getFilename());
        if (spag === undefined) {
            return undefined;
        }
        let inferred = undefined;
        for (const r of (spag === null || spag === void 0 ? void 0 : spag.getData().references) || []) {
            if (r.referenceType === _reference_1.ReferenceType.InferredType
                && r.resolved
                && r.position.getStart().equals(firstToken.getStart())
                && r.resolved instanceof _typed_identifier_1.TypedIdentifier
                && ((_a = r.extra) === null || _a === void 0 ? void 0 : _a.foundQualified) === true) {
                inferred = r.resolved;
                break;
            }
        }
        if (inferred === undefined) {
            return undefined;
        }
        if (inferred.getType() instanceof basic_1.ObjectReferenceType) {
            return inferred.getType().toABAP();
        }
        else {
            return (_b = inferred.getType().getQualifiedName()) === null || _b === void 0 ? void 0 : _b.toLowerCase();
        }
    }
    outlineFS(low, high, lowFile, highSyntax) {
        var _a, _b;
        if (!(low.get() instanceof _statement_1.Unknown)
            || (high.get() instanceof Statements.Loop)) {
            return undefined;
        }
        for (const i of high.findAllExpressionsRecursive(Expressions.InlineFS)) {
            const nameToken = (_a = i.findDirectExpression(Expressions.TargetFieldSymbol)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
            if (nameToken === undefined) {
                continue;
            }
            const name = nameToken.getStr();
            let type = "";
            if (high.concatTokens().toUpperCase().startsWith("APPEND INITIAL LINE TO ")) {
                type = "LIKE LINE OF " + ((_b = high.findFirstExpression(Expressions.Target)) === null || _b === void 0 ? void 0 : _b.concatTokens());
            }
            else {
                const spag = highSyntax.spaghetti.lookupPosition(nameToken.getStart(), lowFile.getFilename());
                if (spag === undefined) {
                    continue;
                }
                const found = spag.findVariable(name);
                if (found === undefined) {
                    continue;
                }
                else if (found.getType() instanceof basic_1.VoidType) {
                    return issue_1.Issue.atToken(lowFile, i.getFirstToken(), "Error outlining voided type", this.getMetadata().key, this.conf.severity);
                }
                type = "TYPE ";
                type += found.getType().getQualifiedName() ? found.getType().getQualifiedName().toLowerCase() : found.getType().toABAP();
            }
            const code = `FIELD-SYMBOLS ${name} ${type}.\n` +
                " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), code);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, i.getFirstToken().getStart(), i.getLastToken().getEnd(), name);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, i.getFirstToken(), "Outline FIELD-SYMBOL", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineData(node, lowFile, highSyntax) {
        // hmm, no guard here, as DATA(SDF) is valid in 702
        var _a, _b;
        if (this.lowReg.getConfig().getVersion() === version_1.Version.OpenABAP) {
            return undefined;
        }
        for (const i of node.findAllExpressionsRecursive(Expressions.InlineData)) {
            const nameToken = (_a = i.findDirectExpression(Expressions.TargetField)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
            if (nameToken === undefined) {
                continue;
            }
            const name = nameToken.getStr();
            const spag = highSyntax.spaghetti.lookupPosition(nameToken.getStart(), lowFile.getFilename());
            if (spag === undefined) {
                continue;
            }
            const found = spag.findVariable(name);
            if (found === undefined) {
                continue;
            }
            else if (found.getType() instanceof basic_1.VoidType && found.getType().getQualifiedName() === undefined) {
                continue;
            }
            else if (found.getType() instanceof basic_1.StructureType && found.getType().getQualifiedName() === undefined) {
                continue;
            }
            let type = found.getType().getQualifiedName()
                ? (_b = found.getType().getQualifiedName()) === null || _b === void 0 ? void 0 : _b.toLowerCase()
                : found.getType().toABAP();
            if (found.getType() instanceof basic_1.ObjectReferenceType) {
                type = found.getType().toABAP();
            }
            if (type === "") {
                continue;
            }
            const code = `DATA ${name} TYPE ${type}.\n` +
                " ".repeat(node.getFirstToken().getStart().getCol() - 1);
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, node.getFirstToken().getStart(), code);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, i.getFirstToken().getStart(), i.getLastToken().getEnd(), name);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, i.getFirstToken(), "Outline DATA", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    outlineCond(low, high, lowFile, highSyntax) {
        var _a, _b;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const i of high.findAllExpressionsRecursive(Expressions.Source)) {
            if (i.getFirstToken().getStr().toUpperCase() !== "COND") {
                continue;
            }
            const body = i.findDirectExpression(Expressions.CondBody);
            if (body === undefined) {
                continue;
            }
            const uniqueName = this.uniqueName(i.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            let type = this.findType(i, lowFile, highSyntax);
            if (type === undefined) {
                if (high.get() instanceof Statements.Move
                    && high.findDirectExpression(Expressions.Source) === i
                    && ((_a = high.findDirectExpression(Expressions.Target)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.TargetField)) !== undefined) {
                    type = "LIKE " + ((_b = high.findDirectExpression(Expressions.Target)) === null || _b === void 0 ? void 0 : _b.concatTokens());
                }
                if (type === undefined) {
                    continue;
                }
            }
            else {
                type = "TYPE " + type;
            }
            const indent = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const bodyCode = this.buildCondBody(body, uniqueName, indent, lowFile, highSyntax);
            const last = i.findDirectTokenByText(")");
            const abap = `DATA ${uniqueName} ${type}.\n` + bodyCode;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), abap);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, i.getFirstToken().getStart(), last.getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, i.getFirstToken(), "Downport COND", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    buildCondBody(body, uniqueName, indent, lowFile, highSyntax) {
        let code = "";
        let first = true;
        let addElse = true;
        for (const c of body.getChildren()) {
            if (c instanceof nodes_1.TokenNode) {
                switch (c.getFirstToken().getStr().toUpperCase()) {
                    case "WHEN":
                        if (first === true) {
                            code += indent + "IF ";
                            first = false;
                        }
                        else {
                            code += indent + "ELSEIF ";
                        }
                        break;
                    case "THEN":
                        code += ".\n";
                        break;
                    case "ELSE":
                        code += indent + "ELSE.\n";
                        addElse = false;
                        break;
                    default:
                        throw "buildCondBody, unexpected token";
                }
            }
            else if (c.get() instanceof Expressions.Cond) {
                code += c.concatTokens();
            }
            else if (c.get() instanceof Expressions.Let) {
                code += this.outlineLet(c, indent, highSyntax, lowFile);
            }
            else if (c.get() instanceof Expressions.Source) {
                code += indent + "  " + uniqueName + " = " + c.concatTokens() + ".\n";
            }
            else if (c.get() instanceof Expressions.Throw) {
                code += indent + "  " + c.concatTokens().replace(/THROW /i, "RAISE EXCEPTION NEW ") + ".\n";
            }
            else {
                throw "buildCondBody, unexpected expression, " + c.get().constructor.name;
            }
        }
        if (addElse) {
            // COND might be called inside a loop
            code += indent + "ELSE.\n";
            code += indent + `  CLEAR ${uniqueName}.\n`;
        }
        code += indent + "ENDIF.\n";
        code += indent;
        return code;
    }
    outlineConv(low, high, lowFile, highSyntax) {
        var _a;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const i of high.findAllExpressionsRecursive(Expressions.Source)) {
            if (i.getFirstToken().getStr().toUpperCase() !== "CONV") {
                continue;
            }
            const end = i.findDirectTokenByText(")");
            if (end === undefined) {
                continue;
            }
            const body = (_a = i.findDirectExpression(Expressions.ConvBody)) === null || _a === void 0 ? void 0 : _a.concatTokens();
            if (body === undefined) {
                continue;
            }
            const uniqueName = this.uniqueName(i.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const type = this.findType(i, lowFile, highSyntax);
            const indent = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const abap = `DATA ${uniqueName} TYPE ${type}.\n` +
                indent + `${uniqueName} = ${body}.\n` +
                indent;
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), abap);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, i.getFirstToken().getStart(), end.getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, i.getFirstToken(), "Downport CONV", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    // "CAST" to "?="
    outlineCast(low, high, lowFile, highSyntax) {
        var _a;
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const i of high.findAllExpressionsRecursive(Expressions.Cast)) {
            const uniqueName = this.uniqueName(i.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const type = this.findType(i, lowFile, highSyntax, true);
            const body = (_a = i.findDirectExpression(Expressions.Source)) === null || _a === void 0 ? void 0 : _a.concatTokens();
            const abap = `DATA ${uniqueName} TYPE ${type}.\n` +
                " ".repeat(high.getFirstToken().getStart().getCol() - 1) +
                `${uniqueName} ?= ${body}.\n` +
                " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), abap);
            const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, i.getFirstToken().getStart(), i.getLastToken().getEnd(), uniqueName);
            const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            return issue_1.Issue.atToken(lowFile, i.getFirstToken(), "Downport CAST", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    uniqueName(position, filename, highSyntax) {
        const spag = highSyntax.spaghetti.lookupPosition(position, filename);
        if (spag === undefined) {
            const name = "temprr" + this.counter;
            this.counter++;
            return name;
        }
        let postfix = "";
        if (spag.getIdentifier().stype === _scope_type_1.ScopeType.ClassDefinition) {
            // try making sure this name is not used in subclasses
            const hash = crypto.createHash("sha1").update(spag.getIdentifier().sname).digest("hex");
            postfix = "_" + hash.substring(0, 10);
        }
        while (true) {
            const name = "temp" + this.counter + postfix;
            const exists = this.existsRecursive(spag, name);
            this.counter++;
            if (exists === false) {
                return name;
            }
        }
    }
    // todo, optimize, the findVariable() and findType() does a lot of redundant checks
    existsRecursive(spag, name) {
        const existsDirect = spag.findVariable(name) || spag.findType(name);
        if (existsDirect) {
            return true;
        }
        for (const child of spag.getChildren()) {
            if (child.findVariable(name) || child.findType(name) || this.existsRecursive(child, name)) {
                return true;
            }
        }
        return false;
    }
    replaceXsdBool(node, lowFile, highSyntax) {
        if (this.lowReg.getConfig().getVersion() === version_1.Version.OpenABAP) {
            return undefined;
        }
        const spag = highSyntax.spaghetti.lookupPosition(node.getFirstToken().getStart(), lowFile.getFilename());
        for (const r of (spag === null || spag === void 0 ? void 0 : spag.getData().references) || []) {
            if (r.referenceType === _reference_1.ReferenceType.BuiltinMethodReference
                && r.position.getName().toUpperCase() === "XSDBOOL") {
                const token = r.position.getToken();
                let source = undefined;
                for (const s of node.findAllExpressionsRecursive(Expressions.Source)) {
                    if (s.getFirstToken().getStart().equals(token.getStart())) {
                        source = s;
                        break;
                    }
                }
                const children = source === null || source === void 0 ? void 0 : source.getChildren();
                if (source === undefined || (children === null || children === void 0 ? void 0 : children.length) !== 4) {
                    continue;
                }
                // make sure to convert to the correct type, RTTI might be used on the result of XSDBOOL
                const code = "CONV xsdboolean( boolc( " + children[2].concatTokens() + " ) )";
                const fix = edit_helper_1.EditHelper.replaceRange(lowFile, source.getFirstToken().getStart(), source.getLastToken().getEnd(), code);
                return issue_1.Issue.atToken(lowFile, token, "Use BOOLC", this.getMetadata().key, this.conf.severity, fix);
            }
        }
        return undefined;
    }
    findMethodCallExpression(node, token) {
        var _a;
        for (const m of node.findAllExpressions(Expressions.MethodCall)) {
            if ((_a = m.findDirectExpression(Expressions.MethodName)) === null || _a === void 0 ? void 0 : _a.getFirstToken().getStart().equals(token.getStart())) {
                return m;
            }
        }
        return undefined;
    }
    replaceMethodConditional(low, high, lowFile, highSyntax) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        for (const c of high.findAllExpressionsRecursive(Expressions.Compare)) {
            const chain = c.findDirectExpression(Expressions.MethodCallChain);
            if (chain === undefined) {
                continue;
            }
            const concat = chain.concatTokens().toUpperCase();
            if (concat.startsWith("LINE_EXISTS( ") || concat.startsWith("LINE_INDEX( ")) {
                // these are handled separately
                continue;
            }
            let predicate = false;
            const spag = highSyntax.spaghetti.lookupPosition(high.getFirstToken().getStart(), lowFile.getFilename());
            for (const r of (spag === null || spag === void 0 ? void 0 : spag.getData().references) || []) {
                if (r.referenceType === _reference_1.ReferenceType.BuiltinMethodReference &&
                    _builtin_1.BuiltIn.isPredicate(chain.getFirstToken().getStr().toUpperCase())) {
                    predicate = true;
                    break;
                }
            }
            const end = chain.getLastToken().getEnd();
            let fix = edit_helper_1.EditHelper.insertAt(lowFile, end, " IS NOT INITIAL");
            if (predicate === true) {
                fix = edit_helper_1.EditHelper.insertAt(lowFile, end, " ) = abap_true");
                const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, chain.getFirstToken().getStart(), "boolc( ");
                fix = edit_helper_1.EditHelper.merge(fix, fix1);
            }
            return issue_1.Issue.atToken(lowFile, chain.getFirstToken(), "Downport method conditional", this.getMetadata().key, this.conf.severity, fix);
        }
        return undefined;
    }
    getReference(node, lowFile, _highSyntax) {
        var _a, _b, _c;
        if (!(node.get() instanceof Statements.GetReference)) {
            return undefined;
        }
        const inline = (_a = node.findDirectExpression(Expressions.Target)) === null || _a === void 0 ? void 0 : _a.findDirectExpression(Expressions.InlineData);
        if (inline === undefined) {
            return undefined;
        }
        const targetName = (_b = inline.findDirectExpression(Expressions.TargetField)) === null || _b === void 0 ? void 0 : _b.concatTokens();
        const sourceName = (_c = node.findDirectExpression(Expressions.Source)) === null || _c === void 0 ? void 0 : _c.concatTokens();
        if (targetName === undefined || sourceName === undefined) {
            return undefined;
        }
        const code = `DATA ${targetName} LIKE REF TO ${sourceName}.\n`;
        const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, node.getFirstToken().getStart(), code);
        const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, inline.getFirstToken().getStart(), inline.getLastToken().getEnd(), targetName);
        const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
        return issue_1.Issue.atToken(lowFile, inline.getFirstToken(), "Downport, outline DATA ref", this.getMetadata().key, this.conf.severity, fix);
    }
    replaceContains(node, lowFile, highSyntax) {
        const spag = highSyntax.spaghetti.lookupPosition(node.getFirstToken().getStart(), lowFile.getFilename());
        // only downport if its an single method call condition
        let found = false;
        for (const c of node.findAllExpressionsRecursive(Expressions.Compare)) {
            found = c.findDirectExpression(Expressions.MethodCallChain) !== undefined;
            if (found === true) {
                break;
            }
        }
        if (found === false) {
            return undefined;
        }
        for (const r of (spag === null || spag === void 0 ? void 0 : spag.getData().references) || []) {
            if (r.referenceType !== _reference_1.ReferenceType.BuiltinMethodReference) {
                continue;
            }
            const func = r.position.getName().toUpperCase();
            if (func === "CONTAINS") {
                const token = r.position.getToken();
                const expression = this.findMethodCallExpression(node, token);
                if (expression === undefined) {
                    continue;
                }
                const sList = expression.findAllExpressions(Expressions.Source).map(e => e.concatTokens());
                if (sList.length !== 2) {
                    continue;
                }
                const code = sList[0] + " CS " + sList[1];
                const start = expression.getFirstToken().getStart();
                const end = expression.getLastToken().getEnd();
                const fix = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, code);
                return issue_1.Issue.atToken(lowFile, token, "Downport contains()", this.getMetadata().key, this.conf.severity, fix);
            }
        }
        return undefined;
    }
    replaceLineFunctions(node, lowFile, highSyntax, highFile) {
        var _a, _b;
        if (this.lowReg.getConfig().getVersion() === version_1.Version.OpenABAP) {
            return undefined;
        }
        const spag = highSyntax.spaghetti.lookupPosition(node.getFirstToken().getStart(), lowFile.getFilename());
        for (const r of (spag === null || spag === void 0 ? void 0 : spag.getData().references) || []) {
            if (r.referenceType !== _reference_1.ReferenceType.BuiltinMethodReference) {
                continue;
            }
            const func = r.position.getName().toUpperCase();
            if (func === "LINE_EXISTS" || func === "LINE_INDEX") {
                const token = r.position.getToken();
                const expression = this.findMethodCallExpression(node, token);
                if (expression === undefined) {
                    continue;
                }
                let condition = "";
                for (const c of ((_a = expression === null || expression === void 0 ? void 0 : expression.findFirstExpression(Expressions.TableExpression)) === null || _a === void 0 ? void 0 : _a.getChildren()) || []) {
                    if (c.getFirstToken().getStr() === "[" || c.getFirstToken().getStr() === "]") {
                        continue;
                    }
                    else if (c.get() instanceof Expressions.ComponentChainSimple && condition === "") {
                        condition = "WITH KEY ";
                    }
                    else if (c.get() instanceof Expressions.Source && condition === "") {
                        condition = "INDEX ";
                    }
                    condition += c.concatTokens() + " ";
                }
                const tableName = (_b = expression.findFirstExpression(Expressions.Source)) === null || _b === void 0 ? void 0 : _b.concatTokens().split("[")[0];
                const uniqueName = this.uniqueName(node.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
                const indentation = " ".repeat(node.getFirstToken().getStart().getCol() - 1);
                const sy = func === "LINE_EXISTS" ? "sy-subrc" : "sy-tabix";
                const code = `DATA ${uniqueName} LIKE sy-subrc.\n` +
                    indentation + `READ TABLE ${tableName} ${condition}TRANSPORTING NO FIELDS.\n` +
                    indentation + uniqueName + ` = ${sy}.\n` +
                    indentation;
                let insertAt = node.getFirstToken().getStart();
                if (node.get() instanceof statements_1.ElseIf) {
                    // assumption: no side effects in IF conditions
                    insertAt = this.findStartOfIf(node, highFile);
                    if (insertAt === undefined) {
                        continue;
                    }
                }
                const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, insertAt, code);
                const start = expression.getFirstToken().getStart();
                const end = expression.getLastToken().getEnd();
                const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, start, end, uniqueName + (func === "LINE_EXISTS" ? " = 0" : ""));
                const fix = edit_helper_1.EditHelper.merge(fix2, fix1);
                return issue_1.Issue.atToken(lowFile, token, "Replace line function", this.getMetadata().key, this.conf.severity, fix);
            }
        }
        return undefined;
    }
    findStartOfIf(node, highFile) {
        const structure = highFile.getStructure();
        for (const c of (structure === null || structure === void 0 ? void 0 : structure.findAllStructuresRecursive(Structures.If)) || []) {
            for (const ei of c.findDirectStructures(Structures.ElseIf)) {
                if (ei.getFirstStatement() === node) {
                    return c.getFirstToken().getStart();
                }
            }
        }
        return undefined;
    }
    newToCreateObject(low, high, lowFile, highSyntax) {
        if (!(low.get() instanceof _statement_1.Unknown)) {
            return undefined;
        }
        const source = high.findDirectExpression(Expressions.Source);
        let fix = undefined;
        if (high.get() instanceof Statements.Move
            && source
            && source.getFirstToken().getStr().toUpperCase() === "NEW") {
            const target = high.findDirectExpression(Expressions.Target);
            const found = source === null || source === void 0 ? void 0 : source.findFirstExpression(Expressions.NewObject);
            // must be at top level of the source for quickfix to work(todo: handle more scenarios)
            if (target
                && found
                && source.concatTokens() === found.concatTokens()
                && target.findDirectExpression(Expressions.InlineData) === undefined) {
                const abap = this.newParameters(found, target.concatTokens(), highSyntax, lowFile);
                if (abap !== undefined) {
                    fix = edit_helper_1.EditHelper.replaceRange(lowFile, high.getFirstToken().getStart(), high.getLastToken().getEnd(), abap);
                }
            }
        }
        if (fix === undefined && high.findAllExpressions(Expressions.NewObject)) {
            const found = high.findFirstExpression(Expressions.NewObject);
            if (found === undefined) {
                return undefined;
            }
            const name = this.uniqueName(found.getFirstToken().getStart(), lowFile.getFilename(), highSyntax);
            const abap = this.newParameters(found, name, highSyntax, lowFile);
            if (abap === undefined) {
                return undefined;
            }
            const type = this.findType(found, lowFile, highSyntax);
            const indentation = " ".repeat(high.getFirstToken().getStart().getCol() - 1);
            const data = `DATA ${name} TYPE REF TO ${type}.\n` +
                indentation + abap + "\n" +
                indentation;
            if (found.getFirstToken().getStart().equals(high.getFirstToken().getStart())
                && found.getLastToken().getEnd().equals(high.getLastToken().getStart())) {
                // full statement = standalone NEW expression
                fix = edit_helper_1.EditHelper.replaceRange(lowFile, high.getFirstToken().getStart(), high.getLastToken().getEnd(), abap);
            }
            else {
                const fix1 = edit_helper_1.EditHelper.insertAt(lowFile, high.getFirstToken().getStart(), data);
                const fix2 = edit_helper_1.EditHelper.replaceRange(lowFile, found.getFirstToken().getStart(), found.getLastToken().getEnd(), name);
                fix = edit_helper_1.EditHelper.merge(fix2, fix1);
            }
        }
        if (fix) {
            return issue_1.Issue.atToken(lowFile, high.getFirstToken(), "Use CREATE OBJECT instead of NEW", this.getMetadata().key, this.conf.severity, fix);
        }
        else {
            return undefined;
        }
    }
    newParameters(found, name, highSyntax, lowFile) {
        var _a, _b, _c;
        const typeToken = (_a = found.findDirectExpression(Expressions.TypeNameOrInfer)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
        let extra = (typeToken === null || typeToken === void 0 ? void 0 : typeToken.getStr()) === "#" ? "" : " TYPE " + (typeToken === null || typeToken === void 0 ? void 0 : typeToken.getStr());
        const parameters = found.findFirstExpression(Expressions.ParameterListS);
        if (parameters) {
            extra = parameters ? extra + " EXPORTING " + parameters.concatTokens() : extra;
        }
        else if (typeToken) {
            const source = (_b = found.findDirectExpression(Expressions.Source)) === null || _b === void 0 ? void 0 : _b.concatTokens();
            if (source) {
                // find the default parameter name for the constructor
                const spag = highSyntax.spaghetti.lookupPosition(typeToken === null || typeToken === void 0 ? void 0 : typeToken.getStart(), lowFile.getFilename());
                let cdef = undefined;
                for (const r of (spag === null || spag === void 0 ? void 0 : spag.getData().references) || []) {
                    if ((r.referenceType === _reference_1.ReferenceType.InferredType
                        || r.referenceType === _reference_1.ReferenceType.ObjectOrientedReference)
                        && r.resolved && r.position.getStart().equals(typeToken.getStart())) {
                        cdef = r.resolved;
                    }
                }
                if (cdef instanceof _typed_identifier_1.TypedIdentifier) {
                    const foo = cdef.getType();
                    if (foo instanceof basic_1.ObjectReferenceType) {
                        cdef = spag === null || spag === void 0 ? void 0 : spag.findClassDefinition(foo.getQualifiedName());
                        if (cdef === undefined) {
                            const cglobal = this.highReg.getObject("CLAS", foo.getQualifiedName());
                            cdef = cglobal === null || cglobal === void 0 ? void 0 : cglobal.getDefinition();
                        }
                    }
                    else {
                        throw new Error("newParameters, downport, unexpected");
                    }
                }
                if (cdef && cdef.getMethodDefinitions === undefined) {
                    return undefined; // something wrong
                }
                const importing = (_c = this.findConstructor(cdef, spag)) === null || _c === void 0 ? void 0 : _c.getParameters().getDefaultImporting();
                if (importing) {
                    extra += " EXPORTING " + importing + " = " + source;
                }
                else if (spag === undefined) {
                    extra += " SpagUndefined ERROR";
                }
                else if (cdef === undefined) {
                    extra += " ClassDefinitionNotFound ERROR";
                }
                else {
                    extra += " SomeError ERROR";
                }
            }
        }
        const abap = `CREATE OBJECT ${name}${extra}.`;
        return abap;
    }
    findConstructor(cdef, spag) {
        let def = cdef;
        while (def !== undefined) {
            const method = def === null || def === void 0 ? void 0 : def.getMethodDefinitions().getByName("CONSTRUCTOR");
            if (method) {
                return method;
            }
            const name = def.getSuperClass();
            if (name) {
                def = spag === null || spag === void 0 ? void 0 : spag.findClassDefinition(name);
            }
            else {
                return undefined;
            }
        }
    }
}
exports.Downport = Downport;
