// @abaplint/core/build/src/rules/fm_global_parameters_obsolete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FMGlobalParametersObsolete = exports.FMGlobalParametersObsoleteConf = void 0;
const issue_1 = require("../issue");
const Objects = require("../objects");
const _basic_rule_config_1 = require("./_basic_rule_config");
const position_1 = require("../position");
class FMGlobalParametersObsoleteConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.FMGlobalParametersObsoleteConf = FMGlobalParametersObsoleteConf;
class FMGlobalParametersObsolete {
    constructor() {
        this.conf = new FMGlobalParametersObsoleteConf();
    }
    getMetadata() {
        return {
            key: "fm_global_parameters_obsolete",
            title: "FM Global Parameters Obsolete",
            shortDescription: `Check for function modules with global parameteers`,
            extendedInformation: `https://help.sap.com/doc/abapdocu_750_index_htm/7.50/en-US/abenglobal_parameters_obsolete.htm`,
            tags: [],
        };
    }
    initialize(_reg) {
        return this;
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    run(obj) {
        if (!(obj instanceof Objects.FunctionGroup)) {
            return [];
        }
        const issues = [];
        for (const module of obj.getModules()) {
            if (module.isGlobalParameters() === true) {
                const file = obj.getMainABAPFile();
                if (file === undefined) {
                    continue;
                }
                const message = `Function Module "${module.getName()}" uses obsolete global parameters`;
                issues.push(issue_1.Issue.atPosition(file, new position_1.Position(1, 1), message, this.getMetadata().key, this.conf.severity));
            }
        }
        return issues;
    }
}
exports.FMGlobalParametersObsolete = FMGlobalParametersObsolete;
