// @abaplint/core/build/src/rules/forbidden_void_type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForbiddenVoidType = exports.ForbiddenVoidTypeConf = void 0;
const _basic_rule_config_1 = require("./_basic_rule_config");
const issue_1 = require("../issue");
const _abap_object_1 = require("../objects/_abap_object");
const syntax_1 = require("../abap/5_syntax/syntax");
const void_type_1 = require("../abap/types/basic/void_type");
const _scope_type_1 = require("../abap/5_syntax/_scope_type");
const basic_1 = require("../abap/types/basic");
const _reference_1 = require("../abap/5_syntax/_reference");
class ForbiddenVoidTypeConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** List of forbidden void types, array of string regex, case in-sensitive
         * @uniqueItems true
        */
        this.check = [];
    }
}
exports.ForbiddenVoidTypeConf = ForbiddenVoidTypeConf;
class ForbiddenVoidType {
    constructor() {
        this.conf = new ForbiddenVoidTypeConf();
    }
    getMetadata() {
        return {
            key: "forbidden_void_type",
            title: "Forbidden Void Types",
            shortDescription: `Avoid usage of specified void types.`,
            extendedInformation: `Inspiration:
BOOLEAN, BOOLE_D, CHAR01, CHAR1, CHAR10, CHAR12, CHAR128, CHAR2, CHAR20, CHAR4, CHAR70,
DATS, TIMS, DATUM, FLAG, INT4, NUMC3, NUMC4, SAP_BOOL, TEXT25, TEXT80, X255, XFELD`,
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
        if (this.conf.check === undefined) {
            this.conf.check = [];
        }
    }
    initialize(reg) {
        this.reg = reg;
        return this;
    }
    run(obj) {
        if (!(obj instanceof _abap_object_1.ABAPObject) || this.conf.check.length === 0) {
            return [];
        }
        return this.traverse(new syntax_1.SyntaxLogic(this.reg, obj).run().spaghetti.getTop());
    }
    ///////////////
    traverse(node) {
        var _a, _b, _c;
        let ret = [];
        const message = "Forbidden void type: ";
        if (node.getIdentifier().stype !== _scope_type_1.ScopeType.BuiltIn) {
            for (const r of node.getData().references) {
                if (r.referenceType === _reference_1.ReferenceType.ObjectOrientedVoidReference
                    && ((_a = r.extra) === null || _a === void 0 ? void 0 : _a.ooName) !== undefined
                    && this.isForbiddenName((_b = r.extra) === null || _b === void 0 ? void 0 : _b.ooName)) {
                    ret.push(issue_1.Issue.atIdentifier(r.position, message + ((_c = r.extra) === null || _c === void 0 ? void 0 : _c.ooName), this.getMetadata().key, this.conf.severity));
                }
                if ((r.referenceType === _reference_1.ReferenceType.VoidType
                    || r.referenceType === _reference_1.ReferenceType.TableVoidReference)
                    && this.isForbiddenName(r.position.getName())) {
                    ret.push(issue_1.Issue.atIdentifier(r.position, message + r.position.getName(), this.getMetadata().key, this.conf.severity));
                }
            }
        }
        for (const c of node.getChildren()) {
            ret = ret.concat(this.traverse(c));
        }
        return ret;
    }
    isForbiddenType(type) {
        if (type instanceof basic_1.StructureType) {
            return type.getComponents().some(c => this.isForbiddenType(c.type));
        }
        else if (!(type instanceof void_type_1.VoidType)) {
            return false;
        }
        const name = type.getVoided();
        return this.isForbiddenName(name);
    }
    isForbiddenName(name) {
        if (name === undefined) {
            return false;
        }
        for (const c of this.conf.check) {
            const reg = new RegExp(c, "i");
            const match = reg.test(name);
            if (match === true) {
                return true;
            }
        }
        return false;
    }
}
exports.ForbiddenVoidType = ForbiddenVoidType;
