// @abaplint/core/build/src/rules/form_tables_obsolete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormTablesObsolete = exports.FormTablesObsoleteConf = void 0;
const issue_1 = require("../issue");
const objects_1 = require("../objects");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _abap_rule_1 = require("./_abap_rule");
const Expressions = require("../abap/2_statements/expressions");
const _irule_1 = require("./_irule");
class FormTablesObsoleteConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.FormTablesObsoleteConf = FormTablesObsoleteConf;
class FormTablesObsolete extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new FormTablesObsoleteConf();
    }
    getMetadata() {
        return {
            key: "form_tables_obsolete",
            title: "TABLES parameters are obsolete",
            shortDescription: `Checks for TABLES parameters in forms.`,
            extendedInformation: `https://help.sap.com/doc/abapdocu_751_index_htm/7.51/en-US/abapform_tables.htm`,
            tags: [_irule_1.RuleTag.SingleFile],
        };
    }
    getMessage() {
        return "FORM TABLES parameters are obsolete";
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file, obj) {
        const ret = [];
        const stru = file.getStructure();
        if (obj instanceof objects_1.Class || stru === undefined) {
            return ret;
        }
        for (const form of stru.findAllExpressions(Expressions.FormTables)) {
            const token = form.getFirstToken();
            const issue = issue_1.Issue.atToken(file, token, this.getMessage(), this.getMetadata().key, this.conf.severity);
            ret.push(issue);
        }
        return ret;
    }
}
exports.FormTablesObsolete = FormTablesObsolete;
