// @abaplint/core/build/src/rules/fully_type_itabs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FullyTypeITabs = exports.FullyTypeITabsConf = void 0;
const _basic_rule_config_1 = require("./_basic_rule_config");
const _abap_rule_1 = require("./_abap_rule");
const issue_1 = require("../issue");
const Statements = require("../abap/2_statements/statements");
const Expressions = require("../abap/2_statements/expressions");
const _irule_1 = require("./_irule");
class FullyTypeITabsConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.FullyTypeITabsConf = FullyTypeITabsConf;
class FullyTypeITabs extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new FullyTypeITabsConf();
    }
    getMetadata() {
        return {
            key: "fully_type_itabs",
            title: "Fully type internal tables",
            shortDescription: `No implict table types or table keys`,
            extendedInformation: `https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#use-the-right-table-type
https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#avoid-default-key`,
            badExample: `DATA lt_foo TYPE TABLE OF ty.
DATA lt_bar TYPE STANDARD TABLE OF ty.`,
            goodExample: `DATA lt_foo TYPE STANDARD TABLE OF ty WITH EMPTY KEY.`,
            tags: [_irule_1.RuleTag.SingleFile],
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file) {
        const issues = [];
        for (const statement of file.getStatements()) {
            if (!(statement.get() instanceof Statements.Data || statement.get() instanceof Statements.Type)) {
                continue;
            }
            const tt = statement.findFirstExpression(Expressions.TypeTable);
            if (tt === undefined) {
                continue;
            }
            const concat = tt.concatTokens().toUpperCase();
            if (concat.includes("TYPE TABLE OF")) {
                const message = "Specify table type";
                issues.push(issue_1.Issue.atPosition(file, tt.getFirstToken().getStart(), message, this.getMetadata().key, this.conf.severity));
            }
            else if (concat.includes(" TABLE FOR ")) {
                continue;
            }
            else if (concat.includes(" WITH ") === false && concat.includes(" RANGE OF ") === false) {
                const message = "Specify table key";
                issues.push(issue_1.Issue.atPosition(file, tt.getFirstToken().getStart(), message, this.getMetadata().key, this.conf.severity));
            }
        }
        return issues;
    }
}
exports.FullyTypeITabs = FullyTypeITabs;
