// @abaplint/core/build/src/rules/fully_type_constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FullyTypeConstants = exports.FullyTypeConsantsConf = void 0;
const _basic_rule_config_1 = require("./_basic_rule_config");
const _abap_rule_1 = require("./_abap_rule");
const issue_1 = require("../issue");
const Statements = require("../abap/2_statements/statements");
const expressions_1 = require("../abap/2_statements/expressions");
const _irule_1 = require("./_irule");
class FullyTypeConsantsConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** Add check for implicit data definition, require full typing. */
        this.checkData = true;
    }
}
exports.FullyTypeConsantsConf = FullyTypeConsantsConf;
class FullyTypeConstants extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new FullyTypeConsantsConf();
    }
    getMetadata() {
        return {
            key: "fully_type_constants",
            title: "Fully type constants",
            shortDescription: `Checks constants for full typing - no implicit typing allowed.`,
            badExample: "CONSTANTS foo VALUE 'a'.",
            goodExample: "CONSTANTS foo TYPE c LENGTH 1 VALUE 'a'.",
            tags: [_irule_1.RuleTag.SingleFile],
        };
    }
    getDescription(type) {
        return `Fully type ${type}, no implicit typing`;
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file) {
        var _a, _b;
        const issues = [];
        for (const stat of file.getStatements()) {
            if ((stat.get() instanceof Statements.Constant
                || (this.conf.checkData === true && stat.get() instanceof Statements.Data))
                && !this.isTyped(stat)) {
                const type = stat.get() instanceof Statements.Constant ? "constant definition" : "data definition";
                let token = (_a = stat.findFirstExpression(expressions_1.NamespaceSimpleName)) === null || _a === void 0 ? void 0 : _a.getFirstToken();
                if (token === undefined) {
                    token = (_b = stat.findFirstExpression(expressions_1.DefinitionName)) === null || _b === void 0 ? void 0 : _b.getFirstToken();
                }
                if (token === undefined) {
                    throw new Error("fully type constants, unexpected node");
                }
                issues.push(issue_1.Issue.atToken(file, token, this.getDescription(type), this.getMetadata().key, this.conf.severity));
            }
        }
        return issues;
    }
    isTyped(stat) {
        return stat.findFirstExpression(expressions_1.Type) || stat.findFirstExpression(expressions_1.TypeTable);
    }
}
exports.FullyTypeConstants = FullyTypeConstants;
