// @abaplint/core/build/src/rules/keep_single_parameter_on_one_line.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeepSingleParameterCallsOnOneLine = exports.KeepSingleParameterCallsOnOneLineConf = void 0;
const issue_1 = require("../issue");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _abap_rule_1 = require("./_abap_rule");
const Expressions = require("../abap/2_statements/expressions");
const _irule_1 = require("./_irule");
class KeepSingleParameterCallsOnOneLineConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** Max line length, in characters */
        this.length = 120;
    }
}
exports.KeepSingleParameterCallsOnOneLineConf = KeepSingleParameterCallsOnOneLineConf;
class KeepSingleParameterCallsOnOneLine extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new KeepSingleParameterCallsOnOneLineConf();
    }
    getMetadata() {
        return {
            key: "keep_single_parameter_on_one_line",
            title: "Keep single parameters on one line",
            shortDescription: `Keep single parameter calls on one line`,
            extendedInformation: `https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#keep-single-parameter-calls-on-one-line`,
            tags: [_irule_1.RuleTag.Whitespace, _irule_1.RuleTag.Styleguide, _irule_1.RuleTag.SingleFile],
            badExample: `call_method(\n  2 ).`,
            goodExample: `call_method( 2 ).`,
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file, obj) {
        let issues = [];
        if (obj.getType() === "INTF") {
            return [];
        }
        const stru = file.getStructure();
        if (stru === undefined) {
            return [];
        }
        for (const s of file.getStatements()) {
            if (this.isMultiLine(s) === false
                || this.calcStatementLength(s) > this.getConfig().length
                || this.containsNewLineValue(s)
                || this.containsNewLineTableExpression(s)
                || this.containsFieldAssigments(s)
                || this.containsNewLineTemplate(s)) {
                continue;
            }
            for (const c of s.findAllExpressions(Expressions.MethodCallParam)) {
                issues = issues.concat(this.check(c, file));
            }
        }
        return issues;
    }
    ///////////////////////////////////////
    containsFieldAssigments(s) {
        const fs = s.findAllExpressions(Expressions.FieldAssignment);
        return fs.length > 1;
    }
    containsNewLineTableExpression(s) {
        for (const st of s.findAllExpressions(Expressions.TableExpression)) {
            if (st.getFirstToken().getRow() !== st.getLastToken().getRow()) {
                return true;
            }
        }
        return false;
    }
    containsNewLineValue(s) {
        for (const st of s.findAllExpressions(Expressions.Source)) {
            const first = st.getFirstToken().getStr().toUpperCase();
            if (first === "VALUE" && st.getFirstToken().getRow() !== st.getLastToken().getRow()) {
                return true;
            }
        }
        return false;
    }
    containsNewLineTemplate(s) {
        for (const st of s.findAllExpressions(Expressions.StringTemplate)) {
            for (const t of st.getAllTokens()) {
                if (t.getStr().includes("\\n")) {
                    return true;
                }
            }
        }
        return false;
    }
    check(c, file) {
        if (this.isSingleParameter(c) === true && this.isMultiLine(c) === true) {
            for (const sub of c.findAllExpressions(Expressions.MethodCallParam)) {
                if (this.isSingleParameter(sub) === false
                    && this.isWithoutParameters(sub) === false) {
                    return [];
                }
            }
            const message = "Keep single parameter on one line";
            return [issue_1.Issue.atToken(file, c.getFirstToken(), message, this.getMetadata().key, this.conf.severity)];
        }
        return [];
    }
    // including first indentation, worst case calculation add space after each token
    calcStatementLength(c) {
        let length = 0;
        for (const t of c.getTokens()) {
            if (length === 0) {
                length = length + t.getStart().getCol();
            }
            length = length + t.getStr().length + 1;
        }
        return length;
    }
    isMultiLine(c) {
        const first = c.getFirstToken();
        const last = c.getLastToken();
        return first.getStart().getRow() < last.getStart().getRow();
    }
    isWithoutParameters(c) {
        return c.getChildren().length === 2;
    }
    isSingleParameter(c) {
        if (c.findDirectExpression(Expressions.Source)) {
            for (const params of c.findAllExpressions(Expressions.ParameterListS)) {
                if (params.getChildren().length > 1) {
                    return false;
                }
            }
            return true;
        }
        const list = c.findDirectExpression(Expressions.ParameterListS);
        if (list) {
            return list.getChildren().length === 1;
        }
        const param = c.findDirectExpression(Expressions.MethodParameters);
        if (param) {
            if (param.getChildren().length > 2) {
                return false;
            }
            const paramsource = param.findDirectExpression(Expressions.ParameterListS);
            if (paramsource && paramsource.getChildren().length === 1) {
                return true;
            }
            const paramtarget = param.findDirectExpression(Expressions.ParameterListT);
            if (paramtarget && paramtarget.getChildren().length === 1) {
                return true;
            }
        }
        return false;
    }
}
exports.KeepSingleParameterCallsOnOneLine = KeepSingleParameterCallsOnOneLine;
