// @abaplint/core/build/src/rules/line_break_multiple_parameters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LineBreakMultipleParameters = exports.LineBreakMultipleParametersConf = void 0;
const issue_1 = require("../issue");
const _abap_rule_1 = require("./_abap_rule");
const _basic_rule_config_1 = require("./_basic_rule_config");
const Expressions = require("../abap/2_statements/expressions");
const _irule_1 = require("./_irule");
const edit_helper_1 = require("../edit_helper");
const virtual_position_1 = require("../virtual_position");
class LineBreakMultipleParametersConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** Amount of allowed parameters on one line */
        this.count = 1;
    }
}
exports.LineBreakMultipleParametersConf = LineBreakMultipleParametersConf;
class LineBreakMultipleParameters extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new LineBreakMultipleParametersConf();
    }
    getMetadata() {
        return {
            key: "line_break_multiple_parameters",
            title: "Line break multiple parameters",
            shortDescription: `Line break multiple parameters`,
            extendedInformation: `https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#line-break-multiple-parameters`,
            badExample: `method( parameter1 = value parameter2 = value ).`,
            goodExample: `method( parameter1 = value\n        parameter2 = value ).`,
            tags: [_irule_1.RuleTag.Whitespace, _irule_1.RuleTag.Styleguide, _irule_1.RuleTag.Quickfix, _irule_1.RuleTag.SingleFile],
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file, obj) {
        const issues = [];
        if (obj.getType() === "INTF") {
            return [];
        }
        for (const s of file.getStatements()) {
            for (const e of s.findAllExpressions(Expressions.ParameterListS)) {
                if (s.getFirstToken().getStart() instanceof virtual_position_1.VirtualPosition) {
                    continue; // skip macro content
                }
                const parameters = e.findDirectExpressions(Expressions.ParameterS);
                if (parameters.length <= 1) {
                    continue;
                }
                let previous = parameters[0];
                for (let i = 1; i < parameters.length; i++) {
                    const current = parameters[i];
                    if (this.conf.count && i < this.conf.count) {
                        previous = current;
                        continue;
                    }
                    const first = current.getFirstToken();
                    if (previous.getFirstToken().getRow() === first.getRow()) {
                        const fix = edit_helper_1.EditHelper.insertAt(file, first.getStart(), "\n" + " ".repeat(parameters[0].getFirstToken().getStart().getCol() - 1));
                        issues.push(issue_1.Issue.atToken(file, current.getFirstToken(), this.getMetadata().title, this.getMetadata().key, this.conf.severity, fix));
                    }
                    previous = current;
                }
            }
        }
        return issues;
    }
}
exports.LineBreakMultipleParameters = LineBreakMultipleParameters;
