// @abaplint/core/build/src/rules/max_one_method_parameter_per_line.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MaxOneMethodParameterPerLine = exports.MaxOneMethodParameterPerLineConf = void 0;
const Expressions = require("../abap/2_statements/expressions");
const Statements = require("../abap/2_statements/statements");
const issue_1 = require("../issue");
const _abap_rule_1 = require("./_abap_rule");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _irule_1 = require("./_irule");
class MaxOneMethodParameterPerLineConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.MaxOneMethodParameterPerLineConf = MaxOneMethodParameterPerLineConf;
class MaxOneMethodParameterPerLine extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new MaxOneMethodParameterPerLineConf();
    }
    getMetadata() {
        return {
            key: "max_one_method_parameter_per_line",
            title: "Max one method parameter definition per line",
            shortDescription: `Keep max one method parameter description per line`,
            tags: [_irule_1.RuleTag.SingleFile, _irule_1.RuleTag.Whitespace],
            badExample: `
METHODS apps_scope_token
  IMPORTING
    body TYPE bodyapps_scope_token client_id TYPE str.`,
            goodExample: `
METHODS apps_scope_token
  IMPORTING
    body      TYPE bodyapps_scope_token
    client_id TYPE str.`,
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file) {
        var _a;
        const issues = [];
        for (const statement of ((_a = file.getStructure()) === null || _a === void 0 ? void 0 : _a.findAllStatements(Statements.MethodDef)) || []) {
            let prev = undefined;
            for (const p of statement.findAllExpressions(Expressions.MethodParam)) {
                if (prev === undefined) {
                    prev = p;
                    continue;
                }
                if (prev.getFirstToken().getStart().getRow() === p.getFirstToken().getStart().getRow()) {
                    const issue = issue_1.Issue.atToken(file, prev.getFirstToken(), this.getMetadata().title, this.getMetadata().key, this.conf.severity);
                    issues.push(issue);
                }
                prev = p;
            }
        }
        return issues;
    }
}
exports.MaxOneMethodParameterPerLine = MaxOneMethodParameterPerLine;
