// @abaplint/core/build/src/rules/mix_returning.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MixReturning = exports.MixReturningConf = void 0;
/* eslint-disable max-len */
const issue_1 = require("../issue");
const _abap_rule_1 = require("./_abap_rule");
const Statements = require("../abap/2_statements/statements");
const Expressions = require("../abap/2_statements/expressions");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _irule_1 = require("./_irule");
class MixReturningConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.MixReturningConf = MixReturningConf;
class MixReturning extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new MixReturningConf();
    }
    getMetadata() {
        return {
            key: "mix_returning",
            title: "Mix of returning and exporting",
            shortDescription: `Checks that methods don't have a mixture of returning and exporting/changing parameters`,
            extendedInformation: `https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#use-either-returning-or-exporting-or-changing-but-not-a-combination

This syntax is not allowed on versions earlier than 740sp02, https://help.sap.com/doc/abapdocu_752_index_htm/7.52/en-us/abennews-740-abap_objects.htm#!ABAP_MODIFICATION_1@1@`,
            tags: [_irule_1.RuleTag.Styleguide, _irule_1.RuleTag.SingleFile, _irule_1.RuleTag.Syntax],
            badExample: `CLASS lcl DEFINITION.
  PUBLIC SECTION.
    METHODS
      foobar
        EXPORTING foo TYPE i
        RETURNING VALUE(rv_string) TYPE string.
ENDCLASS.`,
        };
    }
    getMessage() {
        return "Don't mix RETURNING and EXPORTING/CHANGING parameters in a single method.";
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file) {
        const ret = [];
        const stru = file.getStructure();
        if (stru === undefined) {
            return [];
        }
        for (const def of stru.findAllStatements(Statements.MethodDef)) {
            if (!def.findFirstExpression(Expressions.MethodDefReturning)) {
                continue;
            }
            if (def.findFirstExpression(Expressions.MethodDefExporting)
                || def.findFirstExpression(Expressions.MethodDefChanging)) {
                const token = def.getFirstToken();
                const issue = issue_1.Issue.atToken(file, token, this.getMessage(), this.getMetadata().key, this.conf.severity);
                ret.push(issue);
            }
        }
        return ret;
    }
}
exports.MixReturning = MixReturning;
