// @abaplint/core/build/src/rules/prefer_returning_to_exporting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreferReturningToExporting = exports.PreferReturningToExportingConf = void 0;
const issue_1 = require("../issue");
const _abap_rule_1 = require("./_abap_rule");
const Statements = require("../abap/2_statements/statements");
const Expressions = require("../abap/2_statements/expressions");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _irule_1 = require("./_irule");
class PreferReturningToExportingConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.PreferReturningToExportingConf = PreferReturningToExportingConf;
class PreferReturningToExporting extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new PreferReturningToExportingConf();
    }
    getMetadata() {
        return {
            key: "prefer_returning_to_exporting",
            title: "Prefer RETURNING to EXPORTING",
            shortDescription: `Prefer RETURNING to EXPORTING. Generic types cannot be RETURNING.`,
            extendedInformation: `https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#prefer-returning-to-exporting
https://docs.abapopenchecks.org/checks/44/`,
            tags: [_irule_1.RuleTag.Styleguide, _irule_1.RuleTag.SingleFile],
            badExample: `CLASS lcl DEFINITION.
  PUBLIC SECTION.
    METHODS test EXPORTING ev_foo TYPE i.
ENDCLASS.`,
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file) {
        const ret = [];
        const stru = file.getStructure();
        if (stru === undefined) {
            return [];
        }
        for (const def of stru.findAllStatements(Statements.MethodDef)) {
            if (def.findFirstExpression(Expressions.MethodDefChanging)) {
                continue;
            }
            const exporting = def.findFirstExpression(Expressions.MethodDefExporting);
            if (exporting === undefined) {
                continue;
            }
            const returning = def.findFirstExpression(Expressions.MethodDefReturning);
            if (returning !== undefined) {
                continue;
            }
            const params = exporting.findDirectExpressions(Expressions.MethodParam);
            if (params.length !== 1) {
                continue;
            }
            const concat = params[0].concatTokens().toUpperCase();
            if (concat.endsWith("TYPE ANY")
                || concat.endsWith("TYPE ANY TABLE")
                || concat.endsWith("TYPE C")
                || concat.endsWith("TYPE CLIKE")
                || concat.endsWith("TYPE CSEQUENCE")
                || concat.endsWith("TYPE DATA")
                || concat.endsWith("TYPE DECFLOAT")
                || concat.endsWith("TYPE HASHED TABLE")
                || concat.endsWith("TYPE INDEX TABLE")
                || concat.endsWith("TYPE N")
                || concat.endsWith("TYPE NUMERIC")
                || concat.endsWith("TYPE OBJECT")
                || concat.endsWith("TYPE P")
                || concat.endsWith("TYPE SIMPLE")
                || concat.endsWith("TYPE SORTED TABLE")
                || concat.endsWith("TYPE STANDARD TABLE")
                || concat.endsWith("TYPE TABLE")
                || concat.endsWith("TYPE X")
                || concat.endsWith("TYPE XSEQUENCE")) {
                continue;
            }
            const token = params[0].getFirstToken();
            const issue = issue_1.Issue.atToken(file, token, "Prefer RETURNING to EXPORTING", this.getMetadata().key, this.conf.severity);
            ret.push(issue);
        }
        return ret;
    }
}
exports.PreferReturningToExporting = PreferReturningToExporting;
