// @abaplint/core/build/src/rules/tabl_enhancement_category.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TABLEnhancementCategory = exports.TABLEnhancementCategoryConf = void 0;
const issue_1 = require("../issue");
const _basic_rule_config_1 = require("./_basic_rule_config");
const objects_1 = require("../objects");
const position_1 = require("../position");
class TABLEnhancementCategoryConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.TABLEnhancementCategoryConf = TABLEnhancementCategoryConf;
class TABLEnhancementCategory {
    constructor() {
        this.conf = new TABLEnhancementCategoryConf();
    }
    getMetadata() {
        return {
            key: "tabl_enhancement_category",
            title: "TABL enhancement category must be set",
            shortDescription: `Checks that tables do not have the enhancement category 'not classified'.`,
            extendedInformation: `SAP note 3063227 changes the default to 'Cannot be enhanced'.

You may use standard report RS_DDIC_CLASSIFICATION_FINAL for adjustment.`,
            tags: [],
        };
    }
    getDescription(name) {
        return "TABL enhancement category not classified in " + name;
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    initialize(_reg) {
        return this;
    }
    run(obj) {
        if (!(obj instanceof objects_1.Table)) {
            return [];
        }
        if (obj.getEnhancementCategory() === objects_1.EnhancementCategory.NotClassified) {
            const position = new position_1.Position(1, 1);
            const issue = issue_1.Issue.atPosition(obj.getFiles()[0], position, this.getDescription(obj.getName()), this.getMetadata().key, this.conf.severity);
            return [issue];
        }
        return [];
    }
}
exports.TABLEnhancementCategory = TABLEnhancementCategory;
