// @abaplint/core/build/src/rules/type_form_parameters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeFormParameters = exports.TypeFormParametersConf = void 0;
const issue_1 = require("../issue");
const objects_1 = require("../objects");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _abap_rule_1 = require("./_abap_rule");
const Expressions = require("../abap/2_statements/expressions");
const _irule_1 = require("./_irule");
class TypeFormParametersConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.TypeFormParametersConf = TypeFormParametersConf;
class TypeFormParameters extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new TypeFormParametersConf();
    }
    getMetadata() {
        return {
            key: "type_form_parameters",
            title: "Type FORM parameters",
            shortDescription: `Checks for untyped FORM parameters`,
            tags: [_irule_1.RuleTag.SingleFile],
            badExample: `FORM foo USING bar.
ENDFORM.`,
            goodExample: `FORM foo USING bar TYPE string.
ENDFORM.`,
        };
    }
    getDescription(parameterName) {
        return "Add TYPE for FORM parameter \"" + parameterName + "\"";
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file, obj) {
        const ret = [];
        const stru = file.getStructure();
        if (obj instanceof objects_1.Class || stru === undefined) {
            return ret;
        }
        for (const formparam of stru.findAllExpressions(Expressions.FormParam)) {
            if (formparam.findFirstExpression(Expressions.FormParamType) === undefined) {
                const token = formparam.getFirstToken();
                const issue = issue_1.Issue.atToken(file, token, this.getDescription(token.getStr()), this.getMetadata().key, this.conf.severity);
                ret.push(issue);
            }
        }
        return ret;
    }
}
exports.TypeFormParameters = TypeFormParameters;
