// @abaplint/core/build/src/rules/types_naming.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypesNaming = exports.TypesNamingConf = void 0;
const issue_1 = require("../issue");
const _abap_rule_1 = require("./_abap_rule");
const Statements = require("../abap/2_statements/statements");
const Expressions = require("../abap/2_statements/expressions");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _irule_1 = require("./_irule");
const objects_1 = require("../objects");
class TypesNamingConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** The pattern for TYPES, case insensitive */
        this.pattern = "^TY_.+$";
    }
}
exports.TypesNamingConf = TypesNamingConf;
class TypesNaming extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new TypesNamingConf();
    }
    getMetadata() {
        return {
            key: "types_naming",
            title: "TYPES naming conventions",
            shortDescription: `Allows you to enforce a pattern for TYPES definitions`,
            extendedInformation: `Does not run for TYPE POOLS`,
            tags: [_irule_1.RuleTag.Naming, _irule_1.RuleTag.SingleFile],
            badExample: `TYPES foo TYPE i.`,
            goodExample: `TYPES ty_foo TYPE i.`,
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file, obj) {
        const issues = [];
        const testRegex = new RegExp(this.conf.pattern, "i");
        let nesting = 0;
        if (obj instanceof objects_1.TypePool) {
            return [];
        }
        for (const stat of file.getStatements()) {
            let expr = undefined;
            if (stat.get() instanceof Statements.Type && nesting === 0) {
                expr = stat.findFirstExpression(Expressions.NamespaceSimpleName);
            }
            else if (stat.get() instanceof Statements.TypeBegin
                || stat.get() instanceof Statements.TypeEnumBegin) {
                if (nesting === 0) {
                    expr = stat.findFirstExpression(Expressions.NamespaceSimpleName);
                }
                nesting = nesting + 1;
            }
            else if (stat.get() instanceof Statements.TypeEnd
                || stat.get() instanceof Statements.TypeEnumEnd) {
                nesting = nesting - 1;
                continue;
            }
            else {
                continue;
            }
            if (expr === undefined) {
                continue;
            }
            const token = expr.getFirstToken();
            if (testRegex.exec(token.getStr())) {
                continue;
            }
            else {
                const message = "Bad TYPES naming, expected \"" + this.conf.pattern + "\", got \"" + token.getStr() + "\"";
                const issue = issue_1.Issue.atToken(file, token, message, this.getMetadata().key, this.conf.severity);
                issues.push(issue);
            }
        }
        return issues;
    }
}
exports.TypesNaming = TypesNaming;
