// @abaplint/core/build/src/rules/unknown_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnknownTypes = exports.UnknownTypesConf = void 0;
const syntax_1 = require("../abap/5_syntax/syntax");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _abap_object_1 = require("../objects/_abap_object");
const issue_1 = require("../issue");
const BasicTypes = require("../abap/types/basic");
const _irule_1 = require("./_irule");
const _typed_identifier_1 = require("../abap/types/_typed_identifier");
const _scope_type_1 = require("../abap/5_syntax/_scope_type");
const _reference_1 = require("../abap/5_syntax/_reference");
const basic_1 = require("../abap/types/basic");
class UnknownTypesConf extends _basic_rule_config_1.BasicRuleConfig {
}
exports.UnknownTypesConf = UnknownTypesConf;
class UnknownTypes {
    constructor() {
        this.conf = new UnknownTypesConf();
    }
    getMetadata() {
        return {
            key: "unknown_types",
            title: "Unknown types",
            shortDescription: `Enables check for unknown data types, respects errorNamespace`,
            tags: [_irule_1.RuleTag.Syntax],
        };
    }
    initialize(reg) {
        this.reg = reg;
        return this;
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    run(obj) {
        if (!(obj instanceof _abap_object_1.ABAPObject)) {
            return [];
        }
        const spaghetti = new syntax_1.SyntaxLogic(this.reg, obj).run().spaghetti;
        const found = this.traverse(spaghetti.getTop());
        return this.removeDuplicates(found);
    }
    /////////////////////
    removeDuplicates(list) {
        const deduplicated = [];
        for (const result of list) {
            let cont = false;
            for (const d of deduplicated) {
                if (result.getStart().equals(d.getStart())) {
                    cont = true;
                    break;
                }
            }
            if (cont === true) {
                continue;
            }
            deduplicated.push(result);
        }
        return deduplicated;
    }
    traverse(node) {
        var _a;
        const ret = [];
        const nodeData = node.getData();
        for (const r of nodeData.references) {
            if (r.referenceType === _reference_1.ReferenceType.ObjectOrientedUnknownReference && ((_a = r.extra) === null || _a === void 0 ? void 0 : _a.ooName)) {
                const message = r.extra.ooName + " unknown";
                ret.push(issue_1.Issue.atIdentifier(r.position, message, this.getMetadata().key, this.conf.severity));
            }
            if (r.referenceType === _reference_1.ReferenceType.TypeReference
                && r.resolved instanceof _typed_identifier_1.TypedIdentifier
                && r.resolved.getType() instanceof basic_1.UnknownType) {
                const message = r.resolved.getType().getError();
                ret.push(issue_1.Issue.atIdentifier(r.position, message, this.getMetadata().key, this.conf.severity));
            }
        }
        if (node.getIdentifier().stype !== _scope_type_1.ScopeType.ClassImplementation) {
            const vars = nodeData.vars;
            for (const name in vars) {
                const identifier = vars[name];
                const found = this.containsUnknown(identifier.getType());
                if (found) {
                    const message = "Variable \"" + name + "\" contains unknown: " + found;
                    ret.push(issue_1.Issue.atIdentifier(identifier, message, this.getMetadata().key, this.conf.severity));
                }
            }
            const types = nodeData.types;
            for (const name in types) {
                const identifier = types[name];
                const found = this.containsUnknown(identifier.getType());
                if (found) {
                    const message = "Type \"" + name + "\" contains unknown: " + found;
                    ret.push(issue_1.Issue.atIdentifier(identifier, message, this.getMetadata().key, this.conf.severity));
                }
            }
        }
        for (const name in nodeData.idefs) {
            const v = nodeData.idefs[name];
            const found = this.checkParameters(v);
            if (found) {
                const message = "Contains unknown, " + found.found;
                ret.push(issue_1.Issue.atIdentifier(found.id, message, this.getMetadata().key, this.conf.severity));
            }
        }
        for (const name in nodeData.cdefs) {
            const v = nodeData.cdefs[name];
            const found = this.checkParameters(v);
            if (found) {
                const message = "Contains unknown, " + found.found;
                ret.push(issue_1.Issue.atIdentifier(found.id, message, this.getMetadata().key, this.conf.severity));
            }
        }
        for (const n of node.getChildren()) {
            ret.push(...this.traverse(n));
        }
        return ret;
    }
    checkParameters(idef) {
        var _a;
        for (const m of ((_a = idef.getMethodDefinitions()) === null || _a === void 0 ? void 0 : _a.getAll()) || []) {
            for (const p of m.getParameters().getAll()) {
                const found = this.containsUnknown(p.getType());
                if (found) {
                    return { id: p, found };
                }
            }
        }
        for (const e of idef.getEvents() || []) {
            for (const p of e.getParameters()) {
                const found = this.containsUnknown(p.getType());
                if (found) {
                    return { id: p, found };
                }
            }
        }
        return undefined;
    }
    containsUnknown(type) {
        if (type instanceof BasicTypes.UnknownType) {
            return type.getError();
        }
        else if (type instanceof BasicTypes.StructureType) {
            for (const c of type.getComponents()) {
                const found = this.containsUnknown(c.type instanceof _typed_identifier_1.TypedIdentifier ? c.type.getType() : c.type);
                if (found) {
                    return found;
                }
            }
        }
        else if (type instanceof BasicTypes.TableType) {
            return this.containsUnknown(type.getRowType());
        }
        return undefined;
    }
}
exports.UnknownTypes = UnknownTypes;
