// @abaplint/core/build/src/rules/unnecessary_return.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnnecessaryReturn = exports.UnnecessaryReturnConf = void 0;
const issue_1 = require("../issue");
const _abap_rule_1 = require("./_abap_rule");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _irule_1 = require("./_irule");
const Statements = require("../abap/2_statements/statements");
const edit_helper_1 = require("../edit_helper");
const _statement_1 = require("../abap/2_statements/statements/_statement");
class UnnecessaryReturnConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** Allow empty METHODs + FORMs + FUNCTION-MODULEs */
        this.allowEmpty = false;
    }
}
exports.UnnecessaryReturnConf = UnnecessaryReturnConf;
class UnnecessaryReturn extends _abap_rule_1.ABAPRule {
    constructor() {
        super(...arguments);
        this.conf = new UnnecessaryReturnConf();
    }
    getMetadata() {
        return {
            key: "unnecessary_return",
            title: "Unnecessary Return",
            shortDescription: `Finds unnecessary RETURN statements`,
            extendedInformation: `Finds unnecessary RETURN statements`,
            tags: [_irule_1.RuleTag.SingleFile, _irule_1.RuleTag.Quickfix],
            badExample: `FORM hello1.
  WRITE 'world'.
  RETURN.
ENDFORM.

FORM foo.
  IF 1 = 2.
    RETURN.
  ENDIF.
ENDFORM.`,
            goodExample: `FORM hello2.
  WRITE 'world'.
ENDFORM.`,
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
    }
    runParsed(file) {
        const issues = [];
        const structure = file.getStructure();
        if (structure === undefined) {
            return [];
        }
        const message = "Unnecessary RETURN";
        const statements = file.getStatements();
        let statementCounter = 0;
        for (let i = 0; i < statements.length; i++) {
            const node = statements[i];
            const nodeType = node.get();
            if (nodeType instanceof Statements.MethodImplementation
                || nodeType instanceof Statements.Form
                || nodeType instanceof Statements.FunctionModule) {
                statementCounter = 0;
                continue;
            }
            if (!(nodeType instanceof _statement_1.Comment)) {
                statementCounter++;
            }
            if (!(nodeType instanceof Statements.EndMethod
                || nodeType instanceof Statements.EndForm
                || nodeType instanceof Statements.EndFunction)) {
                continue;
            }
            const prev = statements[i - 1];
            if (prev
                && prev.get() instanceof Statements.Return
                && prev.getChildren().length === 2) {
                if (this.conf.allowEmpty === true && statementCounter === 2) {
                    continue;
                }
                const fix = edit_helper_1.EditHelper.deleteStatement(file, prev);
                issues.push(issue_1.Issue.atStatement(file, prev, message, this.getMetadata().key, this.getConfig().severity, fix));
            }
            // note: ENDTRY is not checked, it will usually just make it an empty catch handler, which is also bad
            const prevprev = statements[i - 2];
            if (prev && prevprev
                && prevprev.get() instanceof Statements.Return
                && prevprev.getChildren().length === 2
                && prev.get() instanceof Statements.EndIf) {
                const fix = edit_helper_1.EditHelper.deleteStatement(file, prevprev);
                issues.push(issue_1.Issue.atStatement(file, prevprev, message, this.getMetadata().key, this.getConfig().severity, fix));
            }
        }
        return issues;
    }
}
exports.UnnecessaryReturn = UnnecessaryReturn;
