// @abaplint/core/build/src/rules/unused_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnusedTypes = exports.UnusedTypesConf = void 0;
const issue_1 = require("../issue");
const _basic_rule_config_1 = require("./_basic_rule_config");
const _irule_1 = require("./_irule");
const syntax_1 = require("../abap/5_syntax/syntax");
const _abap_object_1 = require("../objects/_abap_object");
const _scope_type_1 = require("../abap/5_syntax/_scope_type");
const edit_helper_1 = require("../edit_helper");
const _reference_1 = require("../abap/5_syntax/_reference");
const _statement_1 = require("../abap/2_statements/statements/_statement");
class WorkArea {
    constructor() {
        this.workarea = [];
    }
    push(id) {
        for (const w of this.workarea) {
            if (id.equals(w)) {
                return;
            }
        }
        this.workarea.push(id);
    }
    removeIfExists(id) {
        for (let i = 0; i < this.workarea.length; i++) {
            if (id.equals(this.workarea[i])) {
                this.workarea.splice(i, 1);
                return;
            }
        }
    }
    get() {
        return this.workarea;
    }
    count() {
        return this.workarea.length;
    }
}
class UnusedTypesConf extends _basic_rule_config_1.BasicRuleConfig {
    constructor() {
        super(...arguments);
        /** skip specific names, case insensitive
         * @uniqueItems true
         */
        this.skipNames = [];
    }
}
exports.UnusedTypesConf = UnusedTypesConf;
class UnusedTypes {
    constructor() {
        this.conf = new UnusedTypesConf();
    }
    getMetadata() {
        return {
            key: "unused_types",
            title: "Unused types",
            shortDescription: `Checks for unused TYPE definitions`,
            extendedInformation: `Unused types are not reported if the object contains parser or syntax errors.`,
            tags: [_irule_1.RuleTag.Quickfix],
            pragma: "##NEEDED",
            pseudoComment: "EC NEEDED",
        };
    }
    getConfig() {
        return this.conf;
    }
    setConfig(conf) {
        this.conf = conf;
        if (this.conf.skipNames === undefined) {
            this.conf.skipNames = [];
        }
    }
    initialize(reg) {
        this.reg = reg;
        return this;
    }
    run(obj) {
        if (!(obj instanceof _abap_object_1.ABAPObject)) {
            return [];
        }
        for (const file of obj.getABAPFiles()) {
            for (const statement of file.getStatements()) {
                if (statement.get() instanceof _statement_1.Unknown) {
                    return []; // contains parser errors
                }
            }
        }
        // dont report unused variables when there are syntax errors
        const syntax = new syntax_1.SyntaxLogic(this.reg, obj).run();
        if (syntax.issues.length > 0) {
            return [];
        }
        this.workarea = new WorkArea();
        this.traverse(syntax.spaghetti.getTop(), obj, true);
        this.traverse(syntax.spaghetti.getTop(), obj, false);
        if (this.workarea.count() === 0) {
            return []; // exit early if all types are used in the current object
        }
        for (const o of this.reg.getObjects()) {
            if (o === obj) {
                continue;
            }
            else if (o instanceof _abap_object_1.ABAPObject) {
                if (this.reg.isDependency(o)) {
                    continue; // do not search in dependencies
                }
                const syntax = new syntax_1.SyntaxLogic(this.reg, o).run();
                this.traverse(syntax.spaghetti.getTop(), o, false);
            }
            if (this.workarea.count() === 0) {
                return []; // exit early if all types are used
            }
        }
        // what is left is unused
        const ret = [];
        for (const t of this.workarea.get()) {
            const message = "Type \"" + t.getName() + "\" not used";
            const file = obj.getABAPFileByName(t.getFilename());
            if (file === undefined) {
                continue;
            }
            const statement = edit_helper_1.EditHelper.findStatement(t.getToken(), file);
            if (statement === undefined) {
                continue;
            }
            if (statement.getPragmas().some(t => t.getStr() === this.getMetadata().pragma)) {
                continue;
            }
            else if (this.suppressedbyPseudo(statement, file)) {
                continue;
            }
            const fix = this.buildFix(file, statement);
            ret.push(issue_1.Issue.atIdentifier(t, message, this.getMetadata().key, this.conf.severity, fix));
        }
        return ret;
    }
    suppressedbyPseudo(statement, file) {
        if (statement === undefined) {
            return false;
        }
        if (file === undefined) {
            return false;
        }
        let next = false;
        for (const s of file.getStatements()) {
            if (next === true && s.get() instanceof _statement_1.Comment) {
                return s.concatTokens().includes(this.getMetadata().pseudoComment + "");
            }
            if (s === statement) {
                next = true;
            }
        }
        return false;
    }
    ////////////////////////////
    traverse(node, obj, add) {
        if (node.getIdentifier().stype !== _scope_type_1.ScopeType.BuiltIn) {
            this.checkNode(node, obj, add);
        }
        for (const c of node.getChildren()) {
            this.traverse(c, obj, add);
        }
    }
    checkNode(node, obj, add) {
        const ret = [];
        if (add === true) {
            const types = node.getData().types;
            for (const name in types) {
                const identifier = types[name];
                if (obj.containsFile(identifier.getFilename()) === false) {
                    continue;
                }
                else if (this.conf.skipNames
                    && this.conf.skipNames.length > 0
                    && this.conf.skipNames.some((a) => a.toUpperCase() === name)) {
                    continue;
                }
                else if (name !== identifier.getName().toUpperCase()) {
                    continue; // may have aliases via interfaces
                }
                this.workarea.push(identifier);
            }
        }
        for (const r of node.getData().references) {
            if (r.referenceType === _reference_1.ReferenceType.TypeReference && r.resolved) {
                this.workarea.removeIfExists(r.resolved);
            }
        }
        return ret;
    }
    buildFix(file, statement) {
        if (statement.concatTokens().toUpperCase().includes("BEGIN OF")) {
            return undefined;
        }
        return edit_helper_1.EditHelper.deleteStatement(file, statement);
    }
}
exports.UnusedTypes = UnusedTypes;
